---
title: "Pocillopora_microbiome_QC"
author: "Mike Connelly"
date: "2/16/2021"
output: 
  html_document:
      code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```
```{r packages}
library("tidyverse")
# extended visualization functions
source("./R/ggrare.R")
```

## Quality control plots
```{r rd}
# Read in read depth information
readdepth <- read_tsv("./outputs/qiime2/QC/per-sample-fastq-counts.tsv", col_names = c("Sample_ID", "forward_count", "reverse_count"), col_types = cols(Sample_ID = col_character(),
  forward_count = col_number(),
  reverse_count = col_number()), skip = 1)
readdepth <- left_join(sample_metadata, readdepth, by = "Sample_ID")
```

#  Read depth QC plot
```{r rd_qc}
# readdepth <- dplyr::filter(readdepth, Treatment == "control" | Treatment == "Heat" | Treatment == "Antibiotics" | Treatment == "Antibiotics.Heat" | Treatment == "positive" | Treatment == "negative")
###
###
# readdepth$Treatment <- factor(readdepth$Treatment, levels = treatment_levels, ordered = TRUE)
# readdepth <- dplyr::arrange(readdepth, Treatment, Colony)
# readdepth$SampleID <- factor(readdepth$SampleID, levels = readdepth$SampleID, ordered = TRUE)
###
ggdepth <- ggplot(readdepth, aes(Sample_ID, forward_count)) +
  geom_bar(stat = "identity", aes(color = Treatment, fill = Treatment)) +
  # scale_y_continuous(trans = 'log10') +
  # scale_color_manual(values = c(treatcolors, "black", "black")) +
  # scale_fill_manual(values = c(treatcolors, "dark gray", "light gray")) +
  ylab("16S Read Count") +
  theme(axis.text.x = element_text(angle = 90)) +
  ggtitle("Total number of 16S reads per sample")
# 
print(ggdepth)
ggsave(file = "./outputs/phyloseq_results/figures/readdepth.pdf", width = 400, height = 150,  units = "mm", device = "pdf")
```
```{r}
readdepth %>% 
  filter(forward_count > 1000) %>%
  summarise(mean = mean(forward_count),
            sd = sd(forward_count))
```

### Supplementary Figure 6
```{r rare_curve}
pdf("./outputs/figures/FigS6_rarefaction_curve.pdf", width = 6.5, height = 4.5)
# Rarefaction curve
rarefy_title <- "Rarefaction curve of 16S amplicon sequence variants"
grare <- ggrare(ps_samples, step = 100, color = "Treatment", plot = F, se = F)
grare <- grare + 
  scale_color_manual(values = c(treatcolors, rep("black", 4))) +
  scale_x_continuous(breaks = seq(0, 40000, 5000), name = "Sequence read depth (count)") +
  scale_y_continuous(breaks = seq(0, 400, 50), name = "ASVs (count)") +
  # facet_wrap(~Treatment) +
  ggtitle(rarefy_title)
print(grare)
dev.off()
```

