---
title: "phyloseq_microbiome_analysis"
author: "Mike Connelly"
date: "02/25/2021"
output: 
  html_document:
      code_folding: hide
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
# some aspects of this microbiome analysis follow: https://github.com/meyermicrobiolab/Stony-Coral-Tissue-Loss-Disease-SCTLD-Project
```
# Project and analysis setup
## Setup packages and working directories
```{r packages, include=FALSE}
library("plyr")
library("tidyverse")
# Essential microbiome analysis packages
library("phyloseq")
library("microbiome")
library("zCompositions")
library("CoDaSeq") # devtools::install_github('ggloor/CoDaSeq/CoDaSeq')
library("vegan")
library("ape")
library("ALDEx2")
# library("lefser") # BiocManager::install("lefser")
library("decontam") # devtools::install_github("benjjneb/decontam")
# library("microbiomeMarker")
library("UpSetR")
library("eulerr")
# Data visualization packages
library("pheatmap")
# Graphics packages
library("ggtree")
library("ggtreeExtra")
library("ggbiplot") # library(devtools); install_github("vqv/ggbiplot")
library("ggh4x") # devtools::install_github("teunbrand/ggh4x") 
library("plotrix")
library("cowplot")
library("patchwork")
library("gtable")
library("grid")
library("gridExtra")
library("ggthemes")
library("ggpubr")
library("ggrepel")
library("ggnewscale")
library("RColorBrewer")
library("randomcoloR")
library("wesanderson")
library("viridis")
library("circlize")
library("stringr")
library("extrafont")
library("extrafontdb")
# extended visualization functions
source("./R/ggrare.R")
source("./R/anti_phys_microbiome_functions.R")
source("./R/ismej_theme.R")
```
```{r colors, message=FALSE}
### Set overall theme, colors, and shapes for ggplot2
# colors for experimental treatments
treatcolors <- c("grey", "dodgerblue", "firebrick1")
#treatfills <- c()
shapes <- c(15:20) # may be useful for genotype variables: Gulf, Location, mtORF, symbiont, etc.
# colors for genotypes
genotype_colors <- read_csv("data/genotype_colors.csv") # color names csv corresponding to excel sheet colors
genocolors <- c(genotype_colors$color) 
# default ggplot fill
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}
genocolors <- gg_color_hue(14)
#
# null colors for presentation figures
condcolors_null <- c(rep("black", 3))
colshapes_null <- c(rep(20, 8))
# color vector for barplot facet strips
cond_col_colors <- c(treatcolors[1], genocolors,
                     treatcolors[2], genocolors,
                     treatcolors[3], genocolors)# colcolors_null
# color vector for barplot facet strips w/ gulf
# genotype_colors <- left_join(genotypes, genotype_colors, by = "Genotype")
# geno_chiriqui_colors <- genotype_colors$color[genotype_colors$Gulf == "Chiriqui"]
# geno_panama_colors <- genotype_colors$color[genotype_colors$Gulf == "Panama"]
# cond_col_colors <- c(treatcolors[1], "white", geno_chiriqui_colors, "white", geno_panama_colors,
#                      treatcolors[2], "white", geno_chiriqui_colors, "white", geno_panama_colors,
#                      treatcolors[3], "white", geno_chiriqui_colors, "white", geno_panama_colors)# colcolors_null

# read in preserved colors for taxa barplots
colors_families <- read_csv("./Rmd/taxa_colors/colors_families.csv")
# colors_families$Family
```
```{r theme}
theme_set(theme_ismej())
```

## Import sample metadata
```{r genotypes_fragments}
genotypes <- read_csv("data/genotypes.csv")
#
fragments <- read_csv("data/fragments.csv", col_types = cols(
  Fragment_ID = col_character()))
```
```{r factors}
# factor levels
treatment_levels <- c("Baseline", "Control", "Antibiotics", "FSW_Controls", "SW_Controls", "Neg_Controls", "ZYMO_Controls")
sampling_levels <- c("Day 0", "Day 7")
genotype_levels <- c("PAN-05", "PAN-10", "PAN-32", "PAN-34", "PAN-35", "PAN-37", "PAN-39", "PAN-41", "PAN-43","PAN-44", "PAN-78", "PAN-79", "PAN-83", "URA-51", NA)
gulf_levels <- c("Chiriqui", "Panama")
loc_levels <- c("Uva", "Saboga", "Uraba")
mtorf_levels <- c("Type_1", "Type_3")#, "unknown")
sym_levels <- c("C1bc", "C1d", "D1")#, "unknown")
genus_levels <- c("Cladocopium", "Durusdinium")
```

## Read in ASV table, taxonomy, sample data
```{r asv}
# Read in ASV table
ASV <- read.table("./outputs/qiime2/export_silva/all_silva_feature-table.tsv", row.names = 1, header = TRUE)
ASV <- t(ASV)
# Correct sample names for matching to other analyses
rownames(ASV) <- gsub("\\.", "-", rownames(ASV))
rownames(ASV) <- gsub("PAN-5", "PAN-05", rownames(ASV))
```
```{r taxomony}
# Read in taxononomic classification information
taxonomy <- as.matrix(read.table("./outputs/qiime2/export_silva/taxonomy4.tsv", sep = "\t", row.names = 1, header = FALSE, fill = TRUE))
colnames(taxonomy) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
taxonomy[taxonomy==""] <- NA
taxonomy <- gsub("[dpcofgs]__", "", taxonomy)
taxonomy <- gsub("^Euryarchaeota archaeon SCGC ", "", taxonomy) # remove excessively long family name for ease in plotting later
taxa_df <- data.frame(taxonomy) %>% rownames_to_column("ASV")
```
```{r tree_metadata}
# Import phylogenetic tree
tree <- read_tree("./outputs/qiime2/export_silva/tree.nwk")
# Import sample metadata 
sample_metadata <- read_delim("./data/qiime2_metadata.tsv", delim ="\t", col_names = c("Sample_ID", "Genotype", "Treatment", "Sampling_Time", "Seq_Control", "Neg_Control"), comment = "#")#, row.names = 1, header = FALSE)
sample_metadata <- sample_metadata %>% full_join(genotypes, by = "Genotype")
# Set factor orders
sample_metadata$Genotype <- factor(sample_metadata$Genotype, levels = genotype_levels, ordered = TRUE)
sample_metadata$Treatment <- factor(sample_metadata$Treatment, levels = treatment_levels, ordered = TRUE)
sample_metadata$Gulf <- factor(sample_metadata$Gulf, levels = gulf_levels, ordered = TRUE)
sample_metadata$Location <- factor(sample_metadata$Location, levels = loc_levels, ordered = TRUE)
sample_metadata$mtORF <- factor(sample_metadata$mtORF, levels = mtorf_levels, ordered = TRUE)
sample_metadata$ITS2 <- factor(sample_metadata$ITS2, levels = sym_levels, ordered = TRUE)
sample_metadata$Symbiont_Genus <- ifelse(sample_metadata$ITS2 == "D1", "Durusdinium", "Cladocopium")
# Create phyloseq-compatible sample map
map <- sample_metadata %>% column_to_rownames("Sample_ID")
```
## Create overall phyloseq object
```{r phyloseq}
ps <- phyloseq(otu_table(ASV, taxa_are_rows = FALSE),
               sample_data(map),
               tax_table(taxonomy),
               phy_tree(tree))
ps
# summarize_phyloseq(ps)
# otu_table()   OTU Table:         [ 2153 taxa and 185 samples ]
# sample_data() Sample Data:       [ 185 samples by 11 sample variables ]
# tax_table()   Taxonomy Table:    [ 2153 taxa by 7 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 2153 tips and 2124 internal nodes ]
```
```{r ps_process}
# Pre-process, filter, and QC phyloseq object
nsamples(ps) # 185
samples <- sample_names(ps)
# samples
ps <- prune_samples(grep("PAN-79_Baseline_049", samples, value = TRUE, invert = TRUE), ps)
```
## Remove contaminant ASVs and inspect sequencing controls
```{r subset_sequencing_controls}
# make phyloseq object with just sequencing controls
samples_seq_controls <- grep("_", samples, value = TRUE, invert = TRUE)
ps_seq_controls <- prune_samples(samples_seq_controls, ps)
ps_seq_controls
```
```{r decontam}
contam_df_prev <- isContaminant(ps, method="prevalence", neg="Neg_Control", threshold = 0.1)
table(contam_df_prev$contaminant)
# There are 9 contaminant ASVs that are higher in negative controls than all other samples
# 516f20f4d56455e0f3d845218c58e9cc - B. subtilis
# f9a27fcf92575fe1239dbef43e15b33f - Staphylococcus
# c2f0cd548d1a87ba171d517cfe087f1a - 
# c3342d2a9a35fb23f494d299efedfcb9 - 
# cc8eecf16bd4f5b8a05dc6d33a8b8447 - 
# 6075297080c6b17dd53eedfcb7a52289 - 
# 65fe65aa8cb2b78361e22c2836209481 - 
# 45586716398bf0a1eb30d93b9e908845 - 
# 923a96fd1debbd811788f38a5f363a37 - 
```
```{r remove_contaminants}
# remove contaminant sequences from overall phyloseq object
ps_decontam <- prune_taxa(!contam_df_prev$contaminant, ps)
# remove contaminant sequences and negative control samples from sequencing controls phyloseq
samples_seq_controls_decontam <- grep("Negative", samples_seq_controls, value = TRUE, invert = TRUE)
ps_seq_controls_decontam <- prune_taxa(!contam_df_prev$contaminant, ps_seq_controls) %>% 
prune_samples(samples_seq_controls_decontam, .)
```
```{r negative_controls}
# otu_table(ps_seq_controls_decontam) %>% t() %>% View()
```
```{r}
# ps_seq_controls_decontam_rel <- transform_sample_counts(ps_seq_controls_decontam, function(x) {x / sum(x)})
# plot_bar(ps_seq_controls_decontam_rel, fill = "Phylum")
# colors_phyla <- taxa_color_seq(ps_seq_controls_rel, taxa_rank = "Phylum")
# taxa_barplot(ps_seq_controls_rel, "Phylum", colors_phyla$Phylum, colors_phyla)
# seq_controls_ordination <- ordinate(ps_seq_controls, distance = "bray", method = "NMDS")
# plot_ordination(ps_seq_controls, seq_controls_ordination, color = "Treatment", shape = "Sampling_Time")
```
```{r zymo_positive_controls}

```
```{r seawater_controls}

```

## Remove unassigned ASVs and subset by treatment
```{r remove_unassigned_taxa}
# Retain only assigned taxa
ps_assigned <- subset_taxa(ps_decontam, Kingdom != "Unassigned") # ## unassigned taxa filtered out
ps_assigned <- subset_taxa(ps_assigned, Order != "Chloroplast") # ## taxa filtered out
ps_assigned <- subset_taxa(ps_assigned, Family != "Mitochondria") # ## taxa filtered out
ps_assigned
# otu_table()   OTU Table:         [ 1835 taxa and 184 samples ]
# sample_data() Sample Data:       [ 184 samples by 12 sample variables ]
# tax_table()   Taxonomy Table:    [ 1835 taxa by 7 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 1835 tips and 1823 internal nodes ]
```
```{r inspect_assigned_phyloseq}
ntaxa(ps_assigned) # 1,835
# 
get_taxa_unique(ps_assigned, "Phylum") %>% length() #36
get_taxa_unique(ps_assigned, "Class") %>% length() #73
get_taxa_unique(ps_assigned, "Order") %>% length() #179
get_taxa_unique(ps_assigned, "Family") %>% length() #295
get_taxa_unique(ps_assigned, "Genus") %>% length() #484
```
```{r subset_phyloseq_treatments}
# regex to select treatments 
samples_baseline <- grep("Baseline", samples, value = TRUE, invert = FALSE)
samples_control <- grep("Control", samples, value = TRUE, invert = FALSE)
samples_antibiotics <- grep("Antibiotics", samples, value = TRUE, invert = FALSE)
# prune samples to create phyloseq objects for each treatment
ps_baseline <- prune_samples(samples_baseline, ps_assigned)
ps_control <- prune_samples(samples_control, ps_assigned)
ps_antibiotics <- prune_samples(samples_antibiotics, ps_assigned)
ps_baseline
ps_control
ps_antibiotics
```
## Filter phyloseq object
```{r filter_phyloseq}
# filtered ASVs with very low abundance with phyloseq
# Filter taxa that have a mean count of 3 across all coral samples
ps_filtered <- filter_taxa(ps_assigned, function(x) mean(x) > 3, TRUE)
# 197/1835 taxa with a mean count greater than 3
```
```{r inspect_assigned_phyloseq}
ntaxa(ps_filtered) # 197
# 
get_taxa_unique(ps_filtered, "Phylum") %>% length() #13
get_taxa_unique(ps_filtered, "Class") %>% length() #21
get_taxa_unique(ps_filtered, "Order") %>% length() #58
get_taxa_unique(ps_filtered, "Family") %>% length() #80
get_taxa_unique(ps_filtered, "Genus") %>% length() #93
```
```{r other_filters}
#remove ASVs that do not occur more than 2 times in more than half the samples
# f1 <- genefilter_sample(ps, filterfun_sample(function(x) x > 2), A=0.5*nsamples(ps))
# #remove ASVs that do not occur more than 1 time in 10% of samples
# f2 <- filter_taxa(ps_assigned_AXH, function(x) sum(x > 1) > (0.1*length(x)), TRUE)
# # f2
# ps_filtered_AXH <- f2
# #remove OTU not seen more than one in at least 10% of samples
# ps_filtered <- filter_taxa(ps_assigned, function(x) sum(x > 1) > (0.1*length(x)), TRUE)
# psf2
```
```{r keep_unfiltered}
# or keep unfiltered 
ps_filtered <- ps_assigned
```

## Inspect baseline samples' bacteria communities
```{r}
ps_baseline_rel <- transform_sample_counts(ps_baseline, function(x) {x / sum(x)})
colors_phlya <- taxa_color_seq(ps_baseline_rel, taxa_rank = "Phylum")
colors_families_baseline <- taxa_color_seq(ps_baseline_rel, taxa_rank = "Family")
taxa_barplot(ps_baseline_rel, taxa_rank = "Phylum", colors_phyla$Phylum, colors_phyla)
# taxa_barplot(ps_baseline_rel, taxa_rank = "Family", colors_families$Family, colors_families)
```
```{r}
pdf("./outputs/phyloseq_results/figures/family_barplot_baseline_samples.pdf", height = 8.5, width = 11)
taxa_barplot(ps_baseline_rel, "Family", colors_families$Family, colors_families)
```
```{r codaseq_export}
sample_metadata_baseline <- sample_metadata %>%
  filter(Treatment == "Baseline") %>%
  filter(Sample_ID != "URA-51_Baseline_041")
# export otu and taxa tables from phyloseq for codaseq
otu <- as(otu_table(ps_baseline), "matrix")
taxon <- as(tax_table(ps_baseline), "matrix")
taxa_df_filtered <- data.frame(taxon) %>% rownames_to_column("ASV")
# sample_metadata <- data.frame(sample_data(ps_filtered)) %>% rownames_to_column("Sample_ID")
family <- as.character(taxon[,"Family"])
genus <- as.character(taxon[,"Genus"])
```
```{r centered_log-ratio_transform}
# Transform zero counts using the count zero multiplicative method in zCompositions, Palarea-Albaladejo and Martín-Fernández, 2015: https://doi.org/10.1016/j.chemolab.2015.02.019
# First, replace 0 values with an estimate (because normalization is taking log, can't have 0)
# Also transpose here, need samples as rows
otu_czm <- cmultRepl(t(otu), method = "CZM", label = 0) # 89,870 corrected values
# Perform the center-log-ratio (CLR) transformation, recommended by Gloor et al. 2017: https://doi.org/10.3389/fmicb.2017.02224
otu_clr <- codaSeq.clr(otu_czm)
```
### Beta diversity ordination - centered log-ratio PCA
```{r codaseq}
# transpose matrix of CLR transformed data for ordination and dendrogram
e_clr <- t(otu_clr)
# plot compositional PCA biplot (perform a singular value decomposition)
d_pcx <- prcomp(e_clr)
# inspect pca results
summary(d_pcx)
str(d_pcx)
screeplot(d_pcx)
# calculate percent variance explained for the axis labels
pc1 <- round(d_pcx$sdev[1]^2/sum(d_pcx$sdev^2),3)*100
pc2 <- round(d_pcx$sdev[2]^2/sum(d_pcx$sdev^2),3)*100
xlab <- paste("PC1: ", pc1, "%", sep="")
ylab <- paste("PC2: ", pc2, "%", sep="")
```
```{r clr_ggbiplot_genotype}
pca_clr_biplot <- d_pcx %>% 
  gg_biplot(., choices = c(1,2),
            groups = sample_metadata_baseline$haplotype,
            groups2 = sample_metadata$haplotype,
            ellipse = TRUE,
            var.axes = FALSE,
            taxlabs = "Phylum",
            varname.abbrev = F,
            varname.size = 2.6) #+
  # scale_color_manual(values = genocolors, name = "Genotype") +
  # scale_fill_manual(values = treatcolors, name = "Treatment") +
  # scale_shape_manual(values = shapes, name = "Location") +
  # guides(color = guide_legend(override.aes = list(color = treatcolors, alpha = 1, stroke = 1))) +
  # coord_fixed(xlim = c(-2, 2.5), ylim = c(-2.6, 2)
              # )
pca_clr_biplot 
# ggsave(pca_clr_biplot, filename = "./outputs/phyloseq_results/figures/clr_ggbiplot_baseline_genotype.pdf", width = 140, height = 120, units = "mm", device = "pdf")
```
#### Baseline samples distance to centroids boxplots
```{r betadisper}
# Calculate Aitchinson distance matrix
dist_clr <- dist(e_clr)
# Create factors for statistical testing
treatments <- as.character(sample_metadata_baseline$Treatment)
locations <- as.character(sample_metadata_baseline$Location)
genotypes <- as.character(sample_metadata_baseline$Genotype)
its <-  as.character(sample_metadata$ITS2)

# PERMDISP2 procedure for the analysis of multivariate homogeneity of group dispersions (variances).
mod <- betadisper(dist_clr, genotypes, type = "centroid")
mod
# 
anova(mod)
# Permutation test for F
permutest(mod, pairwise = TRUE)
(mod.HSD <- TukeyHSD(mod))
# inspect
plot(mod)
boxplot(mod)
```
```{r beta_boxplot}
# Draw a boxplot of the distances to centroid for each group
# # first extract the distances to centroids
# mod$distance
mod_distance <- data.frame(mod$distance) %>%
  rownames_to_column(., "Sample_ID") %>%
  left_join(., sample_metadata_baseline, by = "Sample_ID", .keep = TRUE)
# # 
beta_boxplot <- mod_distance %>%
  ggplot(aes(Genotype, mod.distance)) + 
  geom_boxplot(aes(fill = Genotype)) +
  geom_point(aes(fill = Genotype), color = "black", width = 0.1, shape = 21, size = 2, stroke = 1, show.legend = FALSE) +
  # facet_grid(~ ITS2, scales = "free") +
  scale_fill_manual(values = genocolors) +
  scale_color_manual(values = genocolors) +
  # scale_shape_manual(values = colshapes, name="Colony") + 
  ylim(c(0,100)) +
  xlab(NULL) +
  ylab("Beta diversity distance to centroid") +
  ggtitle("Beta diversity distance to centroid violin plot") +
  theme(legend.position = "none",
          axis.text.x = element_text(angle = 0),
        strip.background = element_blank(),
        strip.text = element_blank())
beta_boxplot
# ggsave(beta_boxplot, filename = "./outputs/phyloseq_results/figures/beta_dist-centroid_boxplot.pdf", width = 120, height = 80, units = "mm")
```
## Inspect control samples' bacteria communities
```{r}
ps_control_rel <- transform_sample_counts(ps_control, function(x) {x / sum(x)})
colors_families <- taxa_color_seq(ps_control_rel, taxa_rank = "Family")
taxa_barplot(ps_control_rel, taxa_rank = "Family", colors_families$Family, colors_families)
```
```{r}
pdf("./outputs/phyloseq_results/figures/family_barplot_control_samples.pdf", height = 8.5, width = 11)
taxa_barplot(ps_control_rel, "Family", colors_families$Family, colors_families)
```
```{r codaseq_export}
sample_metadata_control <- sample_metadata %>%
  filter(Treatment == "Control")
# export otu and taxa tables from phyloseq for codaseq
otu <- as(otu_table(ps_control), "matrix")
taxon <- as(tax_table(ps_control), "matrix")
taxa_df_filtered <- data.frame(taxon) %>% rownames_to_column("ASV")
# sample_metadata <- data.frame(sample_data(ps_filtered)) %>% rownames_to_column("Sample_ID")
family <- as.character(taxon[,"Family"])
genus <- as.character(taxon[,"Genus"])
```
```{r centered_log-ratio_transform}
# Transform zero counts using the count zero multiplicative method in zCompositions, Palarea-Albaladejo and Martín-Fernández, 2015: https://doi.org/10.1016/j.chemolab.2015.02.019
# First, replace 0 values with an estimate (because normalization is taking log, can't have 0)
# Also transpose here, need samples as rows
otu_czm <- cmultRepl(t(otu), method = "CZM", label = 0) # 93,172 corrected values
# Perform the center-log-ratio (CLR) transformation, recommended by Gloor et al. 2017: https://doi.org/10.3389/fmicb.2017.02224
otu_clr <- codaSeq.clr(otu_czm)
```
```{r codaseq}
# transpose matrix of CLR transformed data for ordination and dendrogram
e_clr <- t(otu_clr)
# plot compositional PCA biplot (perform a singular value decomposition)
d_pcx <- prcomp(e_clr)
# inspect pca results
summary(d_pcx)
str(d_pcx)
screeplot(d_pcx)
# calculate percent variance explained for the axis labels
pc1 <- round(d_pcx$sdev[1]^2/sum(d_pcx$sdev^2),3)*100
pc2 <- round(d_pcx$sdev[2]^2/sum(d_pcx$sdev^2),3)*100
xlab <- paste("PC1: ", pc1, "%", sep="")
ylab <- paste("PC2: ", pc2, "%", sep="")
```
```{r clr_ggbiplot_genotype}
pca_clr_biplot <- d_pcx %>% 
  gg_biplot(., choices = c(1,2),
            groups = sample_metadata_control$Genotype,
            # groups2 = sample_metadata$Gulf,
            ellipse = TRUE,
            var.axes = FALSE,
            taxlabs = "Phylum",
            varname.abbrev = F,
            varname.size = 2.6) +
  scale_color_manual(values = genocolors, name = "Genotype") #+
  # scale_fill_manual(values = treatcolors, name = "Treatment") +
  # scale_shape_manual(values = shapes, name = "Location") +
  # guides(color = guide_legend(override.aes = list(color = treatcolors, alpha = 1, stroke = 1))) +
  # coord_fixed(xlim = c(-2, 2.5), ylim = c(-2.6, 2)
              # )
pca_clr_biplot 
# ggsave(pca_clr_biplot, filename = "./outputs/phyloseq_results/figures/clr_ggbiplot_baseline_genotype.pdf", width = 140, height = 120, units = "mm", device = "pdf")
```
#### Control samples distance to centroids boxplots
```{r betadisper}
# Calculate Aitchinson distance matrix
dist_clr <- dist(e_clr)
# Create factors for statistical testing
treatments <- as.character(sample_metadata_baseline$Treatment)
locations <- as.character(sample_metadata_baseline$Location)
genotypes <- as.character(sample_metadata_control$Genotype)
its <-  as.character(sample_metadata$ITS2)

# PERMDISP2 procedure for the analysis of multivariate homogeneity of group dispersions (variances).
mod <- betadisper(dist_clr, genotypes, type = "centroid")
mod
# 
anova(mod)
# Permutation test for F
permutest(mod, pairwise = TRUE)
(mod.HSD <- TukeyHSD(mod))
# inspect
plot(mod)
boxplot(mod)
```
```{r beta_boxplot}
# Draw a boxplot of the distances to centroid for each group
# # first extract the distances to centroids
# mod$distance
mod_distance <- data.frame(mod$distance) %>%
  rownames_to_column(., "Sample_ID") %>%
  left_join(., sample_metadata_control, by = "Sample_ID", .keep = TRUE)
# # 
beta_boxplot <- mod_distance %>%
  ggplot(aes(Genotype, mod.distance)) + 
  geom_boxplot(aes(fill = Genotype)) +
  geom_point(aes(fill = Genotype), color = "black", width = 0.1, shape = 21, size = 2, stroke = 1, show.legend = FALSE) +
  # facet_grid(~ ITS2, scales = "free") +
  scale_fill_manual(values = genocolors) +
  scale_color_manual(values = genocolors) +
  # scale_shape_manual(values = colshapes, name="Colony") + 
  ylim(c(0,100)) +
  xlab(NULL) +
  ylab("Beta diversity distance to centroid") +
  ggtitle("Beta diversity distance to centroid violin plot") +
  theme(legend.position = "none",
          axis.text.x = element_text(angle = 0),
        strip.background = element_blank(),
        strip.text = element_blank())
beta_boxplot
# ggsave(beta_boxplot, filename = "./outputs/phyloseq_results/figures/beta_dist-centroid_boxplot.pdf", width = 120, height = 80, units = "mm")
```
## Inspect antibiotics samples' bacteria communities
```{r}
ps_antibiotics_rel <- transform_sample_counts(ps_antibiotics, function(x) {x / sum(x)})
colors_families_anti <- taxa_color_seq(ps_antibiotics_rel, taxa_rank = "Family")
colors_genera_anti <- taxa_color_seq(ps_antibiotics_rel, taxa_rank = "Genus")
taxa_barplot(ps_antibiotics_rel, taxa_rank = "Family", colors_families$Family, colors_families)
```
```{r}
pdf("./outputs/phyloseq_results/figures/family_barplot_antibiotics_samples.pdf", height = 8.5, width = 11)
taxa_barplot(ps_antibiotics_rel, "Family", colors_families$Family, colors_families)
taxa_barplot(ps_antibiotics_rel, "Genus", colors_genera_anti$Genus, colors_genera_anti)
```
```{r codaseq_export}
sample_metadata_antibiotics <- sample_metadata %>%
  filter(Treatment == "Antibiotics")
# export otu and taxa tables from phyloseq for codaseq
otu <- as(otu_table(ps_antibiotics), "matrix")
taxon <- as(tax_table(ps_antibiotics), "matrix")
taxa_df_filtered <- data.frame(taxon) %>% rownames_to_column("ASV")
# sample_metadata <- data.frame(sample_data(ps_filtered)) %>% rownames_to_column("Sample_ID")
family <- as.character(taxon[,"Family"])
genus <- as.character(taxon[,"Genus"])
```
```{r centered_log-ratio_transform}
# Transform zero counts using the count zero multiplicative method in zCompositions, Palarea-Albaladejo and Martín-Fernández, 2015: https://doi.org/10.1016/j.chemolab.2015.02.019
# First, replace 0 values with an estimate (because normalization is taking log, can't have 0)
# Also transpose here, need samples as rows
otu_czm <- cmultRepl(t(otu), method = "CZM", label = 0) # 98,547 corrected values
# Perform the center-log-ratio (CLR) transformation, recommended by Gloor et al. 2017: https://doi.org/10.3389/fmicb.2017.02224
otu_clr <- codaSeq.clr(otu_czm)
```
```{r codaseq}
# transpose matrix of CLR transformed data for ordination and dendrogram
e_clr <- t(otu_clr)
# plot compositional PCA biplot (perform a singular value decomposition)
d_pcx <- prcomp(e_clr)
# inspect pca results
summary(d_pcx)
str(d_pcx)
screeplot(d_pcx)
# calculate percent variance explained for the axis labels
pc1 <- round(d_pcx$sdev[1]^2/sum(d_pcx$sdev^2),3)*100
pc2 <- round(d_pcx$sdev[2]^2/sum(d_pcx$sdev^2),3)*100
xlab <- paste("PC1: ", pc1, "%", sep="")
ylab <- paste("PC2: ", pc2, "%", sep="")
```
```{r clr_ggbiplot_genotype}
pca_clr_biplot <- d_pcx %>% 
  gg_biplot(., choices = c(1,2),
            groups = sample_metadata_antibiotics$Gulf,
            # groups2 = sample_metadata$Gulf,
            ellipse = TRUE,
            var.axes = FALSE,
            taxlabs = "Phylum",
            varname.abbrev = F,
            varname.size = 2.6) +
  scale_color_manual(values = genocolors, name = "Genotype") #+
  # scale_fill_manual(values = treatcolors, name = "Treatment") +
  # scale_shape_manual(values = shapes, name = "Location") +
  # guides(color = guide_legend(override.aes = list(color = treatcolors, alpha = 1, stroke = 1))) +
  # coord_fixed(xlim = c(-2, 2.5), ylim = c(-2.6, 2)
              # )
pca_clr_biplot 
# ggsave(pca_clr_biplot, filename = "./outputs/phyloseq_results/figures/clr_ggbiplot_baseline_genotype.pdf", width = 140, height = 120, units = "mm", device = "pdf")
```
#### Antibiotics samples distance to centroids boxplots
```{r betadisper}
# Calculate Aitchinson distance matrix
dist_clr <- dist(e_clr)
# Create factors for statistical testing
treatments <- as.character(sample_metadata_baseline$Treatment)
locations <- as.character(sample_metadata_baseline$Location)
genotypes <- as.character(sample_metadata_control$Genotype)
its <-  as.character(sample_metadata$ITS2)

# PERMDISP2 procedure for the analysis of multivariate homogeneity of group dispersions (variances).
mod <- betadisper(dist_clr, genotypes, type = "centroid")
mod
# 
anova(mod)
# Permutation test for F
permutest(mod, pairwise = TRUE)
(mod.HSD <- TukeyHSD(mod))
# inspect
plot(mod)
boxplot(mod)
```
```{r beta_boxplot}
# Draw a boxplot of the distances to centroid for each group
# # first extract the distances to centroids
# mod$distance
mod_distance <- data.frame(mod$distance) %>%
  rownames_to_column(., "Sample_ID") %>%
  left_join(., sample_metadata_antibiotics, by = "Sample_ID", .keep = TRUE)
# # 
beta_boxplot <- mod_distance %>%
  ggplot(aes(Genotype, mod.distance)) + 
  geom_boxplot(aes(fill = Genotype)) +
  geom_point(aes(fill = Genotype), color = "black", width = 0.1, shape = 21, size = 2, stroke = 1, show.legend = FALSE) +
  # facet_grid(~ ITS2, scales = "free") +
  scale_fill_manual(values = genocolors) +
  scale_color_manual(values = genocolors) +
  # scale_shape_manual(values = colshapes, name="Colony") + 
  ylim(c(0,100)) +
  xlab(NULL) +
  ylab("Beta diversity distance to centroid") +
  ggtitle("Beta diversity distance to centroid violin plot") +
  theme(legend.position = "none",
          axis.text.x = element_text(angle = 0),
        strip.background = element_blank(),
        strip.text = element_blank())
beta_boxplot
# ggsave(beta_boxplot, filename = "./outputs/phyloseq_results/figures/beta_dist-centroid_boxplot.pdf", width = 120, height = 80, units = "mm")
```

# Analysis of Baseline, Control, and Antibiotics treatments
```{r}
samples_only <- grep("_", samples, value = TRUE, invert = FALSE)
ps_filtered_samples <- prune_samples(samples_only, ps_filtered)
ps_samples <- prune_samples(samples_only, ps)
#
samples_baselinecontrol <- grep("Antibiotics", samples_only, value = TRUE, invert = TRUE)
ps_filtered_baselinecontrol <- prune_samples(samples_baselinecontrol, ps_filtered)
```
## Compositional Data Analysis (CoDA) 
```{r codaseq_export}
# export otu and taxa tables from phyloseq for codaseq
otu <- as(otu_table(ps_filtered_samples), "matrix")
taxon <- as(tax_table(ps_filtered_samples), "matrix")
#
taxa_df_filtered <- data.frame(taxon) %>% rownames_to_column("ASV")
taxa_df_filtered$Genus[is.na(taxa_df_filtered$Genus)] <- "unclassified"
#
sample_metadata <- data.frame(sample_data(ps_filtered_samples)) %>% rownames_to_column("Sample_ID")
#
family <- as.character(taxon[,"Family"])
genus <- as.character(taxon[,"Genus"])
```
```{r centered_log-ratio_transform}
# Transform zero counts using the count zero multiplicative method in zCompositions, Palarea-Albaladejo and Martín-Fernández, 2015: https://doi.org/10.1016/j.chemolab.2015.02.019
# First, replace 0 values with an estimate (because normalization is taking log, can't have 0)
# Also transpose here, need samples as rows
otu_czm <- cmultRepl(t(otu), method = "CZM", label = 0) # 279,747 corrected values unfiltered, 13832 filtered 
# Perform the center-log-ratio (CLR) transformation, recommended by Gloor et al. 2017: https://doi.org/10.3389/fmicb.2017.02224
otu_clr <- codaSeq.clr(otu_czm)
```
### Beta diversity ordination - centered log-ratio PCA
```{r codaseq}
# transpose matrix of CLR transformed data for ordination and dendrogram
e_clr <- t(otu_clr)
# plot compositional PCA biplot (perform a singular value decomposition)
d_pcx <- prcomp(e_clr)
# inspect pca results
# summary(d_pcx)
# str(d_pcx)
# screeplot(d_pcx)
# calculate percent variance explained for the axis labels
pc1 <- round(d_pcx$sdev[1]^2/sum(d_pcx$sdev^2),3)*100
pc2 <- round(d_pcx$sdev[2]^2/sum(d_pcx$sdev^2),3)*100
xlab <- paste("PC1: ", pc1, "%", sep="")
ylab <- paste("PC2: ", pc2, "%", sep="")
```
```{r clr_pca_biplot}
# pdf("./outputs/phyloseq_results/figures/CoDA_PCA_biplot.pdf")
biplot(d_pcx, cex=c(0.6,0.4), var.axes=F,scale=1, xlab=xlab, ylab=ylab, ylabs=family)
ggbiplot(d_pcx,
         choices = c(1,2),
         # groups = sample_metadata$Sample_ID,
         varname.abbrev = T) #+ scale_color_manual(values = c(treatcolors, rep("black", 4)))
```
```{r clr_ggbiplot}
pca_clr_biplot <- d_pcx %>% 
  gg_biplot(., choices = c(1,2),
            groups = sample_metadata$Treatment,
            groups2 = sample_metadata$Treatment,
            ellipse = TRUE,
            var.axes = TRUE,
            taxlabs = "Family",
            varname.abbrev = F,
            varname.size = 2.6) +
  scale_color_manual(values = treatcolors[1:3], name = "Treatment") +
  scale_fill_manual(values = treatcolors[1:3], name = "Treatment") +
  # scale_shape_manual(values = shapes, name = "Location") +
  ggtitle("Bacteria communities PCA plot") +
  guides(color = guide_legend(override.aes = list(color = treatcolors, alpha = 1, stroke = 1)))
#
pca_clr_biplot
ggsave(pca_clr_biplot, filename = "./outputs/phyloseq_results/figures/clr_ggbiplot_decontam_filtered_families.pdf", width = 6.5, height = 5, units = "in", device = "pdf")
```
```{r clr_ggbiplot_treatment_genotype}
pca_clr_biplot <- d_pcx %>% 
  gg_biplot(., choices = c(1,2),
            groups = sample_metadata$mtORF,
            groups2 = sample_metadata$Genotype,
            ellipse = TRUE,
            var.axes = TRUE,
            taxlabs = "Class",
            varname.abbrev = F,
            varname.size = 2.6) #+
  # scale_color_manual(values = treatcolors[1:3], name = "Treatment") +
  # scale_fill_manual(values = genocolors, name = "Genotype") +
  # scale_shape_manual(values = shapes, name = "Location") +
  # ggtitle("PCA plot of bacteria community beta diversity") +
  # guides(color = guide_legend(override.aes = list(color = treatcolors, alpha = 1, stroke = 1)))
#
pca_clr_biplot
# ggsave(pca_clr_biplot, filename = "./outputs/phyloseq_results/figures/clr_ggbiplot_treatment_genotype.pdf", width = 140, height = 120, units = "mm", device = "pdf")
```
### Figure 3A
```{r figure 3A: powerpoint}
pdf("./outputs/figures/bacteria_pca_clr_decontam_no-bones.pdf", height = 4.5, width = 6.5)
plot_title <- expression(paste(italic("Pocillopora"), ' bacteria communities PCA plot', sep = ""))
pca_clr_biplot <- d_pcx %>% 
  gg_biplot(., choices = c(1,2),
            groups = sample_metadata$Treatment,
            groups2 = sample_metadata$Treatment,
            ellipse = TRUE,
            var.axes = FALSE,
            taxlabs = "Genus",
            varname.abbrev = F,
            varname.size = 2.6) +
  scale_color_manual(values = treatcolors[1:3], name = "Treatment") +
  scale_fill_manual(values = treatcolors[1:3], name = "Treatment") +
  theme(plot.title = element_text(size = 16),
        strip.text = element_text(size = 14),
        axis.title = element_text(size = 14),
        axis.text = element_text(size = 12),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 12)) + 
  # scale_shape_manual(values = shapes, name = "Location") +
  ggtitle(plot_title) +
  guides(color = guide_legend(override.aes = list(color = treatcolors, alpha = 1, stroke = 1)))
#
pca_clr_biplot
dev.off()
```
```{r figure 3A: manuscript}
pdf("./outputs/figures/Fig3A_bacteria_pca_clr_decontam.pdf", height = 88/25.4, width = 88/25.4) # two-column max size
plot_title <- expression(paste(italic("Pocillopora"), ' bacteria communities PCA', sep = ""))
pca_clr_biplot <- d_pcx %>% 
  gg_biplot(., choices = c(1,2),
            groups = sample_metadata$Treatment,
            groups2 = sample_metadata$Treatment,
            convexhull = TRUE,
            centroidlabs = TRUE,
            ellipse = FALSE,
            var.axes = FALSE,
            taxlabs = "Genus",
            varname.abbrev = F,
            varname.size = 2.6) +
  scale_color_manual(values = treatcolors[1:3], name = "Treatment") +
  scale_fill_manual(values = treatcolors[1:3], name = "Treatment") +
  # scale_shape_manual(values = shapes, name = "Location") +
  ggtitle(plot_title) +
  guides(color = guide_legend(override.aes = list(color = treatcolors, alpha = 1, stroke = 1)))
#
pca_clr_biplot
dev.off()
```

### Beta diversity statistical testing
```{r distance_matrix}
# Calculate Aitchinson distance matrix
dist_clr <- dist(e_clr)
# The Aitchison distance is superior to both the widely used Jensen-Shannon divergence and the Bray-Curtis dissimilarity metrics, being more stable to subsetting and aggregating of the data, and being a true linear distance (Gloor et al. 2017)
# Create factors for statistical testing
treatments <- as.character(sample_metadata$Treatment)
gulfs <- as.character(sample_metadata$Gulf)
locations <- as.character(sample_metadata$Location)
genotypes <- as.character(sample_metadata$Genotype)
mtorfs <- as.character(sample_metadata$mtORF)
genus <-  as.character(sample_metadata$Symbiont_Genus)
```
#### ANOSIM
```{r anosim}
# anosim between groups using Aitchison distance
# Treatment
anosim_treatment <- anosim(dist_clr, treatments, permutations = 999)
anosim_treatment
plot(anosim_treatment)
# Reef
anosim_location <- anosim(dist_clr, locations, permutations = 999)
plot(anosim_location)
# Colony
anosim_genotype <- anosim(dist_clr, genotypes, permutations = 999)
plot(anosim_genotype)
# Symbiont
anosim_sym <- anosim(dist_clr, genus, permutations = 999)
plot(anosim_sym)
# All ANOSIM tests for each factor yield significant results
```
#### PERMDISP
```{r permdisp}
# PERMDISP tests
beta_treatment <- betadisper(dist_clr, treatments)
permutest(beta_treatment)
beta_treatment <- betadisper(dist_clr, gulfs)
permutest(beta_treatment)
beta_location <- betadisper(dist_clr, locations)
permutest(beta_location)
beta_colony <- betadisper(dist_clr, genotypes)
permutest(beta_colony)
beta_colony <- betadisper(dist_clr, mtorfs)
permutest(beta_colony)
beta_colony <- betadisper(dist_clr, genus)
permutest(beta_colony)
# All PERMDISP tests reject hypothesis for homogeneity of multivariate dispersions
```
#### PERMANOVA
```{r permanova}
# Adonis/PERMANOVA tests
# Treatment
permanova_treatment <- adonis(dist_clr ~ Treatment, data = sample_metadata)
print(permanova_treatment)
# 
permanova_site <- adonis(dist_clr ~ locations, data = sample_metadata)
print(permanova_site)
# 
permanova_colony <- adonis(dist_clr ~ genotypes, data = sample_metadata)
print(permanova_colony)
#
permanova_mtorf <- adonis(dist_clr ~ mtorfs, data = sample_metadata)
print(permanova_colony)
#
permanova_sym <- adonis(dist_clr ~ genus, data = sample_metadata)
print(permanova_sym)
```
```{r pairwise_permanova}
# Pairwise Adonis/PERMANOVA tests
# Treatment
pairwise_permanova_treatment <- pairwise.adonis(dist_clr, treatments)
print(pairwise_permanova_treatment)
# write_csv(pairwise_permanova_treatment, "./outputs/pairwise_permanova_treatment_symbiont.csv")
# Site
pairwise_permanova_site <- pairwise.adonis(dist_clr, locations)
print(pairwise_permanova_site)
# Colony (Genotype)
pairwise_permanova_colony <- pairwise.adonis(dist_clr, genotypes)
print(pairwise_permanova_colony)
# write_csv(pairwise_permanova_colony, "./outputs/pairwise_permanova_colony_symbiont.csv")
```
#### Distances to centroids boxplots
```{r betadisper}
# PERMDISP2 procedure for the analysis of multivariate homogeneity of group dispersions (variances).
mod <- betadisper(dist_clr, treatments, type = "centroid")
mod
# 
anova(mod)
# Permutation test for F
permutest(mod, pairwise = TRUE)
(mod.HSD <- TukeyHSD(mod))
# inspect
plot(mod)
boxplot(mod)
```
### Figure 3C
```{r beta_boxplot}
# # first extract the distances to centroids
# mod$distance
mod_distance <- data.frame(mod$distance) %>%
  rownames_to_column(., "Sample_ID") %>%
  left_join(., sample_metadata, by = "Sample_ID", .keep = TRUE)
```
```{r figure 3C: powerpoint}
pdf("./outputs/figures/beta_dist-centroid_boxplot_no-bones.pdf", height = 2.25, width = 4)
beta_title <- expression(Bacteria~community~beta*-diversity)
# Draw a boxplot of the distances to centroid for each group
beta_boxplot <- mod_distance %>%
  ggplot(aes(Treatment, mod.distance)) + 
  geom_violin(aes(fill = Treatment), alpha = 0.5) +
  geom_jitter(aes(fill = Treatment), color = "black", width = 0.05, shape = 21, size = 2, alpha = 1, stroke = 0.25, show.legend = TRUE) +
  scale_fill_manual(values = treatcolors) +
  scale_color_manual(values = genocolors) +
  #ylim(c(0,100)) +
  xlab(NULL) +
  ylab("Distance to centroid") +
  coord_cartesian(ylim = c(0,125)) +
  ggtitle(beta_title) +
  theme(plot.title = element_text(size = 16),
        strip.text = element_text(size = 14),
        axis.title = element_text(size = 14),
        axis.text = element_text(size = 12),
        axis.text.x = element_text(angle = 0),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 12),
        legend.position = "none",
        strip.background = element_blank())
beta_boxplot
dev.off()
```
```{r figure 3C: manuscript}
pdf("./outputs/figures/Fig3C_beta_dist-centroid_boxplot.pdf", height = 2.25, width = 4)
beta_title <- expression(Bacteria~community~beta*-diversity)
# Draw a boxplot of the distances to centroid for each group
beta_boxplot <- mod_distance %>%
  ggplot(aes(Treatment, mod.distance)) + 
  geom_violin(aes(fill = Treatment), alpha = 0.5) +
  geom_jitter(aes(fill = Treatment), color = "black", width = 0.05, shape = 21, size = 2, alpha = 1, stroke = 0.25, show.legend = TRUE) +
  scale_fill_manual(values = treatcolors) +
  scale_color_manual(values = genocolors) +
  #ylim(c(0,100)) +
  xlab(NULL) +
  ylab("Distance to centroid") +
  coord_cartesian(ylim = c(0,125)) +
  ggtitle(beta_title) +
  theme(legend.position = "none",
        strip.background = element_blank()) 
beta_boxplot
dev.off()
```
```{r beta_kruskal_test}
kruskal.test(data=mod_distance, mod.distance ~ Treatment)
```
## Taxa composition
```{r relative_abundance_transform}
# Create phyloseq object with relative abundances
# Important for many downstream analyses - barplots, core microbiome bubbleplots, etc.
ps_filtered_rel <- transform_sample_counts(ps_filtered_samples, function(x) {x / sum(x)})
ps_filtered_rel
#
ps_filtered_baselinecontrol_rel <- transform_sample_counts(ps_filtered_baselinecontrol, function(x) {x / sum(x)})
```

### Taxa composition summaries
```{r}

toptaxa_summary <- function(physeq, taxa_rank) {
  glom <- tax_glom(physeq, taxrank = taxa_rank)
  psmelted <- psmelt(glom)
  psmelted[, taxa_rank] <- as.character(psmelted[ , taxa_rank])
  # psmelted[psmelted$Abundance < 0.01, taxa_rank] <- "<1% abundance"
  # psmelted[psmelted$Abundance < 0.05, taxa_rank] <- "<5% abundance"
  #
  taxa_rank_name <- sym(taxa_rank)
  toptaxa_summary_df <- psmelted %>% 
    dplyr::group_by({{ taxa_rank_name }}) %>% 
    dplyr::summarise(total_abundance = sum(Abundance)) %>% 
    dplyr::mutate(total_relative_abundance = as.numeric(total_abundance/sum(total_abundance))) %>% 
    arrange(desc(total_relative_abundance))
    return(toptaxa_summary_df)
}
  
```
```{r}
phyla_baselinecontrol <- toptaxa_summary(ps_filtered_baselinecontrol, "Phylum")
#
orders_baselinecontrol <- toptaxa_summary(ps_filtered_baselinecontrol, "Order")
families_baselinecontrol <- toptaxa_summary(ps_filtered_baselinecontrol, "Family")
genera_baselinecontrol <- toptaxa_summary(ps_filtered_baselinecontrol, "Genus")
```
```{r}
phyla_antibiotics <- toptaxa_summary(ps_antibiotics, "Phylum")
families_antibiotics <- toptaxa_summary(ps_antibiotics, "Family")
```


```{r taxa colors, include=FALSE}
# run below taxa colors command until you get a set of colors you like, then comment out and write to csv file for custom edits!
# then read in csv to preserve colors across sessions

# # Bacteria taxa
# # Phyla
colors_phyla <- taxa_color_seq(ps_filtered_rel, taxa_rank = "Phylum") 
write_csv(colors_phyla, "./Rmd/taxa_colors/colors_phyla.csv")
# # Classes
colors_classes <- taxa_color_seq(ps_filtered_rel, taxa_rank = "Class") 
write_csv(colors_classes, "./Rmd/taxa_colors/colors_class.csv")
# # Orders
colors_orders <- taxa_color_seq(ps_filtered_rel, taxa_rank = "Order")
write_csv(colors_orders, "./Rmd/taxa_colors/colors_order.csv")
# # Families
colors_families <- taxa_color_seq(ps_filtered_rel, taxa_rank = "Family")
# write_csv(colors_families, "./Rmd/taxa_colors/colors_families.csv")
# # Genera
colors_genera <- taxa_color_seq(ps_filtered_rel, taxa_rank = "Genus")
write_csv(colors_genera, "./Rmd/taxa_colors/colors_genera.csv")
# # Specices
colors_species <- taxa_color_seq(ps_filtered_AXH_rel, taxa_rank = "Species")
write_csv(colors_species, "./Rmd/taxa_colors/colors_species.csv")
```
### Taxa barplots
```{r taxa_barplots}
taxa_barplot(ps_filtered_rel, "Phylum", colors_phyla$Phylum, colors_phyla)
taxa_barplot(ps_filtered_rel, "Class", colors_classes$Class, colors_classes)
taxa_barplot(ps_filtered_rel, "Order", colors_orders$Order, colors_orders)
taxa_barplot(ps_filtered_rel, "Family", colors_families$Family, colors_families)
taxa_barplot(ps_filtered_rel, "Genus", colors_genera$Genus, colors_genera)
taxa_barplot(ps_filtered_rel, "Species", colors_species$Species, colors_species)
```

```{r family_barplot_3B}
pdf("./outputs/phyloseq_results/figures/family_barplot.pdf", height = 8.5, width = 18)
taxa_barplot(ps_filtered_rel, "Family", colors_families$Family, colors_families)
# pdf("./manuscript_figures/Fig3C_family_barplot.pdf", height = 3.5, width = 6.5)
# taxa_barplot(ps_filtered_AXH_rel, "Family", colors_families$Family, colors_families)
# pdf("./manuscript_figures/MBE702_family_barplot.pdf", height = 4.3, width = 6.1)
# taxa_barplot(ps_filtered_AXH_rel, "Family", colors_families$Family, colors_families)
```
```{r}
colors_families_defense <- read_csv("./Rmd/taxa_colors/colors_families_defense_edit.csv")
colors_families_defense$Family[2:66] <- stringr::str_c(" ", colors_families_defense$Family[2:66])
colors_families$Family == colors_families_defense$Family
colors_families_defense$hex <- gplots::col2hex(colors_families_defense$colors)
```
```{r}
# taxa_barplot_simple(ps_filtered_rel, "Phylum", colors_phyla$Phylum, colors_phyla)
# pdf("./outputs/phyloseq_results/figures/family_barplot_simple_defense.pdf", height = 7, width = 10)
pdf("./outputs/phyloseq_results/figures/family_barplot_simple_dissertation.pdf", height = 10, width = 9)
taxa_barplot_simple(ps_filtered_rel, "Family", colors_families$Family, colors_families) + ggtitle(title)
dev.off()
```
### Supplementary Figure 7
```{r}
pdf("./outputs/figures/FigS7_family_barplot.pdf", width = 9, height = 10)
title <- expression(paste("Relative abundance of ", italic("Pocillopora"), "-associated bacteria families across treatments", collapse = ""))
# 
figS7 <- taxa_barplot_simple(ps_filtered_rel, "Family", colors_families$Family, colors_families) + ggtitle(title)
# 
print(figS7)
# 
dev.off()
```

## Alpha diversity plots and statistics
```{r estimate_richness}
# Use unfiltered, unrarefied dataset for alpha diversity metrics calculations
ps_richness <- estimate_richness(ps_filtered_samples) %>%
  rownames_to_column("Sample_ID")
ps_richness$Sample_ID <- str_replace_all(ps_richness$Sample_ID, "\\.", "-")

alphameasures <- c("Observed", "Chao1", "Shannon", "Simpson")
```
### Overall alpha diversity boxplots
```{r alpha_boxplots}
alpha_box <- ps_filtered_samples %>% plot_richness(., x="Treatment", measures = alphameasures) +
  geom_boxplot(aes(fill = Treatment)) +
  # scale_fill_manual(values = condfillcolors_AxH) +
  # new_scale_fill() + 
  geom_point(aes(fill = Treatment), size = 1, alpha = 0.8) +
  # geom_point(size = 3, aes(fill = Treatment), stroke = 0.5, show.legend = FALSE) +
  scale_color_manual(values = treatcolors) +
  scale_fill_manual(values = treatcolors)
  # scale_shape_manual(values = colshapes, name="Colony")
alpha_box
# ggsave(alpha_box, filename = "./outputs/phyloseq_results/figures/alpha_boxplot.pdf", width = 220, height = 150, units = "mm", device = "pdf")
```
### Figure 3B
```{r figure 3B: powerpoint}
pdf("./outputs/figures/alpha_chao1_boxplot_no-bones.pdf", height = 2.25, width = 4)
alpha_title <- expression(Bacteria~community~alpha*-diversity)
obs_boxplot <- ps_filtered_samples %>% plot_richness(., x="Treatment", measures=c("Observed")) +
  geom_violin(aes(fill = Treatment), alpha = 0.5) +
  geom_jitter(aes(fill = Treatment), color = "black", width = 0.05, shape = 21, size = 2, alpha = 1, stroke = 0.25, show.legend = TRUE) +
  scale_color_manual(values = genocolors) +
  scale_fill_manual(values = treatcolors) +
  # scale_shape_manual(values = colshapes, name="Colony") + 
  # ylim(c(0,20)) +
  xlab(NULL) +
  ylab("Chao1 ASV richness") +
  ggtitle(alpha_title) +
  theme_bw() +
  theme(plot.title = element_text(size = 16),
        axis.title = element_text(size = 14),
        axis.text = element_text(size = 12),
        axis.text.x = element_text(angle = 0),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 12),
        legend.position = "none",
        strip.background = element_blank(),
        strip.text = element_blank()) 
obs_boxplot
dev.off()
```
```{r figure 3B: manuscript}
pdf("./outputs/figures/Fig3B_alpha_chao1_boxplot.pdf", height = 2.25, width = 4)
alpha_title <- expression(Bacteria~community~alpha*-diversity)
obs_boxplot <- ps_filtered_samples %>% plot_richness(., x="Treatment", measures=c("Observed")) +
  geom_violin(aes(fill = Treatment), alpha = 0.5) +
  geom_jitter(aes(fill = Treatment), color = "black", width = 0.05, shape = 21, size = 2, alpha = 1, stroke = 0.25, show.legend = TRUE) +
  scale_color_manual(values = genocolors) +
  scale_fill_manual(values = treatcolors) +
  # scale_shape_manual(values = colshapes, name="Colony") + 
  # ylim(c(0,20)) +
  xlab(NULL) +
  ylab("Chao1 ASV richness") +
  ggtitle(alpha_title) +
  theme(plot.title = element_text(size = 12),
        strip.text = element_blank(),
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5),
        legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        legend.position = "none",
        strip.background = element_blank()) 
obs_boxplot
dev.off()
```

### Figure 3ABC
```{r figure 3: patchwork}
figure_3 <- (pca_clr_biplot | obs_boxplot / beta_boxplot)
# Need to add plot annotation tags (ie. A,B,C in Inkscape)
pdf("./outputs/figures/Fig3_ABC.pdf", height = 88.9/25.4, width = 175/25.4)
print(figure_3)
dev.off()
```

### Statistical testing of alpha diversity differences
```{r richness_data}
# make combined table with richness estimates and sample data
ps_richness_meta <- left_join(ps_richness, sample_metadata, by = "Sample_ID")
```
```{r}
# Kruskal-Wallis testing for treatment effect
kruskal.test(data=ps_richness_meta, Observed ~ Treatment)
kruskal.test(data=ps_richness_meta, Chao1 ~ Treatment)
kruskal.test(data=ps_richness_meta, Shannon ~ Treatment)
kruskal.test(data=ps_richness_meta, Simpson ~ Treatment)
```
```{r}
se <- function(x) sd(x) / sqrt(length(x))
ps_richness_meta %>% 
  dplyr::group_by(Treatment) %>% 
  dplyr::summarise(mean_obs = mean(Observed),
            sd_obs = sd(Observed))
```
```{r}
# Kruskal-Wallis testing for reef effect
kruskal.test(data=ps_richness_meta, Observed ~ Reef)
kruskal.test(data=ps_richness_meta, Chao1 ~ Reef)
kruskal.test(data=ps_richness_meta, ACE ~ Reef)
kruskal.test(data=ps_richness_meta, Shannon ~ Reef)
kruskal.test(data=ps_richness_meta, Simpson ~ Reef)
kruskal.test(data=ps_richness_meta, InvSimpson ~ Reef)
kruskal.test(data=ps_richness_meta, Fisher ~ Reef)
```
```{r}
# Kruskal-Wallis testing for colony effect
kruskal.test(data=ps_richness_meta, Observed ~ Colony)
kruskal.test(data=ps_richness_meta, Chao1 ~ Colony)
kruskal.test(data=ps_richness_meta, ACE ~ Colony)
kruskal.test(data=ps_richness_meta, Shannon ~ Colony)
kruskal.test(data=ps_richness_meta, Simpson ~ Colony)
kruskal.test(data=ps_richness_meta, InvSimpson ~ Colony)
kruskal.test(data=ps_richness_meta, Fisher ~ Colony)
```

## Differential abundance testing
### ALDeX2
#### Baseline-Control
```{r}
# Treatment sample subsets of phyloseq objects
nsamples(ps_filtered_samples)
# samples
samples.BxC <- grep("Antibiotics", samples, invert = TRUE, value = TRUE)
samples.BxC <- grep("_", samples.BxC, invert = FALSE, value = TRUE)
ps.BxC <- prune_samples(samples.BxC, ps_filtered_samples)
ps.BxC
# summarize_phyloseq(ps.BxC)
# Read out ASV tables from phyloseq objects
aldex.asv <- as.data.frame(t(otu_table(ps.BxC)))
```
```{r}
conds <- as.character(sample_data(ps.BxC)$Treatment)
conds <- factor(conds, levels = c("Baseline", "Control"), ordered = TRUE)
```
```{r}
x.ps <- aldex(reads = aldex.asv, conditions = conds, mc.samples = 128, test = "t", effect = TRUE, include.sample.summary = FALSE, denom = "all", verbose = TRUE)
```
```{r}
aldex_res_base_control <- x.ps %>% 
  rownames_to_column("ASV") %>% 
  # filter(wi.eBH < 0.05) %>% 
  left_join(taxa_df, by = "ASV") %>% 
  arrange(desc(effect))
write_csv(aldex_res_base_control, "./outputs/phyloseq_results/aldex2_base_control.csv")
```
```{r}
#pdf("./outputs/aldex2/anti-aldex2.pdf", height = 4, width = 8)
par(mfrow=c(1,2))
aldex.plot(x.ps, type="MW", test="welch", xlab="Dispersion", ylab="Difference")
aldex.plot(x.ps, type = "MA", test = "welch", xlab = "Log-ratio abundance", ylab = "Difference")

#dev.off()
```
#### Baseline-Antibiotics
```{r}
# Treatment sample subsets of phyloseq objects
nsamples(ps_filtered_samples)
# samples
samples.BxA <- grep("Control", samples, invert = TRUE, value = TRUE)
samples.BxA<- grep("_", samples.BxA, invert = FALSE, value = TRUE)
ps.BxA <- prune_samples(samples.BxA, ps_filtered_samples)
ps.BxA
summarize_phyloseq(ps.BxA)
# Read out ASV tables from phyloseq objects
aldex.asv <- as.data.frame(t(otu_table(ps.BxA)))
```
```{r}
conds <- as.character(sample_data(ps.BxA)$Treatment)
conds <- factor(conds, levels = c("Baseline", "Antibiotics"), ordered = TRUE)
```
```{r}
x.ps <- aldex(reads = aldex.asv, conditions = conds, mc.samples = 128, test = "t", effect = TRUE, include.sample.summary = FALSE, denom = "all", verbose = TRUE)
```
```{r}
aldex_res_base_anti <- x.ps %>% 
  rownames_to_column("ASV") %>% 
  # filter(wi.eBH < 0.05) %>% 
  left_join(taxa_df, by = "ASV") %>% 
  arrange(desc(effect))
write_csv(aldex_res_base_anti, file = "./outputs/phyloseq_results/aldex2_base_anti.csv")
```
```{r}
#pdf("./outputs/aldex2/anti-aldex2.pdf", height = 4, width = 8)
par(mfrow=c(1,2))
aldex.plot(x.ps, type="MW", test="welch", xlab="Dispersion", ylab="Difference")
aldex.plot(x.ps, type = "MA", test = "welch", xlab = "Log-ratio abundance", ylab = "Difference")

#dev.off()
```
#### Antibiotics-Control
```{r}
# Treatment sample subsets of phyloseq objects
nsamples(ps_filtered_samples)
# samples
samples.AxC <- grep("Baseline", samples, invert = TRUE, value = TRUE)
samples.AxC <- grep("_", samples.AxC, invert = FALSE, value = TRUE)
ps.AxC <- prune_samples(samples.AxC, ps_filtered_samples)
ps.AxC
# summarize_phyloseq(ps.AxC)
# Read out ASV tables from phyloseq objects
aldex.asv <- as.data.frame(t(otu_table(ps.AxC)))
```
```{r}
conds <- as.character(sample_data(ps.AxC)$Treatment)
conds <- factor(conds, levels = c("Control", "Antibiotics"), ordered = TRUE)
```
```{r}
x.ps <- aldex(reads = aldex.asv, conditions = conds, mc.samples = 128, test = "t", effect = TRUE, include.sample.summary = FALSE, denom = "all", verbose = TRUE)
```
```{r}
aldex_res_control_anti <- x.ps %>% 
  rownames_to_column("ASV") %>% 
  # filter(wi.eBH < 0.05) %>% 
  left_join(taxa_df, by = "ASV")
write.csv(x.ps, file = "./outputs/phyloseq_results/aldex2_control_anti.csv")
```
```{r}
#pdf("./outputs/aldex2/anti-aldex2.pdf", height = 4, width = 8)
par(mfrow=c(1,2))
aldex.plot(x.ps, type="MW", test="welch", xlab="Dispersion", ylab="Difference")
aldex.plot(x.ps, type = "MA", test = "welch", xlab = "Log-ratio abundance", ylab = "Difference")
#dev.off()
```
```{r}
aldexmw <- aldex.plot_gg(x.ps, type = "MW", test = "welch")
aldexmw_anti <- aldexmw + 
  ggtitle("Antibiotics vs. Control Treatment")
aldexma <- aldex.plot_gg(x.ps, type = "MA", test = "welch")
#aldex2 + ggtitle(subtitle = "Antibiotics + Heat vs. Control Treatment", "Bacteria ASV ALDeX2 Differential Abundance MA Plot") 
 
aldexplots <- (aldexmw | aldexma) 
aldexplots
ggsave(aldexplots, filename = "./outputs/phyloseq_results/figures/anti-asv-aldex2.pdf", width = 8, height = 4, units = "in", device = "pdf")
ggsave(aldexma, filename = "./outputs/phyloseq_results/figures/anti-asv-aldexma.pdf", width = 4, height = 3, units = "in", device = "pdf")
# 90 differentially abundant ASVs!
```
```{r}
aldex_res_control_anti %>% 
  filter(effect < 0) %>% 
  dplyr::group_by(Family) %>%
  dplyr::count() %>% arrange(desc(n))
aldex_res_control_anti %>% 
  dplyr::group_by(Genus) %>% dplyr::count() %>% arrange(desc(n))
```
```{r}
# aldex_df <- aldex_res_base_control
# aldex_df <- aldex_res_base_anti
aldex_df <- aldex_res_control_anti

# 
aldexmw <- aldex_plot_gg(aldex_df, type = "MW", test = "welch") + ggtitle("Control vs. Antibiotics")
aldexma <- aldex_plot_gg(aldex_df, type = "MA", test = "welch")
 
aldexplots_control_anti <- (aldexmw | aldexma) 

aldexplots <- (aldexplots_base_control / aldexplots_base_anti / aldexplots_control_anti)
aldexplots

ggsave(aldexplots, filename = "./outputs/phyloseq_results/figures/aldex2_all.pdf", width = 9, height = 12, units = "in", device = "pdf")
```
#### Venn Diagrams
```{r}
base_control <- intersect(aldex_res_base_anti$ASV, aldex_res_base_control$ASV)
anti_shared <- intersect(aldex_res_base_anti$ASV, aldex_res_control_anti$ASV)

antibiotics_reduced_taxa <- taxa_df %>% filter(ASV %in% anti_shared)
data.frame("ASV" = intersect(base_control, anti_shared)) %>% left_join(taxa_df, by = "ASV") %>% View()
```
```{r}

```
## LEfSe
```{r}

```

## Core microbiome analysis
```{r core_microbiome_phyloseq}
# Determine members of the core microbiota with given abundance and prevalences
core_taxa_standard <- core_members(ps_filtered_rel, detection = 0, prevalence = 0.8) # 10
# 
core_taxa_fifty <- core_members(ps_filtered_rel, detection = 0, prevalence = 0.5) # 56
# 
# Filter the phyloseq object to include only core prevalent taxa
ps_filtered_core_standard <- core(ps_filtered_rel, detection = 0, prevalence = 0.8)
ps_filtered_core_standard
# 
ps_filtered_core_fifty <- core(ps_filtered_rel, detection = 0, prevalence = 0.5)
ps_filtered_core_fifty
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 56 taxa and 166 samples ]
# sample_data() Sample Data:       [ 166 samples by 11 sample variables ]
# tax_table()   Taxonomy Table:    [ 56 taxa by 7 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 56 tips and 55 internal nodes ]
```
### Treatment core microbiomes
```{r treatment_core}
# Treatment-specific core microbiomes
# --> how much of the core microbiome is eliminated by antibiotics treatment?
baseline_core <- core_members(ps_baseline_rel, detection = 0, prevalence = 0.75) # 48
control_core <- core_members(ps_control_rel, detection = 0, prevalence = 0.75) # 40
antibiotics_core <- core_members(ps_antibiotics_rel, detection = 0, prevalence = 0.75) # 14
#
#
combo_core <- union(union(baseline_core, control_core),  antibiotics_core) # 54 total
# prune taxa to just those in the combined treatment core
ps_combo_core_rel <- ps_filtered_rel %>% 
  prune_taxa(combo_core,.)
#
baseline_control_core <- intersect(baseline_core, control_core)
# prune taxa
ps_combo_core_rel <- ps_filtered_rel %>% 
  prune_taxa(baseline_control_core,.)
# prune to just correlated taxa
ps_combo_core_rel <- ps_filtered_rel %>% 
  prune_taxa(wgcna_correlated_core,.)
```
```{r treatment_core_venn}
intersect(baseline_core, control_core)
intersect(baseline_core, antibiotics_core)
intersect(control_core, antibiotics_core)
#
core_taxa_baseline_control <- tax_table(ps_filtered) %>% as.data.frame() %>% rownames_to_column("ID") %>% dplyr::filter(ID %in% intersect(baseline_core, control_core))
#
core_taxa_antibiotics_remove <- tax_table(ps_filtered) %>% as.data.frame() %>% rownames_to_column("ID") %>% dplyr::filter(ID %in% setdiff(intersect(baseline_core, control_core), antibiotics_core))
#
core_taxa_antibiotics_persist <- tax_table(ps_filtered) %>% as.data.frame() %>% rownames_to_column("ID") %>% dplyr::filter(ID %in% intersect(control_core, antibiotics_core))
#
core_taxa_antibiotics_unique <- tax_table(ps_filtered) %>% as.data.frame() %>% rownames_to_column("ID") %>% dplyr::filter(ID %in% setdiff(control_core, antibiotics_core))
```
```{r treatment_core_upset}
listInput <- list("Baseline" = baseline_core, 
                  "Control" = control_core, 
                  "Antibiotics" = antibiotics_core)
#
upset(fromList(listInput),  nsets = 3, 
      order.by = "freq", keep.order = F,
      empty.intersections = "on", nintersects = NA,
      sets.bar.color = treatcolors)
```
```{r treatment_core_euler}
# Input in the form of a named numeric vector
fit1 <- euler(c("Baseline" = 10, "Control" = 3, "Antibiotics" = 3, 
                "Baseline&Control" = 27, "Antibiotics&Baseline" = 1, "Antibiotics&Control" = 0, 
                "Antibiotics&Baseline&Control" = 10), shape = "ellipse")
#
treatment_core_euler <- plot(fit1,
     quantities = list(type = c("counts"), fontsize = 9),
     labels = identical(legend, FALSE),
     fills = list(fill = treatcolors, alpha = 0.5),
     edges = list(lwd = 0.5),
     legend = FALSE, 
     adjust_labels = TRUE)
treatment_core_euler
```
### Figure 3D
```{r figure 3D: manuscript}
pdf("./outputs/figures/Fig3D_treatment_core_upset.pdf", height = 75/25.4, width = 175/25.4)
treatment_upset <- upset(fromList(listInput),  nsets = 3, 
      order.by = "freq", keep.order = F,
      empty.intersections = "on", nintersects = NA,
      sets.bar.color = treatcolors)
print(treatment_upset)
dev.off()
```
```{r figure 3D: manuscript}
pdf("./outputs/figures/Fig3D_treatment_core_euler.pdf", height = 40/25.4, width = 40/25.4)
treatment_core_euler
dev.off()
```

```{r symbiont_type_core}

```


### Supplementary Tables 2 - 4
```{r}
core_taxa_baseline <- tax_table(ps_filtered) %>% as.data.frame() %>% rownames_to_column("ID") %>% dplyr::filter(ID %in% baseline_core)
write_excel_csv(core_taxa_baseline, "./outputs/phyloseq_results/TableS2_core_taxa_baseline.csv")
# 
core_taxa_control <- tax_table(ps_filtered) %>% as.data.frame() %>% rownames_to_column("ID") %>% dplyr::filter(ID %in% control_core)
write_excel_csv(core_taxa_control, "./outputs/phyloseq_results/TableS3_core_taxa_control.csv")
# 
core_taxa_antibiotics <- tax_table(ps_filtered) %>% as.data.frame() %>% rownames_to_column("ID") %>% dplyr::filter(ID %in% antibiotics_core)
write_excel_csv(core_taxa_antibiotics, "./outputs/phyloseq_results/TableS4_core_taxa_antibiotics.csv")
```
### Supplementary Tables 5 - 6
```{r}
#
write_excel_csv(core_taxa_antibiotics_remove, "./outputs/phyloseq_results/TableS5_core_taxa_antibiotics_reduced_table.csv")
#
write_excel_csv(core_taxa_antibiotics_persist, "./outputs/phyloseq_results/TableS6_core_taxa_antibiotics_persist_table.csv")
#
```

```{r core_microbiome_abundance}
core_abundance_standard <- sample_sums(ps_filtered_core_standard)
core_abundance_fifty <- sample_sums(ps_filtered_core_fifty)
#
core_abundance_combo <- sample_sums(ps_combo_core_rel)
```
### Plots of core microbiome relative abundance
```{r}

core_abundance_df <- data.frame(core_abundance_combo) %>% 
  rownames_to_column("Sample_ID") %>% 
  dplyr::rename("core_abundance" = core_abundance_combo) %>% 
  left_join(sample_metadata, by = "Sample_ID")
```
```{r}
core_abundance_df %>% 
  ggplot(aes(Genotype, core_abundance)) +
  facet_grid(Treatment ~ mtORF, space = "free", scales = "free_x") +
  geom_boxplot()
```

## Bubble plots
```{r}
ps_filtered_rel
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1835 taxa and 166 samples ]
# sample_data() Sample Data:       [ 166 samples by 11 sample variables ]
# tax_table()   Taxonomy Table:    [ 1835 taxa by 7 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 1835 tips and 1823 internal nodes ]
sample_data(ps_filtered_rel)
# 166 x 11 df
# View(tax_table(ps_filtered_rel)) 
# 1835 taxa by 7 taxonomic ranks
```
### Bubble plots of core microbiome ASV relative abundance
```{r sample_asv_data}
sample_asv_data <- data.frame(otu_table(ps_filtered_rel)) %>% 
  rownames_to_column("Sample_ID") # ps_filtered_core_fifty
# 
colnames(sample_asv_data) <- gsub("^X", "", colnames(sample_asv_data))
```
```{r bubble_data_ASV}
# create tidy data frame for bubble plots of ASV abundance
bubble_data_asv <- left_join(sample_metadata, sample_asv_data, by = "Sample_ID") %>% 
  pivot_longer(cols = -(Sample_ID:mtORF), names_to = "ASV", values_to = "relative_abundance") %>%
  full_join(taxa_df_filtered, by = "ASV") %>% 
  arrange(desc(relative_abundance))
# ps_filtered_rel: 304,610 obs. of 21 variables
```
```{r bubble_data_ASV_core}
bubble_core_standard <- bubble_data_asv %>%
  dplyr::filter(ASV %in% core_taxa_standard) #%>% 
  # dplyr::filter(Treatment != "Baseline")
# 
bubble_core_fifty <- bubble_data_asv %>%
  dplyr::filter(ASV %in% core_taxa_fifty) #%>% 
  # dplyr::filter(Treatment != "Baseline")
# 
bubble_core_combo <- bubble_data_asv %>%
  dplyr::filter(ASV %in% combo_core) #%>% 
  # dplyr::filter(Treatment != "Baseline")
#
bubble_core_baseline_control <- bubble_data_asv %>%
  dplyr::filter(ASV %in% baseline_control_core) #%>% 
  
unique(bubble_core_fifty$Class)
unique(bubble_core_fifty$Family)
class_order <- c(" Alphaproteobacteria", " Gammaproteobacteria", " Desulfovibrionia",
                                             " Planctomycetes", " Pla4_lineage",
                                             " Phycisphaerae",
                                             " Holophagae",
                                             " Bdellovibrionia", " Desulfobacteria", " Spirochaetia", " Campylobacteria")
bubble_core_fifty$Class <- factor(bubble_core_fifty$Class, 
                                  levels = class_order, 
                                  ordered = TRUE)
#
class_order <- c(" Alphaproteobacteria", " Gammaproteobacteria", " Desulfovibrionia",
                                             " Phycisphaerae", " Planctomycetes", " Bacteroidia", 
                                             " Bdellovibrionia", " Campylobacteria")
bubble_core_combo$Class <- factor(bubble_core_combo$Class, 
                                  levels = class_order, 
                                  ordered = TRUE)
```

```{r phylogenetic_tree_test}
# all ASVs
ps_filtered_rel %>% plot_tree(color = "Treatment", size = "Abundance", label.tips = "Family")
# core microbiome ASVs
ps_filtered_core_fifty %>% plot_tree(color = "Treatment", size = "Abundance", label.tips = "Family")
# combined core ASVs
ps_combo_core_rel %>%  plot_tree(color = "Treatment", size = "Abundance", label.tips = "Family")
```
```{r phylogenetic_tree_change_tip_labels}
# All ASVs
ps_tree <- phy_tree(ps_filtered_rel) # ps_filtered_rel
p <- ggtree(ps_tree)
# ps_tree$tip.label # tip labels are ASV IDs
# create data frame with ASV taxonomy information
taxa_df_rel <- data.frame(tax_table(ps_filtered_rel)) #Genus
# confirm that data frame rownames match tree tip labels, to prevent misassignment
all(rownames(taxa_df_rel) == ps_tree$tip.label) 
# change tip labels on phylogenetic tree
ps_tree$tip.label <- taxa_df_rel$Family
p2 <- ggtree(ps_tree) # create phylogenetic tree
# plot tree test
p2 + geom_tiplab()
```
```{r phylogenetic_tree_change_tip_labels}
# core microbiome ASVs
ps_tree <- phy_tree(ps_combo_core_rel) # ps_filtered_core_fifty
p <- ggtree(ps_tree)
# ps_tree$tip.label # tip labels are ASV IDs
# create data frame with ASV taxonomy information
# taxa_df_core_fifty <- data.frame(tax_table(ps_filtered_core_fifty)) #Genus
taxa_df_combo_core <- data.frame(tax_table(ps_combo_core_rel))
# confirm that data frame rownames match tree tip labels, to prevent misassignment
all(rownames(taxa_df_combo_core) == ps_tree$tip.label) 
# change tip labels on phylogenetic tree
# ps_tree$tip.label <- taxa_df_core_fifty$Class
ps_tree$tip.label <- taxa_df_combo_core$Family
p2 <- ggtree(ps_tree) # create phylogenetic tree
# plot tree test
p2 + geom_tiplab()
```
```{r wgcna_classes_dendrogram}
# This does not work right now...
tips <- c("b0a81d828606840739faecf67253a259", "3a6eb9c5c538c7745ce3677db40c9cba", "d4877dd74621c28f1b1e9271e1ff6e3a", "50b050a19ce41aadc068545ce53a8f64", "5e66fbb557569a2267bc202323b61461", "726a599eb8048c779b1e9d85249bfa6d", "bf763a858ed2aa5e61bf06d402ce1542", "4994d5ada2883f43f94aa07302aa6ffe", "ea9f302c8aaadcee80576b34cf4e8131")
ps_tree_class <- keep.tip(ps_tree, tips)

taxa_short_classes <- taxa_df_core_fifty %>% 
  rownames_to_column("ASV") %>% 
  dplyr::filter(ASV %in% tips) %>% 
  column_to_rownames("ASV")
# confirm that data frame rownames match tree tip labels, to prevent misassignment
all(rownames(taxa_short_classes) == ps_tree_class$tip.label) 

# change tip labels on phylogenetic tree
ps_tree_class$tip.label <- taxa_short_classes$Class

classes_dendrogram <- as.dendrogram.phylo(ps_tree_class)
ps_tree_class$tip.label
ggtree(ps_tree_class) + geom_tiplab()
```

```{r bubble_colors}
# 
colors_classes <- taxa_color_seq(ps_filtered_core_fifty, taxa_rank = "Class")
```
### Figure 3E
```{r figure 3E: manuscript}
# Bubble plot showing differential effects of antibiotics treatment on core bacteria taxa
pdf("./outputs/figures/Fig3D_bubbleplot_tree_all.pdf", height = 133/25.4, width = 175/25.4)
bubble_title <- expression(paste("Relative abundance of ASVs present in >50% of ", italic("Pocillopora"), " samples"))
# 
tree_bubbleplot <- p + 
  # geom_tiplab(size = 2) +
  geom_fruit(data = bubble_core_fifty, # bubble_core_fifty
             geom = geom_point,
             mapping = aes(x = Sample_ID, y = ASV, fill = Class, size = relative_abundance*100),
             offset = 0.05,
             pwidth = 4,
             # lwd = .05,
             alpha = 0.5,
             shape = 21,
             stroke = 0.1,
             color = "black",
             axis.params = list(),
             grid.params = list(vline = FALSE, color = "white"), #  list() if grid lines desired
             show.legend = TRUE) +
  facet_grid(. ~ Treatment, scales = "free_x", space = "free", drop = TRUE) +
  scale_size_continuous(range = c(1, 10), breaks = c(0.1, 50, 100), limits = c(0.1, 100),
                        labels = c("0.1%", "50%", "100%"),
                        guide = guide_legend(override.aes = list(fill = "lightgrey", color = "black", alpha = 1))) +
  # scale_fill_viridis(discrete = T,
  #                    guide = guide_legend(override.aes = list(size = 6, alpha = 0.8))) +
  scale_fill_brewer(palette = "Set3", 
                    guide = guide_legend(override.aes = list(size = 5, alpha = 0.8))) +
  # scale_fill_manual(values = colors_classes,
                    # guide = guide_legend(override.aes = list(size = 6, alpha = 0.8))) +
  labs(size = "Relative Abundance", fill = "Class") +
  theme(axis.text = element_blank(),
        plot.title = element_text(hjust = -1.2, size = 12),
        legend.position = "right",
        panel.background = element_rect(color = "white", fill = NA)) +
  ggtitle(bubble_title)

# modify facet strip colors
strip_colors <- c(treatcolors[1:3])
  gt <- ggplotGrob(tree_bubbleplot)
  strip_both <- which(grepl('strip-', gt$layout$name))
  k <- 1
  for (i in strip_both) {
    j <- which(grepl('rect', gt$grobs[[i]]$grobs[[1]]$childrenOrder))
    gt$grobs[[i]]$grobs[[1]]$children[[j]]$gp$fill <- strip_colors[k]
    k <- k+1
  }
  # # draw gtable object
  gp <- grid.draw(gt)
  print(gp)
dev.off()
```
```{r figure 3E: manuscript}
# Bubble plot showing differential effects of antibiotics treatment on core bacteria taxa - across treatments!
pdf("./outputs/figures/Fig3E_bubbleplot_tree_treatmentcore.pdf", height = 140/25.4, width = 175/25.4)
bubble_title <- expression(paste("Relative abundance of core ASVs detected in >75% of samples in any treatment"))
# 
tree_bubbleplot <- p + 
  # geom_tiplab(size = 2) +
  geom_fruit(data = bubble_core_combo, # bubble_core_fifty
             geom = geom_point,
             mapping = aes(x = Sample_ID, y = ASV, fill = Class, size = relative_abundance*100),
             offset = 0.05,
             pwidth = 4,
             # lwd = .05,
             alpha = 0.5,
             shape = 21,
             stroke = 0.2,
             color = "black",
             axis.params = list(),
             grid.params = list(vline = FALSE, color = "white"), #  list() if grid lines desired
             show.legend = TRUE) +
  facet_grid(. ~ Treatment, scales = "free_x", space = "free", drop = TRUE) +
  scale_size_continuous(range = c(1, 10), breaks = c(0.1, 25, 50, 100), limits = c(0.01, 100),
                        labels = c("0.1%", "25%", "50%", "100%"),
                        guide = guide_legend(override.aes = list(fill = "lightgrey", color = "black", alpha = 1))) +
  # scale_fill_viridis(discrete = T,
  #                    guide = guide_legend(override.aes = list(size = 6, alpha = 0.8))) +
  scale_fill_brewer(palette = "Set3", 
                    guide = guide_legend(override.aes = list(size = 5, alpha = 0.8))) +
  # scale_fill_manual(values = colors_classes,
                    # guide = guide_legend(override.aes = list(size = 6, alpha = 0.8))) +
  labs(size = "Relative Abundance", fill = "Class") +
  theme(axis.text = element_blank(),
        plot.title = element_text(hjust = 1.25, size = 12),
        legend.position = "left",
        legend.spacing.y = unit(1,"mm"),
        legend.title = element_text(size = 10),
        legend.justification = c(0,0),
        panel.background = element_rect(color = "white", fill = NA)) +
  ggtitle(bubble_title)

# modify facet strip colors
strip_colors <- c(treatcolors[1:3])
  gt <- ggplotGrob(tree_bubbleplot)
  strip_both <- which(grepl('strip-', gt$layout$name))
  k <- 1
  for (i in strip_both) {
    j <- which(grepl('rect', gt$grobs[[i]]$grobs[[1]]$childrenOrder))
    gt$grobs[[i]]$grobs[[1]]$children[[j]]$gp$fill <- strip_colors[k]
    k <- k+1
  }
  # # draw gtable object
  gp <- grid.draw(gt)
  print(gp)
dev.off()
```

```{r}
# Bubble plot showing differential effects of antibiotics treatment on all bacteria ASVs
pdf("./outputs/figures/FigS##_bubbleplot_tree_all_ASVs.pdf", height = 8.5, width = 11)
bubble_title <- expression(paste("Relative abundance of bacteria ASVs"))
# 
tree_bubbleplot <- p + 
  # geom_tiplab(size = 2) +
  geom_fruit(data = bubble_data_asv,
             geom = geom_point,
             mapping = aes(x = Sample_ID, y = ASV, fill = Class, size = relative_abundance*100),
             offset = 0.05,
             pwidth = 5,
             # lwd = .05,
             alpha = 0.5,
             shape = 21,
             stroke = 0.1,
             color = "black",
             axis.params = list(),
             grid.params = list(vline = FALSE, color = "white"), #  list() if grid lines desired
             show.legend = FALSE) +
  facet_grid(. ~ Treatment, scales = "free_x", space = "free", drop = TRUE) +
  scale_size_continuous(range = c(1, 10), breaks = c(0.1, 50, 100), limits = c(0.01, 100),
                        labels = c("0.1%", "50%", "100%"),
                        guide = guide_legend(override.aes = list(fill = "lightgrey", color = "black", alpha = 1))) +
  # scale_fill_viridis(discrete = T,
  #                    guide = guide_legend(override.aes = list(size = 6, alpha = 0.8))) +
  # scale_fill_brewer(palette = "Set3", 
  #                   guide = guide_legend(override.aes = list(size = 5, alpha = 0.8))) +
  # scale_fill_manual(values = colors_classes,
                    # guide = guide_legend(override.aes = list(size = 6, alpha = 0.8))) +
  labs(size = "Relative Abundance", fill = "Class") +
  theme(axis.text = element_blank(),
        legend.position = "right",
        panel.background = element_rect(color = "white", fill = NA)) +
  ggtitle(bubble_title)

# modify facet strip colors
strip_colors <- c(treatcolors[1:3])
  gt <- ggplotGrob(tree_bubbleplot)
  strip_both <- which(grepl('strip-', gt$layout$name))
  k <- 1
  for (i in strip_both) {
    j <- which(grepl('rect', gt$grobs[[i]]$grobs[[1]]$childrenOrder))
    gt$grobs[[i]]$grobs[[1]]$children[[j]]$gp$fill <- strip_colors[k]
    k <- k+1
  }
  # # draw gtable object
  gp <- grid.draw(gt)
  print(gp)

dev.off()
```

```{r}
# ?geom_fruit
tree_bubbleplot
```

```{r bubble_all}
# One thing I want to do is have the phylogenetic tree next to the plot
bubble_data %>% ggplot(aes(x = Sample_ID, y = ASV, fill = Family)) +
  geom_point(aes(size = relative_abundance*100), alpha = 0.9, shape = 21) +
  # scale_size_continuous(limits = c(0, 2), range = c(0,10), breaks = c(0.1, 0.25, 0.50, 0.75)) +
  facet_nested(. ~ Treatment + mtORF + haplotype, scales="free", space="free") +
  # scale_fill_manual(values = x) +
  # scale_y_discrete(limits = (levels(bubble_0.05$ASV))) +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  labs(size = "Relative Abundance", color = "Genera") +
  guides(fill = guide_legend(override.aes = list(size = 4)))
```

```{r}

```

### Bubble plots of Family relative abundance
```{r}
# This may be difficult to get the correct phylogenetic tree printed next to the plot
physeq <- ps_filtered_rel
taxa_rank = "Family"
glom <- tax_glom(physeq, taxrank = taxa_rank)
  psmelted <- psmelt(glom)
  psmelted[, taxa_rank] <- as.character(psmelted[ , taxa_rank])
  # psmelted[psmelted$Abundance < 0.01, taxa_rank] <- "<1% abundance" 
  psmelted[psmelted$Abundance < 0.05, taxa_rank] <- "<5% abundance"
  # psmelted[, taxa_rank] <- factor(psmelted[, taxa_rank], levels = rev(taxa_levels), ordered = TRUE)
  psmelted$Treatment <- factor(psmelted$Treatment, levels = c("Baseline", "Control", "Antibiotics", ordered = TRUE))
```
```{r}
ps_filtered_rel
```


## Output bacteria taxa abundance tables for WGCNA module eigengene correlations
```{r}
sample_asv_data <- data.frame(otu_table(ps_filtered_rel)) %>% 
  rownames_to_column("Sample_ID") # ps_filtered_core_fifty
# 
colnames(sample_asv_data) <- gsub("^X", "", colnames(sample_asv_data))
```
### Class cumulative relative abundance of overall core ASVs
```{r bacteria_class_WGCNA, message=FALSE}
# need to filter to only samples analyzed for transcriptomes
class_abundances <- ps_filtered_rel %>%
  subset_samples(., sample_names(ps_filtered_rel) %in% samples_analysis$Sample_ID[match(rownames(coldata), samples_analysis$Sample_ID)]) %>% 
  tax_glom(., taxrank = "Class") %>% 
  psmelt(.) %>%
  # filter(Abundance > 0.001) %>%
  dplyr::filter(OTU %in% core_taxa_fifty) 
# 
class_average_abundances <- class_abundances %>%
  # filter(Family %in% int_families) %>%
  group_by(Sample, Class) %>%
  dplyr::summarise(cum_rel_abundance = sum(Abundance, na.rm = TRUE)) %>%
  ungroup()
# This step is complicated because not all families are found across all samples
# family_average_abundances %>% ungroup() %>% count(Family) %>% View()
# 
datTaxa_Class <- class_average_abundances %>% 
  tidyr::pivot_wider(names_from = Class, values_from = cum_rel_abundance) %>% 
  arrange(Sample) %>%
  # relocate(Endozoicomonadaceae, Rhodobacteraceae, Alteromonadaceae) %>%
  column_to_rownames("Sample") 
#
write_csv(datTaxa_Class, file = "./outputs/phyloseq_results/datTaxa_Class.csv")
```
### Relative abunance of individual baseline and control core ASVs 
```{r}
# need to filter to only samples analyzed for transcriptomes
combo_core_abundances <- ps_filtered_rel %>% # use original relative abundances for these specific core ASVs
  subset_samples(., sample_names(ps_filtered_rel) %in% samples_analysis$Sample_ID[match(rownames(coldata), samples_analysis$Sample_ID)]) %>% 
  # tax_glom(., taxrank = "Class") %>% # keep as individual ASVs for high-resolution hypotheses
  psmelt(.) %>%
  # filter(Abundance > 0.001) %>%
# include only 37 ASVs detected in baseline and control core microbiomes: 27 are depleted by antibiotics, 10 persist
  dplyr::filter(OTU %in% baseline_control_core) 
#
datTaxa_CoreASVs <- combo_core_abundances %>% 
  dplyr::select(OTU, Sample, Abundance) %>% 
  tidyr::pivot_wider(names_from = OTU, values_from = Abundance) %>% 
  dplyr::arrange(Sample) %>% 
  # dplyr::select(Sample, ps_tree$tip.label) %>% 
  column_to_rownames("Sample") 
#
# create extended text labels to display taxonomy information, and not just ASV IDs
CoreASVs_labels <- combo_core_abundances %>% 
    mutate(Label = str_c(OTU, Order, Family, sep = ":")) %>% 
    dplyr::select(OTU, Label) %>% distinct(OTU, .keep_all = T) %>% View()
    tidyr::pivot_wider(names_from = OTU, values_from = Label) %>% 
    dplyr::select(ps_tree$tip.label) %>%
    tidyr::pivot_longer(1:54, names_to = "OTU", values_to = "Label")
#
# add column names to display ASV taxonomic information on correlation heatmap
colnames(datTaxa_CoreASVs) <- CoreASVs_labels$Label
#
# write_csv(datTaxa_CoreASVs, file = "./outputs/phyloseq_results/datTaxa_CoreASVs.csv")
```
```{r wgcna_core_asv_dendrogram}
# ps_tree$tip.label
# confirm that data frame rownames match tree tip labels, to prevent misassignment
all(rownames(taxa_df_combo_core) == ps_tree$tip.label) 
# colnames(datTaxa_CoreASVs)
# ps_tree$tip.label
all(colnames(datTaxa_CoreASVs) == ps_tree$tip.label)
# 
# 
clustASVs <- as.dendrogram.phylo(ps_tree)
#
clustASVs_save <- as.dendrogram.phylo(ps_tree)
```
### Relative abunance of specific core ASVs that are correlated to one or more WGCNA modules
```{r}
# need to filter to only samples analyzed for transcriptomes
combo_core_abundances <- ps_filtered_rel %>% # use original relative abundances for these specific core ASVs
  subset_samples(., sample_names(ps_filtered_rel) %in% samples_analysis$Sample_ID[match(rownames(coldata), samples_analysis$Sample_ID)]) %>% 
  # tax_glom(., taxrank = "Class") %>% # keep as individual ASVs for high-resolution hypotheses
  psmelt(.) %>%
  # filter(Abundance > 0.001) %>%
# include only 37 ASVs detected in baseline and control core microbiomes: 27 are depleted by antibiotics, 10 persist
  dplyr::filter(OTU %in% wgcna_correlated_core) 
#
datTaxa_CoreASVs <- combo_core_abundances %>% 
  dplyr::select(OTU, Sample, Abundance) %>% 
  tidyr::pivot_wider(names_from = OTU, values_from = Abundance) %>% 
  dplyr::arrange(Sample) %>% 
  # dplyr::select(Sample, ps_tree$tip.label) %>% 
  column_to_rownames("Sample") 
#
# create extended text labels to display taxonomy information, and not just ASV IDs
CoreASVs_labels <- combo_core_abundances %>% 
    mutate(Label = str_c(OTU, Order, Family, sep = ":")) %>% 
    dplyr::select(OTU, Label) %>% distinct(OTU, .keep_all = T) %>% 
    tidyr::pivot_wider(names_from = OTU, values_from = Label) %>% 
    # dplyr::select(ps_tree$tip.label) %>%
    tidyr::pivot_longer(1:20, names_to = "OTU", values_to = "Label")
#
# add column names to display ASV taxonomic information on correlation heatmap
colnames(datTaxa_CoreASVs) <- CoreASVs_labels$Label
```


```{r}
combo_core_abundances %>% ggplot(aes(Abundance)) + geom_histogram(binwidth = 0.001)
```


```{r}
# 
family_average_abundances %>% 
  ggplot(aes(x = Sample, y = avg_rel_abundance, fill = Family)) +
  geom_col(position = "fill")
# 
```

```{r int_genus_relative_abundances}
genus_abundances <- ps_filtered_rel %>%
  tax_glom(., taxrank = "Genus") %>% 
  psmelt(.) %>% 
  # filter(Family %in% int_families) %>%
  # dplyr::filter(Family == "Endozoicomonadaceae")
  group_by(SampleID, Genus) %>%
  dplyr::summarise(avg_rel_abundance = mean(Abundance, na.rm = TRUE))
# 
genus_abundances %>% 
  ggplot(aes(x = Genus, y = avg_rel_abundance)) +
  geom_boxplot()
```

### Supplementary Table 
```{r}
tax_table(ps_filtered) %>% as.data.frame() %>% rownames_to_column("ID") %>% dplyr::filter(ID %in% colnames(modP_bacteria)) %>% write_excel_csv(., "./outputs/phyloseq_results/TableSX_correlated_core_taxa_table.csv")
```


## Summary
```{r}
sessionInfo()
```

```{r}
write_csv(families_treatment_contrast, "./outputs/families_treatment_contrast.csv")
families_treatment_contrast$Treatment <- factor(families_treatment_contrast$Treatment, levels = treatment_levels, ordered = TRUE)
plot <- families_treatment_contrast %>% ggplot(aes(Family, Cumulative_Rel_Abundance, color = Treatment)) + geom_point() + scale_color_manual(values = treatcolors) + coord_flip()
print(plot)
```

