---
title: "quantseq_transcriptome_SNPFILTR"
author: "Mike Connelly"
date: "8/25/2022"
output: html_document
---

```{r setup, include=FALSE}
# O'Leary et al. 
# modified from https://github.com/sjoleary/SNPFILT/blob/master/filteringsnapper.Rmd
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())

```
```{r}
library("tidyverse")
library("vcfR")
library("SNPfiltR")
# 
source("./R/VCFfilterstats_edit.R")
source("./R/VCFfilterviz.R")
# need to build custom functions in the repository (ex. multiplot())
```

## Preliminary filtered dataset stats
The raw VCF was filtered on Hydra for loci with <50% missing data, Q > 30, depth 20 - 300, and minor allele frequency > 0.015, 
The preliminary all-sample VCF contains 644 SNPs across 101 samples
The preliminary genotype-specific VCF contains #### SNPs across ~14 samples

Perform filtering target several thousand (~1,000 - 10,000) biallelic SNPs for downstream analyses
## Import VCF file
```{r imporf vcf}
# all samples
vcf <- vcfR::read.vcfR(file = "./outputs/phylotrans_pdam/samples_all_filtered_primary.vcf")
# top samples from each genotype, direct from GenotypeGVCFs
vcf <- vcfR::read.vcfR(file = "./outputs/phylotrans_Pdam/samples_genotypes.vcf.gz")
```
```{r}
### check the metadata present in your vcf
vcf
# All samples VCF
# ***** Object of Class vcfR *****
# 101 samples
# 351 CHROMs
# 644 variants
# Object size: 2.9 Mb
# 0 percent missing data
# *****        *****         *****
# 
#generate popmap file. Two column popmap with the same format as stacks, and the columns must be named 'id' and 'pop'
popmap <- data.frame(id = colnames(vcf@gt)[2:length(colnames(vcf@gt))],
                     pop = substr(colnames(vcf@gt)[2:length(colnames(vcf@gt))], 1,6))
```
## Inspect with vcfR
```{r}
# pdam_genome <- ape::read.dna("/Users/mikeconnelly/computing/sequences/genomes/coral/pocillopora/pdam/pdam_genome.fasta", format = "fasta")
# pdam_gff <- read.table("/Users/mikeconnelly/computing/sequences/genomes/coral/pocillopora/pdam/pdam_genome.gff", sep="\t", quote="")
# 
# pdam_meta_chrom <- create.chromR(name="Supercontig", vcf = vcf, seq = pdam_genome, ann = pdam_gff, verbose = FALSE)
# chrom <- proc.chromR(pdam_meta_chrom, verbose = TRUE)
```

## Filter VCF file
```{r}
vcf %>% SNPfiltR::hard_filter(.)
```
```{r}
# all: 20,30
#hard filter to minimum depth of 10, and minimum genotype quality of 30

vcf_001 <- hard_filter(vcfR = vcf, depth = 10, gq = 20)
# all-sample VCF
#
# Genotype-specific VCF
# 96.73% of genotypes fall below a read depth of 20 and were converted to NA
# 5.98% of genotypes fall below a genotype quality of 30 and were converted to NA
# 
vcf_001
```
```{r}
 vcf_002 <- vcf_001 %>%  
  filter_allele_balance() #%>% 
# Genotype-specific VCF
# 18.36% of het genotypes (3.9% of all genotypes) fall outside of 0.25 - 0.75 allele balance ratio and were converted to NA
vcf_002
```
```{r}
max_depth(vcf_002)
# Genotype-specific VCF
# dashed line indicates a mean depth across all SNPs of 38.9
```

```{r}
#run function to visualize samples
missing_by_sample(vcfR = vcf_002, popmap = popmap)
```
```{r}
vcf_003 <- vcf_002 %>% missing_by_sample(., cutoff = 1) 
# all-samples vcf
# 9 samples are above a 0.95 missing data cutoff, and were removed from VCF
# 21 samples are above a 0.8 missing data cutoff, and were removed from VCF
# 40 samples are above a 0.5 missing data cutoff, and were removed from VCF

#subset popmap to only include retained individuals
popmap<-popmap[popmap$id %in% colnames(vcf_003@gt),]
#if there are still problematic samples, drop them using the following syntax
#vcfR <- vcfR[,colnames(vcfR@gt) != "A_woodhouseii_24711" & colnames(vcfR@gt) != "A_californica_45901"]

#remove invariant sites generated by dropping individuals
vcf_004 <- min_mac(vcf_003, min.mac = 1)
# Genotype-specific VCF
# 6.83% of SNPs fell below a minor allele count of 1 and were removed from the VCF
vcf_004
```
```{r}
# missing_by_snp(vcf_004) # 0 SNPs passing 100% completeness threshold, visualizations cannot be generated, returning filtered vcf
#
vcf_005 <- vcf_004 %>% missing_by_snp(., cutoff = .75)
# 97.5% of SNPs fell below a completeness cutoff of 0.95 and were removed from the VCF
# 94.67% of SNPs fell below a completeness cutoff of 0.9 and were removed from the VCF
# 92.33% of SNPs fell below a completeness cutoff of 0.85 and were removed from the VCF
# 82.83% of SNPs fell below a completeness cutoff of 0.75 and were removed from the VCF
# 12.5% of SNPs fell below a completeness cutoff of 0.5 and were removed from the VCF
vcf_005
# ***** Object of Class vcfR *****
# 61 samples
# 243 CHROMs
# 382 variants
# Object size: 1.4 Mb
# 11.24 percent missing data
# *****        *****         *****
#subset popmap to only include retained individuals
popmap<-popmap[popmap$id %in% colnames(vcf_005@gt),]
```

## Step 5:
investigate the effect of a minor allele count (MAC) cutoff on downstream inferences.
```{r}
#investigate clustering patterns with and without a minor allele cutoff
#use min.mac() to investigate the effect of multiple cutoffs
vcf_mac <- vcf_005 %>% min_mac(., min.mac = 2)
# 36.14% of SNPs fell below a minor allele count of 2 and were removed from the VCF
```
```{r}
#assess clustering without MAC cutoff
miss <- assess_missing_data_tsne(vcf_005, popmap, clustering = FALSE)

#assess clustering wit MAC cutoff
miss <- assess_missing_data_tsne(vcf_mac, popmap, clustering = FALSE)
```

```{r}
#plot depth per snp and per sample
dp <- extract.gt(vcf_005, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)

#plot genotype quality per snp and per sample
gq <- extract.gt(vcf_005, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)

```

## Step 6: 
  Write out files for downstream analysis
```{r}
# all SNPs unlinked
vcf_006 <- distance_thin(vcf_005, min.distance = 1000)
# 382 out of 382 input SNPs were not located within 1000 base-pairs of another SNP and were retained despite filtering
```
```{r}
vcf <- vcf_005
vcf <- vcf_006
vcf <- vcf[is.biallelic(vcf), ] 
```
```{r}
#write out vcf with all SNPs
vcfR::write.vcf(vcf, "./outputs/vcf_poc_anti_genotypes_654.vcf.gz")
```
  
  