---
title: "quantseq transcriptome QC symbionts"
author: "Mike Connelly"
date: "04/07/2022"
output: 
  html_document:
      code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
options(stringsAsFactors = FALSE)
```
```{r packages}
library("tidyverse")
library("ggrepel")
library("scales")
library("patchwork")
library("ggthemes")
#
source("./R/ismej_theme.R")
```
```{r colors}
### Set overall theme, colors, and shapes for ggplot2
# colors for experimental treatments
treatcolors <- c("lightblue", "dodgerblue", "firebrick1")
shapes <- c(15:20) # may be useful for genotype variables: Gulf, Location, mtORF, symbiont, etc.
# colors for genotypes
genotype_colors <- read_csv("data/genotype_colors.csv") #%>%
  # dplyr::filter(Genotype %in% c("PAN-78", "PAN-79", "PAN-83"))# color names csv corresponding to excel sheet colors
genocolors <- c(genotype_colors$color) 
# default ggplot fill
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}
genocolors <- gg_color_hue(14)
#
# null colors for presentation figures
condcolors_null <- c(rep("black", 3))
colshapes_null <- c(rep(20, 8))
```
```{r theme}
theme_set(theme_ismej())
```
```{r genotypes_fragments}
genotypes <- read_csv("data/genotypes.csv")
#
fragments <- read_csv("data/fragments.csv", col_types = cols(
  Fragment_ID = col_character()))
```
```{r factors}
# factor levels
# treatment_levels <- c("Baseline", "Control", "Antibiotics", "FSW_Controls", "SW_Controls", "Neg_Controls", "ZYMO_Controls")
sampling_levels <- c("Day 0", "Day 7")
genotype_levels <- c("PAN-05", "PAN-10", "PAN-32", "PAN-34", "PAN-35", "PAN-37", "PAN-39", "PAN-41", "PAN-43","PAN-44", "PAN-78", "PAN-79", "PAN-83", "URA-51", NA)
gulf_levels <- c("Chiriqui", "Panama")
loc_levels <- c("Uva", "Saboga", "Uraba")
mtorf_levels <- c("Type_1", "Type_3")#, "unknown")
sym_levels <- c("C1bc", "C1d", "D1")#, "unknown")
genus_levels <- c("Cladocopium", "Durusdinium")
```
```{r  sample_data, echo=TRUE}
# Import sample metadata 
sample_metadata <- read_delim("./data/quantseq_metadata.tsv", delim ="\t", col_names = c("Sample_ID", "Genotype", "Treatment", "Sampling_Time", "Sequencing_Success", "Sequencing_Round", "Include_Analysis"), comment = "#")#, row.names = 1, header = FALSE)
# join with genotype metadata
sample_metadata <- sample_metadata %>% left_join(genotypes, by = "Genotype")
# Set factor orders
# sample_metadata$Genotype <- factor(sample_metadata$Genotype, levels = genotype_levels, ordered = TRUE)
# sample_metadata$Treatment <- factor(sample_metadata$Treatment, levels = treatment_levels, ordered = TRUE)
sample_metadata$Gulf <- factor(sample_metadata$Gulf, levels = gulf_levels, ordered = TRUE)
sample_metadata$Location <- factor(sample_metadata$Location, levels = loc_levels, ordered = TRUE)
sample_metadata$mtORF <- factor(sample_metadata$mtORF, levels = mtorf_levels, ordered = TRUE)
sample_metadata$ITS2 <- factor(sample_metadata$ITS2, levels = sym_levels, ordered = TRUE)
sample_metadata$Symbiont_Genus <- ifelse(sample_metadata$ITS2 == "D1", "Durusdinium", "Cladocopium")
# Create analysis-compatible sample map
samples <- sample_metadata %>%
  dplyr::filter(Sequencing_Success == T)
```
```{r analysis_samples}
samples_analysis <- samples %>% 
  dplyr::filter(Include_Analysis == T, Treatment != "Baseline")
# 
samples_analysis$Sample_ID
```

```{r QC data}
# samples <- 
# samples$SampleID <- gsub('-','.', samples$SampleID)
QC_summary <- read_csv("./data/MultiQCs_PocAnti_011322.csv")
sum(QC_summary$`Raw Reads`)
# 1.298 billion reads as of 01/13/2022
```
```{r QC_analysis_samples_summary}
# summary statistics for analysis samples
QC_summary_analysis <- QC_summary %>% 
  dplyr::filter(Sample_ID %in% samples_analysis$Sample_ID)
# 
sum(QC_summary_analysis$`Raw Reads`)
mean(QC_summary_analysis$`Raw Reads`)
sd(QC_summary_analysis$`Raw Reads`)
sd(QC_summary_analysis$`Raw Reads`)/((sqrt(53)))
# 
sum(QC_summary_analysis$`Uniquely Mapped Reads`)/sum(QC_summary_analysis$`Raw Reads`)
```

```{r tidydata}
QC_summary <- QC_summary %>%
  dplyr::select(-`multimapped`, -`unmapped`) %>% 
  pivot_longer(., `Raw Reads`:`Assigned Reads`, names_to = "pipeline_step", values_to = "read_count") %>%
  left_join(., genotypes, by = "Genotype")
# QC_summary$Treatment <- factor(QC_summary$Treatment, levels = c("control", "Heat", "Antibiotics", "Antibiotics.Heat"), ordered = TRUE)
#
QC_summary$pipeline_step <- factor(QC_summary$pipeline_step, levels = c("Raw Reads", "Trimmed Reads", "Uniquely Mapped Reads", "Input Reads", "Assigned Reads"), ordered = TRUE)
```

## Vizualize raw read depth distribution
```{r read_depth_all}
# All sequenced samples
FASTQCg <- QC_summary %>% 
  filter(pipeline_step == "Raw Reads") %>%
  ggplot(aes(Sample_ID, read_count, color = Genotype))
FASTQCg + geom_point() + 
  scale_color_manual(values = genocolors) +
  # scale_shape_manual(values = colshapes) +
  geom_hline(yintercept = 10000000, color  = "green") +
  geom_hline(yintercept = 8000000) + 
  geom_hline(yintercept = 1000000, color = "red")
  #geom_text_repel(data = FASTQC_summary[1:15, ], aes(label = GNomExID), color = "black", size = 3, box.padding = unit(0.5, "lines")
```
```{r read_depth_analysis}
# Only analysis samples
FASTQCg <- QC_summary %>% 
  filter(Sample_ID %in% samples_analysis$Sample_ID) %>% 
  filter(pipeline_step == "Raw Reads") %>%
  ggplot(aes(Sample_ID, read_count, color = Genotype))
FASTQCg + geom_point() + 
  scale_color_manual(values = genocolors) +
  # scale_shape_manual(values = colshapes) +
  geom_hline(yintercept = 10000000, color  = "green") +
  geom_hline(yintercept = 8000000) + 
  geom_hline(yintercept = 1000000, color = "red")
  #geom_text_repel(data = FASTQC_summary[1:15, ], aes(label = GNomExID), color = "black", size = 3, box.padding = unit(0.5, "lines")
```

## Vizualize overall pipeline performance metrics: raw read counts and proportions
```{r read_counts_all}
# All sequenced samples
# pdf("./outputs/figures/transcriptome_QC_counts.pdf")
pipeQC_counts <- QC_summary %>%
  # dplyr::filter(pipeline_step != "Input Reads" & pipeline_step != "Assigned Reads") %>%
  ggplot(aes(pipeline_step, read_count, color = Genotype)) +
  geom_path(aes(group = Sample_ID), show.legend = FALSE) +
  geom_point(aes(fill = Genotype), size = 2, shape = 21, color = "black", alpha = 0.5, stroke = 0.5, show.legend = FALSE) + 
  facet_grid(.~Treatment, scales = "free") +
  # scale_y_continuous(limits = c(0, 2.5e7), breaks = c(0, 5e6, 1e7, 1.5e7, 2e7, 2.5e7)) +
  scale_color_manual(values = genocolors) +
  # scale_shape_manual(values = colshapes) + 
  scale_fill_manual(values = genocolors) +
  theme(axis.text.x = element_text(angle = 315, hjust = 0),
        axis.title.x = element_blank()
        # panel.grid.major = element_line(color = "grey92"),
        ) +
  ylab("read count") +
  xlab("pipeline step") + 
  ggtitle("read counts")
pipeQC_counts
# dev.off()
```
```{r read_counts_analysis}
# Only analysis samples
# pdf("./outputs/figures/transcriptome_QC_counts.pdf")
pipeQC_counts <- QC_summary %>%
  filter(Sample_ID %in% samples_analysis$Sample_ID) %>% 
  # dplyr::filter(pipeline_step != "Input Reads" & pipeline_step != "Assigned Reads") %>%
  ggplot(aes(pipeline_step, read_count, color = Genotype)) +
  geom_path(aes(group = Sample_ID), show.legend = FALSE) +
  geom_point(aes(fill = Genotype), size = 2, shape = 21, color = "black", alpha = 0.5, stroke = 0.5, show.legend = FALSE) + 
  facet_grid(.~Treatment, scales = "free") +
  # scale_y_continuous(limits = c(0, 2.5e7), breaks = c(0, 5e6, 1e7, 1.5e7, 2e7, 2.5e7)) +
  scale_color_manual(values = genocolors) +
  # scale_shape_manual(values = colshapes) + 
  scale_fill_manual(values = genocolors) +
  theme(axis.text.x = element_text(angle = 315, hjust = 0),
        axis.title.x = element_blank()
        # panel.grid.major = element_line(color = "grey92"),
        ) +
  ylab("read count") +
  xlab("pipeline step") + 
  ggtitle("read counts")
pipeQC_counts
# dev.off()
```

```{r proportions}
QC_summary <- QC_summary %>% 
  dplyr::group_by(Sample_ID) %>%
  mutate(proportion = read_count / read_count[pipeline_step == "Raw Reads"])
```
```{r read_proportions_all}
# All sequenced samples
# pdf("./outputs/figures/transcriptome_QC_proportions.pdf")
pipeQC_props <- QC_summary %>%
  ggplot(aes(pipeline_step, proportion, color = Genotype, fill = Genotype)) +
  geom_path(aes(group = Sample_ID), show.legend = TRUE) +
  geom_point(size = 2, shape = 21, color = "black", alpha = 0.5, stroke = 0.5, show.legend = TRUE) + 
  facet_grid(.~Treatment, scales = "free") +
  scale_y_continuous(limits = c(0, 1)) +
  scale_color_manual(values = genocolors) +
  # scale_shape_manual(values = colshapes) + 
  scale_fill_manual(values = genocolors) +
  theme(axis.text.x = element_text(angle = 315, hjust = 0), 
        # panel.grid.major = element_line(color = "grey92")
        legend.position = "right",
        legend.direction = "vertical") +
  # guides(color = guide_legend(override.aes = list(color = condcolors_AxH, alpha = 1, stroke = 1)),
  #          fill = guide_legend(override.aes = list(fill = condcolors_AxH, shape = 21, alpha = 1, stroke = 0.5)),
  #          shape = guide_legend(override.aes = list(shape = colshapes, alpha = 1, stroke = 0.5))) +
  ylab("proportion raw reads") +
  xlab("pipeline step") + 
  ggtitle("read proportions")
pipeQC_props
# dev.off()
```
```{r read_proportions_analysis}
# Only analysis samples
# pdf("./outputs/figures/transcriptome_QC_proportions.pdf")
pipeQC_props <- QC_summary %>%
  filter(Sample_ID %in% samples_analysis$Sample_ID) %>% 
  ggplot(aes(pipeline_step, proportion, color = Genotype, fill = Genotype)) +
  geom_path(aes(group = Sample_ID), show.legend = TRUE) +
  geom_point(size = 2, shape = 21, color = "black", alpha = 0.5, stroke = 0.5, show.legend = TRUE) + 
  facet_grid(.~Treatment, scales = "free") +
  scale_y_continuous(limits = c(0, 1)) +
  scale_color_manual(values = genocolors) +
  # scale_shape_manual(values = colshapes) + 
  scale_fill_manual(values = genocolors) +
  theme(axis.text.x = element_text(angle = 315, hjust = 0), 
        # panel.grid.major = element_line(color = "grey92")
        legend.position = "right",
        legend.direction = "vertical") +
  # guides(color = guide_legend(override.aes = list(color = condcolors_AxH, alpha = 1, stroke = 1)),
  #          fill = guide_legend(override.aes = list(fill = condcolors_AxH, shape = 21, alpha = 1, stroke = 0.5)),
  #          shape = guide_legend(override.aes = list(shape = colshapes, alpha = 1, stroke = 0.5))) +
  ylab("proportion raw reads") +
  xlab("pipeline step") + 
  ggtitle("read proportions")
pipeQC_props
# dev.off()
```

### Supplementary Figure 8
```{r}
pdf("./outputs/figures/FigS8_readcounts.pdf", height = 6, width = 6.5)
qcfig <- (pipeQC_counts / pipeQC_props & theme(legend.position = "right", legend.margin = margin(2,2,2,10, unit = "mm"))) + plot_annotation(tag_levels = "A") + plot_layout(guides = "collect") 
# ggsave(qcfig, filename = "./outputs/figures/FigS8_readcounts.pdf", width = 6.5, height = 5, units = "in", device = "pdf")
print(qcfig)
dev.off()
```

## Vizualize STAR alignment rates across genotypes, mtORF types, haplotypes, etc. 
```{r alignment_boxplot_viz}
QC_summary %>%
  filter(pipeline_step == "Uniquely Mapped Reads") %>% 
  ggplot(aes(Sequencing_Round, proportion)) + 
  geom_boxplot() +
  facet_grid(.~ mtORF + haplotype, scales = "free")

```
### Test for significant differences across genotypes, mtORF types, haplotypes, etc. 
```{r alignment_rate_anova}
STAR_QC <- QC_summary %>%
  filter(pipeline_step == "Uniquely Mapped Reads")

aov.mtORF <- aov(proportion ~ mtORF, data = STAR_QC)
summary(aov.mtORF)
aov.haplotype <- aov(proportion ~ haplotype, data = STAR_QC)
summary(aov.haplotype)
aov.Sequencing_Round <- aov(proportion ~ Sequencing_Round, data = STAR_QC)
summary(aov.Sequencing_Round)
```

