---
title: "quantseq_transcriptome_analysis"
author: "Mike Connelly"
date: "1/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
options(stringsAsFactors = FALSE)
```

## Setup packages and working directories
```{r packages, message=FALSE, include=FALSE}
# Include all packages needed for the entire transcriptome analysis and downstream visualizations
library("tidyverse")
# Essential RNAseq analysis packages
library("doParallel")
library("variancePartition")
library("limma")
library("pairwiseAdonis")
library("DESeq2")
library("PCAtools")
library("apeglm")
library("WGCNA")
library("dynamicTreeCut")
library("flashClust")
library("genefilter")
# library("adegenet")
# GO/KOG enrichment packages and other functions
library("KOGMWU")
source("./R/GO_MWU/gomwu.functions.R")
# Data visualization packages
# library("pheatmap")
library("ComplexHeatmap")
# library("VennDiagram")
# library("eulerr")
library("UpSetR")
# Graphics packages
# library("cowplot")
library("patchwork")
library("gridExtra")
library("ggthemes")
library("ggpubr")
library("ggrepel")
library("ggnewscale")
library("RColorBrewer")
library("viridis")
library("circlize")
library("stringr")
library("extrafont")
library("extrafontdb")
# extended visualization functions
source("./R/anti_phys_transcriptome_functions.R")
source("./R/ismej_theme.R")
```
```{r colors}
### Set overall theme, colors, and shapes for ggplot2
# colors for experimental treatments
treatcolors <- c("lightblue", "dodgerblue", "firebrick1")
shapes <- c(15:20) # may be useful for genotype variables: Gulf, Location, mtORF, symbiont, etc.
# colors for genotypes
genotype_colors <- read_csv("data/genotype_colors.csv") #%>%
  # dplyr::filter(Genotype %in% c("PAN-78", "PAN-79", "PAN-83"))# color names csv corresponding to excel sheet colors
genocolors <- c(genotype_colors$color) 
# default ggplot fill
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}
genocolors <- gg_color_hue(14)
#
# null colors for presentation figures
condcolors_null <- c(rep("black", 3))
colshapes_null <- c(rep(20, 8))
# differential gene expression volcano plot point colors
DEGcolors <- c("red", "blue", "dark grey")
# Create sample heatmap cell color
colonyclustercol <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
# Create gene expression cell color
paletteLength <- 100
# Red --> Blue gradient
geneexpcolors <- colorRampPalette(rev(c("red", "yellow", "white", "cyan", "blue")))(paletteLength)
# Orange --> Cyan gradient
geneexpcolors1 <- colorRampPalette(rev(c("chocolate1","#FEE090","grey10", "cyan3","cyan")))(paletteLength)
# Red --> Blue dark gradient
geneexpcolors2 <- colorRampPalette(rev(c("red", "yellow", "grey10", "cyan", "blue")))(paletteLength)
# Color bars to preview
#color.bar(geneexpcolors, -1, 1)
#color.bar(geneexpcolors1, -1, 1)
#color.bar(geneexpcolors2, -1, 1)
```
```{r theme}
theme_set(theme_ismej())
```

## Import sample metadata
```{r genotypes_fragments}
genotypes <- read_csv("data/genotypes.csv")
#
fragments <- read_csv("data/fragments.csv", col_types = cols(
  Fragment_ID = col_character()))
```
```{r factors}
# factor levels
treatment_levels <- c("Baseline", "Control", "Antibiotics")
sampling_levels <- c("Day 0", "Day 7")
genotype_levels <- c("PAN-05", "PAN-10", "PAN-32", "PAN-34", "PAN-35", "PAN-37", "PAN-39", "PAN-41", "PAN-43","PAN-44", "PAN-78", "PAN-79", "PAN-83", "URA-51", NA)
gulf_levels <- c("Chiriqui", "Panama")
loc_levels <- c("Uva", "Saboga", "Uraba")
mtorf_levels <- c("Type_1", "Type_3")
sym_levels <- c("C1bc", "C1d", "D1")
genus_levels <- c("Cladocopium", "Durusdinium")
```
```{r  sample_data, echo=TRUE}
# Import sample metadata 
sample_metadata <- read_delim("./data/quantseq_metadata.tsv", delim ="\t", col_names = c("Sample_ID", "Genotype", "Treatment", "Sampling_Time", "Sequencing_Success", "Sequencing_Round", "Include_Analysis"), comment = "#")#, row.names = 1, header = FALSE)
# join with genotype metadata
sample_metadata <- sample_metadata %>% left_join(genotypes, by = "Genotype")
# Set factor orders
# sample_metadata$Genotype <- factor(sample_metadata$Genotype, levels = genotype_levels, ordered = TRUE)
# sample_metadata$Treatment <- factor(sample_metadata$Treatment, levels = treatment_levels, ordered = TRUE)
sample_metadata$Gulf <- factor(sample_metadata$Gulf, levels = gulf_levels, ordered = TRUE)
sample_metadata$Location <- factor(sample_metadata$Location, levels = loc_levels, ordered = TRUE)
sample_metadata$mtORF <- factor(sample_metadata$mtORF, levels = mtorf_levels, ordered = TRUE)
sample_metadata$ITS2 <- factor(sample_metadata$ITS2, levels = sym_levels, ordered = TRUE)
sample_metadata$Symbiont_Genus <- ifelse(sample_metadata$ITS2 == "D1", "Durusdinium", "Cladocopium")
# Create analysis-compatible sample map
samples_success <- sample_metadata %>%
  dplyr::filter(Sequencing_Success == T)
```
```{r analysis_samples}
samples_analysis <- samples_success %>% 
  dplyr::filter(Include_Analysis == T, Treatment != "Baseline")
# 
samples_analysis$Sample_ID
```
```{r coldata}
# Get sample data into tibble
coldata <- samples_analysis %>%
  as_tibble(.) %>%
  mutate(mtORF_Treatment = interaction(mtORF, Treatment),
         Genotype_Treatment = interaction(Genotype, Treatment)) %>%
  mutate_if(is.character, as.factor) %>% 
  dplyr::filter(Include_Analysis == TRUE,
                Treatment != "Baseline") %>% 
  dplyr::arrange("Sample_ID") %>% #Order rows by sample name                         
  column_to_rownames(var = "Sample_ID")  
# 53 samples remaining for downstream analysis
```
```{r inspect_coldata}
coldata %>% 
  dplyr::count(Genotype, Treatment)
```

## Import *Pocillopora* coral counts data and gene annotation
```{r coral_countdata}
# Import and tidy counts data
countdata <- read.delim("./outputs/tagseq/STARcounts_Pdam/PocAnti_Pdam.counts", comment.char="#")
# countdata <- read.delim("./outputs/tagseq/STARcounts_Pdam/PocAnti_Pdam_Round2.counts", comment.char="#")
# Set Gene ID's as row names
row.names(countdata) <- countdata$Geneid
# Remove first six columns (Geneid, chr, start, end, strand, length)
countdata <- countdata[ ,7:ncol(countdata)]
#countdata <- countdata[ ,6:ncol(countdata)]
# Remove file prefixes and suffixes
  colnames(countdata) <- gsub("X.", "", colnames(countdata))
  colnames(countdata) <- gsub("scratch.nmnh_corals.connellym.projects.anti_phys.outputs.STARalign_Pdam.", "", colnames(countdata))
  colnames(countdata) <- gsub("_Aligned.sortedByCoord.out.uniq.bam$", "", colnames(countdata))
  colnames(countdata) <- gsub("_R1$", "", colnames(countdata))
  colnames(countdata) <- gsub("\\.", "-", colnames(countdata))
```
```{r coral_countdata_tidy}
# Select counts for samples present in coldata
countdata_select <- countdata[, colnames(countdata) %in% rownames(coldata)]
# Sort countdata by P. damicornis gene ID
countdata_sorted_coral <- as.data.frame(countdata_select[sort(rownames(countdata_select)), order(match(colnames(countdata_select),row.names(coldata)))]) 
# Sanity check
all(colnames(countdata_sorted_coral) == rownames(coldata))
```
```{r coral_gene_annotation}
gene_annotation_coral <- read.delim(file = "./data/pdam_genome_annotations.tsv", header = T) %>% arrange(ID)
rownames(gene_annotation_coral) <- gene_annotation_coral$ID
# Check gene feature annotation and countdata rowname order coherence
all(rownames(countdata_sorted_coral) == gene_annotation_coral$ID)
all(rownames(countdata_sorted_coral) ==  rownames(gene_annotation_coral))
# Obtain KOG annotations for P. damicornis genome
gene2kog_coral <- gene_annotation_coral %>%
  dplyr::select(ID, KOG_Class) %>% 
  filter(KOG_Class != "")
```

## Import *Cladocopium*/*Durusdinium* algal symbiont counts data and gene annotation
Not possible yet, alignments to Cladocopium and Durusdinium genome pending
```{r symbiont_countdata}
# # # Import and tidy counts data
# countdata <- read.delim("./outputs/tagseq/STARcounts_Symbio/PocAnti_Symbio.counts", comment.char="#")
# # ## Tidy counts data
# # #Set Gene ID's as row names
# row.names(countdata) <- countdata$Geneid
# # # Remove first six columns (Geneid, chr, start, end, strand, length)
# countdata <- countdata[ ,7:ncol(countdata)]
# # #countdata <- countdata[ ,6:ncol(countdata)]
# # # Remove file prefixes and suffixes
#   colnames(countdata) <- gsub("X.", "", colnames(countdata))
#   colnames(countdata) <- gsub("scratch.projects.transcriptomics.mikeconnelly.projects.anti_phys.outputs.STARalign_Symbio.", "", colnames(countdata))
#   colnames(countdata) <- gsub("_SymbioAligned.sortedByCoord.out.uniq.bam$", "", colnames(countdata))
#   colnames(countdata) <- gsub("\\.", "-", colnames(countdata))
```
```{r symbiont_countdata_tidy}
# # Select counts for samples present in coldata
# countdata_select <- countdata[, colnames(countdata) %in% rownames(coldata)]
# # Sort countdata by Cladocopium/Durusdinium gene ID
# countdata_sorted_symbiont <- as.data.frame(countdata_select[sort(rownames(countdata_select)), order(match(colnames(countdata_select),row.names(coldata)))]) 
# # Sanity check
# all(colnames(countdata_sorted_coral) == rownames(coldata))
```
```{r symbiont_gene_annotation}
# gene_annotation_symbiont <- read.delim("./data/cladocopium_genome_annotations.tsv", header = T) %>% arrange(ID)
# rownames(gene_annotation_symbiont) <- gene_annotation_symbiont$ID
# # Check gene feature annotation and countdata rowname order coherence
# all(rownames(countdata_sorted_symbiont) == gene_annotation_symbiont$ID)
# all(rownames(countdata_sorted_symbiont) == rownames(gene_annotation_symbiont))
# # Obtain KOG annotations for Cladocopium goreaui genome
# gene2kog_symbiont <- gene_annotation_symbiont %>% dplyr::select(ID, KOG_Class) %>% dplyr::filter(KOG_Class != "")
```

## Create DESeq datasets
### *Pocillopora* coral host
```{r create_dds_coral}
# Create full DESeqDataSet
dds_coral <- DESeqDataSetFromMatrix(countData = countdata_sorted_coral,
                              colData = coldata,
                              design = ~ Treatment)
# Check annotation and dds_coral object rowname order coherence
all(rownames(dds_coral) == rownames(gene_annotation_coral))
# Add gene feature annotation to DESeqDataSets
mcols(dds_coral) <- cbind(mcols(dds_coral), gene_annotation_coral)
# Subset DESeqDataSet
# Option 1: Remove genes with counts less than 10 in 90% of samples
# keep <- rowSums(counts(dds_coral) >= 10) > (ncol(countdata_sorted_coral)*0.9) # 5,527 genes
# Option 2: Remove genes with mean counts less than 10
# keep <- rowMeans(counts(dds_coral)) > 10 # 13,454 genes
# Option 3: Remove genes with mean counts less than 3 - as in Schlecker et al. (2022)
keep <- rowMeans(counts(dds_coral)) > 3 # 17,144 genes
dds_coral <- dds_coral[keep, ]
# Normalize expression data for visualization purposes using VST tranformation
vsd_coral <- vst(dds_coral, blind = TRUE) # use blind = TRUE to not account for experimental design
```

### *Cladocopium*/*Durusdinium* algal symbiont
```{r create_dds_symbiont}
# # Create full DESeqDataSet
dds_symbiont <- DESeqDataSetFromMatrix(countData = countdata_sorted_symbiont,
                              colData = coldata,
                              design = ~ Treatment)
# # Check annotation and dds_symbiont object rowname order coherence
# all(rownames(dds_symbiont) == rownames(gene_annotation_symbiont))
# # Add gene feature annotation to DESeqDataSets
# mcols(dds_symbiont) <- cbind(mcols(dds_symbiont), gene_annotation_symbiont)
# # Subset DESeqDataSet
# # Remove genes with counts less than 10 in 90% of samples
keep <- rowSums(counts(dds_symbiont) >= 10) > (ncol(countdata_sorted_symbiont)*0.9) #2,540 genes
dds_symbiont <- dds_symbiont[keep, ]
# # Normalize expression data for visualization purposes using VST transformation
vsd_symbiont <- vst(dds_symbiont, blind = TRUE) # use blind = TRUE to not account for experimental design
```

## Visualize global gene expression
```{r choose_vsd, include=FALSE}
# Choose vsd for host/symbiont analysis with PCoA and WGCNA
vsd <- vsd_coral
# vsd <- vsd_symbiont
```
```{r vsd_batcheffect_remove}
vsd_matrix_corrected <- limma::removeBatchEffect(assay(vsd), vsd$Sequencing_Round)
# use batch-colony-removed vst counts
dat <- data.frame(vsd_matrix_corrected) 
```

### Principal Components Analysis (PCA)
```{r}
DESeq2::plotPCA(vsd, intgroup = c("Genotype"))
```
```{r pca}
### General style with ellipses around treatment
ntop <- 2000
# pdf("./outputs/figures/transcriptome_pca_control_anti.pdf")
pca_title <- expression(paste(italic("Pocillopora"), " gene expression PCA plot", sep = ""))
# 
ggPCA(vsd_coral, samples, genocolors, pclab = c(1,2), ntop = ntop) +
  # facet_wrap("Genotype", ncol = 7) +
  ggtitle(pca_title)
# dev.off()
```
```{r pca}
### Modified style with convex hulls around treatment
# ntop <- 2000
# pdf("./outputs/figures/transcriptome_pca_control_anti.pdf")
# pca_title <- expression(paste(italic("Pocillopora"), " gene expression PCA plot", sep = ""))
# 
pca1 <- ggPCA_mod1(vsd_coral, samples, genocolors, pclab = c(1,2), ntop = ntop) +
  # facet_wrap("Genotype", ncol = 7) +
  ggtitle("Experimental treatment") +
  # guides(color = guide_legend(title = NULL), color = NULL) +
  theme(legend.title = element_blank(), legend.margin = margin(0,0,0,0, unit = "pt"))
# dev.off()
print(pca1)
```
```{r pca}
### Modified style with spider points connecting genotypes
# ntop <- 2000
# pdf("./outputs/figures/transcriptome_pca_control_anti.pdf")
pca_title <- expression(paste(italic("Pocillopora"), " genotype", sep = ""))
# 
pca2 <- ggPCA_mod2(vsd_coral, samples, genocolors, pclab = c(1,2), ntop = ntop) +
  # facet_wrap("Genotype", ncol = 7) #+
  ggtitle(pca_title) +
  theme(legend.title = element_blank(), legend.margin = margin(0,0,0,0, unit = "pt"))
# dev.off()
print(pca2)
```

### Figure 4AB
```{r figure 4AB: powerpoint}
pdf("./outputs/figures/transcriptome_pca.pdf", height = 4.5, width = 6.5)
pca_title <- expression(paste(italic("Pocillopora"), " transcriptome PCA", sep = ""))
# 
# ggPCA(vsd_coral, samples, genocolors, pclab = c(1,2)) +
#   # facet_wrap("Genotype") +
#   theme_ismej() +
#   theme(plot.title = element_text(size = 16),
#         strip.text = element_text(size = 14),
#         axis.title = element_text(size = 14),
#         axis.text = element_text(size = 12),
#         legend.title = element_text(size = 14),
#         legend.text = element_text(size = 12)) +
#   ggtitle(pca_title)
pca12 <- (pca1 | pca2) + plot_annotation(title = pca_title)
print(pca12)
dev.off()
```
```{r figure 4AB: manuscript}
pdf("./outputs/figures/Fig4AB_transcriptome_pca.pdf", height = 68/25.4, width = 175/25.4)
pca_title <- expression(paste(italic("Pocillopora"), " transcriptome PCA", sep = ""))
# # 
pca12 <- (pca1 | pca2) + plot_layout(guides = "collect")
print(pca12)
dev.off()
```

#### Comprehensively interrogate dataset using PCAtools
```{r pcatools_pca}
pca_vsd <- pca(assay(vsd), metadata = colData(dds_coral), removeVar = ((nrow(vsd) - ntop)/(nrow(vsd))))
#
getComponents(pca_vsd)
#
getVars(pca_vsd)
#
pca_loadings <- getLoadings(pca_vsd) %>% 
  rownames_to_column("ID") %>% 
  dplyr::select(ID:PC10) %>% 
  left_join(gene_annotation_coral, by = "ID")
```
```{r pcatools_scree}
screeplot(pca_vsd, axisLabSize = 10, titleLabSize = 12)
```
```{r pca_biplot_overview}
biplot(pca_vsd, 
       colby = 'Treatment', colkey = c('Baseline' = treatcolors[1], 'Control' = treatcolors[2], 'Antibiotics' = treatcolors[3]),
       encircle = TRUE,
       encircleFill = TRUE,
       showLoadings = TRUE,
       # lab = pca_vsd$metadata$,
       labSize = 2, 
       pointSize = 2,
       sizeLoadingsNames = 2) + theme_ismej()
```
```{r pca_pairsplot}
pairsplot(pca_vsd,
          colby = 'Treatment', colkey = c('Baseline' = treatcolors[1], 'Control' = treatcolors[2], 'Antibiotics' = treatcolors[3]))
```
```{r pca_gene_loadings}
pdf("./outputs/figures/transcriptome_pcatools_loadings.pdf", height = 6, width = 6.5)
plotloadings(pca_vsd, labSize = 1.5) + theme_ismej()
dev.off()
```
```{r pca_eigencor}
# Need to account for numerical variables
eigencorplot(pca_vsd, metavars = c("Treatment", "Gulf", "mtORF", "haplotype", "Genotype", "Symbiont_Genus", "Sequencing_Round"))
```

### Principal Coordinate Analysis (PCoA)
  Must choose a vsd object (coral or symbiont) to use for the PCoA 
```{r mds, eval = TRUE}
# Calculate distances among samples
sampleDists <- dist(t(assay(vsd)), method = "manhattan")
sampleDistMatrix <- as.matrix(sampleDists)
# Calculate MDS
mds <- as.data.frame(colData(vsd)) %>% 
  cbind(cmdscale(sampleDistMatrix))
# Calculate MDS and use eigenvectors to determine proport
mds_eig <- cmdscale(sampleDistMatrix, eig = TRUE)
mds_eigenvectors <- data.frame(mds_eig$eig) %>% 
  mutate(prop_var = mds_eig.eig / sum(mds_eig.eig))
# create axis labels for plotting
xlab <- paste("PC1 (", round(mds_eigenvectors$prop_var[1]*100, digits = 1), "%)", sep = "")
ylab <- paste("PC2 (", round(mds_eigenvectors$prop_var[2]*100, digits = 1), "%)", sep = "")
# Calculate Colony centroids for plotting
mds_col <- mds %>% 
  group_by(Genotype) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)
# set factor orders 
# mds_col$Colony <- factor(mds$Colony, levels = c("HW1", "HW2", "WT1", "WT2"), ordered = TRUE)
# mds_col$Treatment <- factor(mds$Treatment, levels = c("control", "Heat", "Antibiotics", "Antibiotics.Heat"), ordered = TRUE)
# Calculate Treatment centroids for plotting
mds_trmt <- mds %>%
  group_by(Treatment) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)
#set factor orders 
# mds_trmt$Colony <- factor(mds_trmt$Colony, levels = c("HW1", "HW2", "WT1", "WT2"), ordered = TRUE)
mds_trmt$Treatment <- factor(mds_trmt$Treatment, levels = treatment_levels, ordered = TRUE)
```
```{r pcoa_treatment}
pcoa_treatment_coral <- ggPCoA(mds_trmt)
# pcoa_treatment_symbiont <- ggPCoA(mds_trmt)
# 
# Save single treatment PCoA plot
pcoa_plot <- pcoa_treatment_coral +
  # facet_wrap(~Genotype) +
  theme_ismej()
# pcoa_plot <- pcoa_treatment_symbiont
#
# pdf(file = "./outputs/DESeq_results/figures/pcoa_treatment.pdf", width = 6, height = 4)
print(pcoa_plot)
# dev.off()
```

### PERMANOVA tests
  Must choose a vsd object (coral or symbiont) to use for the analysis
```{r transcriptome_distance_matrices}
# obtain vst-transformed counts
countdata_vst <- assay(vsd)
# Convert to matrix and transpose, check dimensions
datExpr <- t(countdata_vst)
# Calculate Aitchinson distance matrix
tx_dist <- vegdist(datExpr, method = "manhattan")
# Now obtain data frame from the sample_data
# head(sample_metadata)
# Create factors for statistical testing
treatments <- as.character(coldata$Treatment)
colonies <- as.character(coldata$Genotype)
locations <- as.character(coldata$Location)
mtorfs <- as.character(coldata$mtORF)
haplotypes <- as.character(coldata$haplotype)
genus <- as.character(coldata$Symbiont_Genus)
```
```{r transcriptome_permanova}
# PERMDISP tests
# Treatment
beta_treatment <- betadisper(tx_dist, treatments)
permutest(beta_treatment)
# Genotype
beta_genotype <- betadisper(tx_dist, colonies)
permutest(beta_genotype)
# Symbiont type
beta_genus <- betadisper(tx_dist, genus)
permutest(beta_genus)
# Adonis/PERMANOVA tests
# Treatment
permanova_treatment <- adonis(tx_dist ~ treatments, data = coldata)
print(permanova_treatment)
# Genotype
permanova_genotype <- adonis(tx_dist ~ colonies, data = coldata)
print(permanova_genotype)
# Symbiont Type
permanova_genus <- adonis(tx_dist ~ genus, data = coldata)
print(permanova_genus)
```
```{r transcriptome_pairwise_permanova}
# Pairwise Adonis/PERMANOVA tests
# Treatment
pairwise_permanova_treatment <- pairwise.adonis(tx_dist, treatments)
print(pairwise_permanova_treatment)
# write_csv(pairwise_permanova_treatment, "./outputs/pairwise_permanova_treatment_symbiont.csv")
# Colony (Genotype)
pairwise_permanova_colony <- pairwise.adonis(tx_dist, colonies)
print(pairwise_permanova_colony)
# write_csv(pairwise_permanova_colony, "./outputs/pairwise_permanova_colony_symbiont.csv")
```

## VariancePartition
```{r variance_partition}
# Partition total transcriptome variance among explanatory factors including sequencing round, coral genotype, mtORF lineage, treatment, and their interaction
formula <- ~ (1|Sequencing_Round) + (1|Gulf) + (1|Genotype) + (1|Symbiont_Genus) + (1|Treatment)
#
cores <- detectCores()
cl <- parallel::makeCluster(cores[1]-1)
doParallel::registerDoParallel(cl)
varpart_total <- fitExtractVarPartModel(assay(vsd), formula, coldata)
stopCluster(cl)
# sort variables (i.e. columns) by median fraction of variance explained
vpt <- sortCols(varpart_total, FUN = mean)
```
```{r varpart_violin}
# Violin plot of contribution of each variable to total variance
varplot <- plotVarPart(vpt)
vartitle <- expression(paste(italic("Pocillopora"), " spp. transcriptome variance partition"))
varplot <- varplot + 
    scale_fill_viridis(discrete = T) +
  ggtitle(vartitle)
print(varplot)
ggsave(varplot, filename = "./outputs/figures/variancePartition_total.pdf", width = 120, height = 85, units = "mm", device = "pdf")
```
```{r varpart_sum}
vptdf <- as.data.frame(vpt) %>% rownames_to_column("gene")
varpct <- vptdf %>% pivot_longer(cols = -gene, names_to = "source", values_to = "value") %>% 
  mutate(relvar = value / sum(value)) %>%
  mutate(source = str_replace(source, "_", " ")) %>% 
  group_by(source) %>% 
  # filter(source != "Residuals") %>%
  summarise(total_pctvar_explained = sum(relvar))
```

### Supplementary Figure 10
```{r figure_S10: manuscript}
pdf("./outputs/figures/FigS10_variance_partition_totals.pdf", height = 4, width = 6.5)
var_title <- expression(paste("Transcriptome variance factors"))
# 
var_sum_plot <- varpct %>% 
  mutate(source = fct_reorder(source, order(colnames(vpt)))) %>% 
  dplyr::filter(source != "Residuals") %>% 
  ggplot(aes(source, total_pctvar_explained)) +
  geom_col(aes(fill = source), color = "black", show.legend = FALSE) +
  ylab("Total percent variance explained") +
  scale_fill_viridis(discrete = T, name = "Factor") +
  scale_y_continuous(limits = c(0,0.1), labels = scales::percent) +
  theme(axis.text.x = element_text(angle = 270, hjust = 0),
        axis.title.x = element_blank()) + 
  ggtitle(var_title)
print(var_sum_plot)
dev.off()
```

## Differential gene expression analysis (DESeq2)
  Must choose a dds object and gene annotation (coral or symbiont) to use for the DESeq2 analysis
```{r choose_dds_annotation}
# # Choose DESeq dataset and annotation for coral or symbiont differential gene expression analysis
dds <- dds_coral
gene_annotation <- gene_annotation_coral
partner <- "coral"
# #
# dds <- dds_symbiont
# gene_annotation <- gene_annotation_symbiont
# partner <- "symbiont"
```
```{r genotype_DESeq}
#First, genotype-specific treatment models
design(dds) <- formula(~ Sequencing_Round + Genotype_Treatment)
#Set control treatment in URA-51 as reference factor level
dds$Genotype_Treatment <- relevel(dds$Genotype_Treatment, ref = "URA-51.Control")
#Perform DESeq2 analysis
dsr1 <- DESeq(dds)
resultsNames(dsr1)
# Define contrasts - only one option to contrast Antibiotics vs. Control
contrasts <- tibble(num = c("Antibiotics"), 
                    den = c("Control"))
# Create tibble with DESeq results for contrasts of Antibiotics vs. Control for each genotype
DE <- crossing(Genotype = c("PAN-32", "PAN-35", "PAN-39", "PAN-78", "PAN-79", "PAN-83", "URA-51"), contrasts) %>%
# A tibble: 7 Ã— 3
#   Genotype num         den    
#   <chr>    <chr>       <chr>  
# 1 PAN-32   Antibiotics Control
# 2 PAN-35   Antibiotics Control
# 3 PAN-39   Antibiotics Control
# 4 PAN-78   Antibiotics Control
# 5 PAN-79   Antibiotics Control
# 6 PAN-83   Antibiotics Control
# 7 URA-51   Antibiotics Control
   dplyr::mutate(dsr = pmap(list(Genotype, num, den), function(x, y, z) {
   results(dsr1, contrast = c("Genotype_Treatment", paste0(x, ".", c(y, z))))}))
#
```
```{r overall_DESeq}
#Second, overall genotype and treatment interaction model
# ~ Sequencing_Round + Genotype + Treatment #+ Colony:Treatment
design(dds) <- formula(~ Sequencing_Round + Genotype + Treatment)
#Set control treatment as reference factor level
dds$Treatment <- factor(dds$Treatment, levels = c("Control", "Antibiotics"))
dds$Treatment <- relevel(dds$Treatment, ref = "Control")
#Perform DESeq2 analysis
dsr2 <- DESeq(dds)
resultsNames(dsr2)
# Get DESeq results for all treatment contrasts
DE <- crossing(Genotype = "all", contrasts) %>%
  mutate(dsr = map2(num, den, ~ results(dsr2, contrast = c("Treatment", .x, .y)))) %>%
  bind_rows(DE)
```
```{r DE_table}
DE <- DE %>%
  mutate(df = map(dsr, ~ rownames_to_column(data.frame(.), "ID")),
         df = map(df, ~ left_join(., gene_annotation, by = "ID")))

# Filter out significant up and down-regulated genes
DE <- DE %>%
  mutate(sig = map(df, ~ filter(., padj < 0.05)),
          up = map(sig, ~ filter(., log2FoldChange > 0)),
        down = map(sig, ~ filter(., log2FoldChange < 0))) 

# Generate signed logP values for KOG MWU tests of all differential expression contrasts 
DE <- DE %>% mutate(logP = map(dsr, ~ data.frame(
  gene = rownames(data.frame(.)),
  logP = -log10(data.frame(.)$pvalue) * sign(data.frame(.)$log2FoldChange))))

# Generate binary values for KOG Fisher's tests of all differential expression contrasts
# Upregulated genes = 1, all others = 0
DE <- DE %>%
  mutate(sig_binary_up = map(dsr, ~ data.frame(
    gene = rownames(data.frame(.)),
    value = ifelse(data.frame(.)$padj < 0.05 & data.frame(.)$log2FoldChange > 0, 1, 0))))

# Downregulated genes = 1, all others = 0
DE <- DE %>%
  mutate(sig_binary_down = map(dsr, ~ data.frame(
    gene = rownames(data.frame(.)),
    value = ifelse(data.frame(.)$padj < 0.05 & data.frame(.)$log2FoldChange < 0, 1, 0))))
```
```{r DE_check}
DE
warnings()
```
```{r DE_store}
# Store DE results tibbles
# DE_coral <- DE
# DE_symbiont <- DE
```
```{r DE_load}
# DE <- DE_coral
# DE <- DE_symbiont
```
```{r DEG_contrasts_table}
# Count number of differentially expressed genes within each colony, and overall, for each contrast
DEtab <- DE %>%
  mutate(nsig = map_dbl(sig, ~ nrow(.)),
         nup = map_dbl(up, ~ nrow(.)),
         ndn = map_dbl(down, ~ nrow(.)),
         `DEGs [up, down]` = paste0(nsig, " [", nup, ", ", ndn, "]")) %>% 
  mutate(Genotype = factor(Genotype, levels = c("PAN-32", "PAN-35", "PAN-39", "PAN-78", "PAN-79", "PAN-83", "URA-51", "all"), ordered = TRUE)) %>%
  arrange(Genotype, num, den) %>% 
  filter(den == "Control") %>% 
  mutate("Contrast" = paste0(num, " vs. ", den)) %>% 
  dplyr::select(Genotype, Contrast, `DEGs [up, down]`) %>% 
  pivot_wider(names_from = Genotype, values_from = `DEGs [up, down]`)
```
```{r DE_contrasts_summary_table}
# Store DE results tables
# DEtab_coral <- DEtab
# DEtab_symbiont <- DEtab
```
```{r Table_1}
# Bind together coral and symbiont DEG contrast results
# DE_table <- bind_rows("Pocillopora" = DEtab_coral, "Cladocopium" = DEtab_symbiont, .id = "Holobiont Partner")
# Write output to file
# DE_table %>%
#   write_csv("./outputs/DESeq_results/tables/DE_table.csv")
```

### Obtain results tables
```{r Write_significant_results_tables}
# Write significant DEG results tables to csv files
DE %>%
  filter(den == "Control") %>%
  mutate(sig_file_out = paste0(num, "_", den, "_", Genotype, "_", partner, "_", "sig", ".csv"),
         sig_file_out_path = file.path("./outputs/DESeq_results/tables/sig/", sig_file_out),
         data = walk2(sig, sig_file_out_path, write_csv))
# Write ALL DEG results tables to csv files
DE %>%
  filter(den == "control") %>%
  mutate(file_out = paste0(num, "_", den, "_", Genotype, "_", partner, "_", "all", ".csv"),
         file_out_path = file.path("./outputs/DESeq_results/tables/all/", file_out),
         data = walk2(df, file_out_path, write_csv))
```

### Assess genotype response similarity
```{r DEgenes_colony_comparison, message=FALSE}
shared_DEGs_genotypes <- DE %>%
  unnest(sig) %>%
  filter(Genotype != "all") %>%
  filter(den == "Control") %>% 
  group_by(num, den) %>%
  dplyr::count(ID, log2FoldChange > 0) %>%  # In how many colonies is same gene DE in same direction?
  summarise(`1 genotype` = sum(n==1), `2 genotypes` = sum(n==2), `3 genotypes` = sum(n==3), `4 genotypes` = sum(n==4), `5 genotypes` = sum(n==5), `6 genotypes` = sum(n==6), `7 genotypes` = sum(n==7))
```
```{r DEG_intersects}
# DE 
ressig_PAN_32 <- DE$sig[DE$Genotype == "PAN-32"][[1]]$ID
ressig_PAN_35 <- DE$sig[DE$Genotype == "PAN-35"][[1]]$ID
ressig_PAN_39 <- DE$sig[DE$Genotype == "PAN-39"][[1]]$ID
ressig_PAN_78 <- DE$sig[DE$Genotype == "PAN-78"][[1]]$ID
ressig_PAN_79 <- DE$sig[DE$Genotype == "PAN-79"][[1]]$ID
ressig_PAN_83 <- DE$sig[DE$Genotype == "PAN-83"][[1]]$ID
ressig_URA_51 <- DE$sig[DE$Genotype == "URA-51"][[1]]$ID
# 
listInput <- list("PAN-32" = ressig_PAN_32, 
                  "PAN-35" = ressig_PAN_35, 
                  "PAN-39" = ressig_PAN_39, 
                  "PAN-78" = ressig_PAN_78, 
                  "PAN-79" = ressig_PAN_79, 
                  "PAN-83" = ressig_PAN_83, 
                  "URA-51" = ressig_URA_51)
#
sets_genotypes <- rev(c("PAN-32", "PAN-35", "PAN-39", "PAN-78", "PAN-79","PAN-83", "URA-51"))
# pdf("./outputs/figures/DEG_upset_plot.pdf", width = 11, height = 8.5)
upset(fromList(listInput), 
      sets = sets_genotypes, keep.order = T, nsets = 7, 
      empty.intersections = "on",
      nintersects = NA,
      # order.by = "degree",
      order.by = c("degree", "freq"),
      number.angles = 30, point.size = 1.5, line.size = 0.5, 
      sets.x.label = "# DEGs per genotype", mainbar.y.label = "# DEGs shared among genotypes")
# dev.off()
```
#### Supplementary Figure 11
```{r DEG_intersects}
# DE 
resup_PAN_32 <- DE$up[DE$Genotype == "PAN-32"][[1]]$ID
resup_PAN_35 <- DE$up[DE$Genotype == "PAN-35"][[1]]$ID
resup_PAN_39 <- DE$up[DE$Genotype == "PAN-39"][[1]]$ID
resup_PAN_78 <- DE$up[DE$Genotype == "PAN-78"][[1]]$ID
resup_PAN_79 <- DE$up[DE$Genotype == "PAN-79"][[1]]$ID
resup_PAN_83 <- DE$up[DE$Genotype == "PAN-83"][[1]]$ID
resup_URA_51 <- DE$up[DE$Genotype == "URA-51"][[1]]$ID
# 
listInput <- list("PAN-32" = resup_PAN_32, 
                  "PAN-35" = resup_PAN_35, 
                  "PAN-39" = resup_PAN_39, 
                  "PAN-78" = resup_PAN_78, 
                  "PAN-79" = resup_PAN_79, 
                  "PAN-83" = resup_PAN_83, 
                  "URA-51" = resup_URA_51)
#
# pdf("./outputs/figures/DEG_upset_plot_up.pdf", width = 11, height = 8.5)
upset_up <- upset(fromList(listInput), 
      sets = sets_genotypes, keep.order = T, nsets = 7, 
      # empty.intersections = "on",
      nintersects = NA,
      order.by = "degree",
      # order.by = c("degree", "freq"),
      number.angles = 0, point.size = 1.5, line.size = 0.5, 
      text.scale = c(1, 1, 1, 1, 1, 0.5),
      sets.x.label = "# DEGs per genotype", mainbar.y.label = "# DEGs shared among genotypes")
# c(intersection size title, intersection size tick labels, set size title, set size tick labels, set names, numbers above bars)

pdf("./outputs/figures/FigS11A_DEG_upset_plot_up.pdf", width = 6.5, height = 4)
upset_up
dev.off()
```
```{r DEG_intersects}
# DE 
resdown_PAN_32 <- DE$down[DE$Genotype == "PAN-32"][[1]]$ID
resdown_PAN_35 <- DE$down[DE$Genotype == "PAN-35"][[1]]$ID
resdown_PAN_39 <- DE$down[DE$Genotype == "PAN-39"][[1]]$ID
resdown_PAN_78 <- DE$down[DE$Genotype == "PAN-78"][[1]]$ID
resdown_PAN_79 <- DE$down[DE$Genotype == "PAN-79"][[1]]$ID
resdown_PAN_83 <- DE$down[DE$Genotype == "PAN-83"][[1]]$ID
resdown_URA_51 <- DE$down[DE$Genotype == "URA-51"][[1]]$ID
# 
listInput <- list("PAN-32" = resdown_PAN_32, 
                  "PAN-35" = resdown_PAN_35, 
                  "PAN-39" = resdown_PAN_39, 
                  "PAN-78" = resdown_PAN_78, 
                  "PAN-79" = resdown_PAN_79, 
                  "PAN-83" = resdown_PAN_83, 
                  "URA-51" = resdown_URA_51)
#
upset_down <- upset(fromList(listInput), 
      sets = sets_genotypes, keep.order = T, nsets = 7, 
      # empty.intersections = "on",
      nintersects = NA,
      order.by = "degree",
      # order.by = c("degree", "freq"),
      number.angles = 0, point.size = 1.5, line.size = 0.5, 
      text.scale = c(1, 1, 1, 1, 1, 0.5),
      sets.x.label = "# DEGs per genotype", mainbar.y.label = "# DEGs shared among genotypes")
#
pdf("./outputs/figures/FigS11B_DEG_upset_plot_down.pdf", width = 6.5, height = 4)
upset_down
dev.off()
```

### DESeq Gene Ontology (GO) enrichment analysis
#### Antibiotics vs. Control 
```{r anti_signed_logP}
# Write overall Antibiotics vs. control logP values to file for GO_MWU analysis
logPfilepath <- paste0("./R/GO_MWU/logP", "_", "antibiotics.control", "_", partner, ".txt")
DE %>%
  filter(Genotype == "all", num == "Antibiotics", den == "Control") %>%
  unnest(logP) %>%
  dplyr::select(gene, logP) %>%
  write.table(., file = logPfilepath, row.names = FALSE, quote = FALSE, sep = ",")
#
logPfile <- paste0("logP", "_", "antibiotics.control", "_", partner, ".txt")
```

```{r anti GO_MWU plot}
# Plot GO_MWU results for Antibiotics vs. control 
commandArgs <- function(...) c("logP_antibiotics.control_coral.txt", "BP", "GO_MWU.anti.control.pdf", 0.01, 1e-3, 1e-4, 0.88)
source("./R/GO_MWU/GO_MWU_plot.R")
commandArgs <- function(...) c("logP_antibiotics.control_coral.txt", "MF", "GO_MWU.anti.control.pdf", 0.01, 0.01, 0.001, 0.88)
source("./R/GO_MWU/GO_MWU_plot.R")
commandArgs <- function(...) c("logP_antibiotics.control_coral.txt", "CC", "GO_MWU.anti.control.pdf", 0.01, 1e-3, 1e-4, 0.88)
source("./R/GO_MWU/GO_MWU_plot.R")
```

```{r anti_significant_binary_up}
# Write overall Antibiotics vs. control upregulated binary values to file for GO_MWU analysis
binaryfilepath <- paste0("./R/GO_MWU/sig_up", "_", "antibiotics.control", "_", partner, ".txt")
DE %>%
  filter(Genotype == "all", num == "Antibiotics", den == "Control") %>%
  unnest(sig_binary_up) %>%
  dplyr::select(gene, value) %>%
  drop_na(value) %>% 
  write.table(., file = binaryfilepath, row.names = FALSE, quote = FALSE, sep = ",")
#
binaryfile <- paste0("sig_up", "_", "antibiotics.control", "_", partner, ".txt")
```
```{r anti_significant_binary_down}
# Write overall Antibiotics vs. control downregulated binary values to file for GO_MWU analysis
binaryfilepath <- paste0("./R/GO_MWU/sig_down", "_", "antibiotics.control", "_", partner, ".txt")
DE %>%
  filter(Genotype == "all", num == "Antibiotics", den == "Control") %>%
  unnest(sig_binary_down) %>%
  dplyr::select(gene, value) %>% 
  drop_na(value) %>% 
  write.table(., file = binaryfilepath, row.names = FALSE, quote = FALSE, sep = ",")
#
binaryfile <- paste0("sig_down", "_", "antibiotics.control", "_", partner, ".txt")
```
```{r anti_GO_MWU}
# Run GO_MWU for Antibiotics vs. control for Biological Process (BP) GO terms
commandArgs <- function(...) c(binaryfile, "BP", "pdam_emapper_gene2go.tab") #logPfile, "cladocopium_gene2go.tab"
source("./R/GO_MWU/GO_MWU.R")

# Run GO_MWU for Antibiotics vs. control for Molecular Function GO terms
commandArgs <- function(...) c(logPfile, "MF", "pdam_emapper_gene2go.tab") #"pdam_emapper_gene2go.tab"
source("./R/GO_MWU/GO_MWU.R")

# Run GO_MWU for Antibiotics vs. control for Cellular Component (CC) GO terms
commandArgs <- function(...) c(binaryfile, "CC", "pdam_emapper_gene2go.tab")
source("./R/GO_MWU/GO_MWU.R")
```
```{r anti.control_GO}
anti_control_GO_logP_coral <- as.list(
  c(BP = "./R/GO_MWU/MWU_BP_logP_antibiotics.control_coral.txt",
    MF = "./R/GO_MWU/MWU_MF_logP_antibiotics.control_coral.txt",
    CC = "./R/GO_MWU/MWU_CC_logP_antibiotics.control_coral.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(p.adj < 0.05)
#
# Table S8
anti_control_GO_logP_coral %>%
  dplyr::select(ontology, delta.rank, p.adj, name, nseqs, level, term) %>%
  arrange(ontology, p.adj) %>% 
  write_csv(path = "./outputs/DESeq_results/TableS8_anti.control.GO.csv")

anti_control_GO_logP_coral %>% 
  filter(p.adj < 0.05) %>% 
  # filter(delta.rank <= 0) %>% 
  group_by(ontology) %>% 
  dplyr::summarise(n = n())
```
```{r anti_control_GO_up_coral}
anti_control_GO_up_coral <- as.list(
  c(BP = "./R/GO_MWU/MWU_BP_sig_up_antibiotics.control_coral.txt",
    MF = "./R/GO_MWU/MWU_MF_sig_up_antibiotics.control_coral.txt",
    CC = "./R/GO_MWU/MWU_CC_sig_up_antibiotics.control_coral.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(pval < 0.05)
#
anti_control_GO_up_coral %>% filter(p.adj < 0.05) %>% 
  group_by(ontology) %>% 
  dplyr::summarise(n = n())
#
anti_control_GO_up_coral %>% filter(p.adj < 0.05, ontology == "BP") 
```
```{r anti_control_GO_down_coral}
anti_control_GO_down_coral <- as.list(
  c(BP = "./R/GO_MWU/MWU_BP_sig_down_antibiotics.control_coral.txt",
    MF = "./R/GO_MWU/MWU_MF_sig_down_antibiotics.control_coral.txt",
    CC = "./R/GO_MWU/MWU_CC_sig_down_antibiotics.control_coral.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(pval < 0.05)
#
anti_control_GO_down_coral %>% filter(p.adj < 0.05) %>% 
  group_by(ontology) %>% 
  dplyr::summarise(n = n())
#
anti_control_GO_down_coral %>% filter(p.adj < 0.1) %>% filter(ontology == "BP") 
```

## Weighted gene co-expression network analysis (WGCNA)
```{r wgcna_setup_datExpr}
allowWGCNAThreads(nThreads = 4)
# obtain vst-transformed counts
countdata_vst <- assay(vsd)
# Convert to matrix and transpose, check dimensions
datExpr <- t(countdata_vst)
dim(datExpr)
```
```{r soft_threshold}
# Find correlation power R^N that satisfies scale free critereon (SFT.R.sq>0.9)
sft <- pickSoftThreshold(datExpr, verbose = 1)
sft$powerEstimate #5
```
### Step-by-step network construction
```{r adjacency_TOM}
# This follows the tutorial: 
# https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/FemaleLiver-02-networkConstr-man.pdf
# 
# I have chosen the following network construction parameters for the following reasons:
# First, following the recommendations of the WGCNA developers (https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html), a signed network was chosen to be able to detect positive and negative gene correlations, and the biweight midcorrelation was used since it is more robust to outliers. 

adjacency <- adjacency(datExpr,
      # Network construction arguments:  correlation, adjacency function,  and topological overlap map options
                       corFnc = "bicor", # bimidweight correlation
                       power = sft$powerEstimate, # 5 for coral, unknown for symbiont
                       type = "signed") # signed
# 
TOM <- TOMsimilarity(adjacency,
                     TOMType = "signed",
                     verbose = 5)
dissTOM <- 1-TOM
# 
rm(adjacency) # may need to delete adjacency, TOM to clear up vector memory
```
```{r gene_dendrogram}
geneTree <- flashClust(as.dist(dissTOM), method = "average")
plot(geneTree, labels = FALSE, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity")
```
```{r dynamic_treecut}
# minModuleSize between 30 to 100 because I prefer large modules that lend themselves to enrichment tests with GO_MWU for easier interpretation.
minModuleSize <- 30
# Module identification using dynamic tree cut, recommendations of the WGCNA developers are cutHeight = 0.99 and deepSplit = 2, I prefer deepSplit = 3 to get more modules here
dynamicMods <- cutreeDynamic(dendro = geneTree,
                             distM = dissTOM,
                             cutHeight = 0.99,
                             deepSplit = 3,
                             pamRespectsDendro = FALSE,
                             minClusterSize = minModuleSize)
table(dynamicMods)
# Convert numeric lables into colors
dynamicColors <- labels2colors(dynamicMods, colorSeq = standardColors())
```
```{r module_dendrogram}
# Calculate eigengenes
MEList <- moduleEigengenes(datExpr, colors = dynamicColors)
MEs <- MEList$eigengenes
# Calculate dissimilarity of module eigengenes
MEDiss <- 1-cor(MEs);
# Cluster module eigengenes
METree <- flashClust(as.dist(MEDiss), method = "average");
# Plot the result
plot(METree, main = "Clustering of module eigengenes",
xlab = "", sub = "")
MEDissThres = 0.3
# Plot the cut line into the dendrogram
abline(h=MEDissThres, col = "red")
```
```{r merge_modules}
# Call an automatic merging function
# merge cutHeight = 0.3 because I want to group together modules with >70% similar module eigengene expression, as in Schlecker et al. (2022)
mergedMods <- mergeCloseModules(datExpr, dynamicColors, cutHeight = 0.3, verbose = 1)
# The merged module colors
mergedColors <- mergedMods$colors
table(mergedColors) 

# 17 modules in final network
```

### Supplementary Figure 12
```{r dendrogram_modules}
pdf(file = "./outputs/figures/FigS12_WGCNA_dendrogram_coral.pdf", width = 165.1/25.4, height = 101.6/25.4)
# pdf(file = "./manuscript_figures/FigS##_WGCNA_dendrogram.pdf", width = 8, height = 6)
# Plot the dendrogram and the module colors underneath
plotDendroAndColors(dendro = geneTree, 
                    colors = mergedColors, #cbind(dynamicColors, mergedColors),
                    groupLabels = "Modules", #c("Dynamic modules", "Merged dynamic"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    abHeight = c(0.99))
dev.off()
```
```{r merged_eigengenes}
MEList <- moduleEigengenes(datExpr, colors = mergedColors)
mergedMEs <- MEList$eigengenes
```
```{r choose_modules}
moduleColors <- mergedColors#netColors#
MEs1 <- mergedMEs#netMEs#
# MEs1 <- MEs1 %>% select(-MEgrey)
MEmodule_colors <- colnames(MEs1)
module_colors <- gsub("ME", "", colnames(MEs1))
colnames(MEs1) <- module_colors
```
```{r ME_clustering}
datME <- MEs1
dissimME <- (1-t(cor(datME, method="p")))/2
hclustME <- flashClust(as.dist(dissimME), method="average" )
# Plot the eigengene dendrogram
par(mfrow=c(1,1))
plot(hclustME, main="Clustering tree based on the module eigengenes")
```
```{r uniq_modules}
# Extract all unique modules
uniqModules <- unique(colnames(MEs1))
# set in order according to ME clustering dendogram
hclustME$order
uniqModules <- uniqModules[hclustME$order]
#
uniqModules
# create ordered factor for downstream analysis
modules_factor <- factor(levels = uniqModules[hclustME$order], ordered = TRUE)
```
### Module genes and hub genes
```{r modules_genes}
# Generate GeneIDs
Genes <- colnames(datExpr)
# Output genes and annotations for each module
for (module in uniqModules)
{
# Select module genes
inModule <- (moduleColors == module)
# Get gene IDs
modGeneIDs <- Genes[inModule] #this is the correct set of gene IDs!
# Write gene annotations into a file
fileName = paste("./outputs/WGCNA_results/notes/", module, ".csv", sep="");
module_annotation <- gene_annotation[modGeneIDs, ]
write_csv(module_annotation, path = fileName)
}
```
```{r hub_genes}
hubs <- chooseTopHubInEachModule(datExpr,
                         mergedColors,
                         corFnc = "bicor", # bimidweight correlation
                         power=sft$powerEstimate, # 4
                         # power = 4,
                         type = "signed")
hub_genes <- gene_annotation[hubs, ]
hub_genes$module <- names(hubs)
row.names(hub_genes) <- NULL
hub_genes <- hub_genes %>% dplyr::select(module, everything())
```
```{r sig_hubs}
hub_genes <- hub_genes %>%
  # filter(module %in% sigModules) %>%
  dplyr::select(module, ID, Gene_Info) %>% 
  dplyr::rename("hub gene" = ID, "UniProt best match" = Gene_Info) 
```
### Supplementary Table 9
```{r Table_S9}
# hub_genes_coral <- hub_genes
# hub_genes_symbiont <- hub_genes
# Bind together coral and symbiont DEG contrast results
# hub_genes_combined <- bind_rows("Pocillopora" = hub_genes_coral, "Cladocopium" = hub_genes_symbiont, .id = "Holobiont Partner")
# Write output to file
write_csv(hub_genes, path = "./outputs/WGCNA_results/hub_genes.csv")
write_csv(hub_genes, path = "./outputs/WGCNA_results/Table_S3_hub_genes.csv")
```
```{r datkME}
#this section is adapted from Rachel Wright's GitHub: https://github.com/rachelwright8/Ahya-White-Syndromes/
colnames(MEs1) <- MEmodule_colors
datKME <- signedKME(datExpr, MEs1, outputColumnName = "") %>% rownames_to_column(., "ID")
genecolors <- data.frame(ID = Genes, moduleColor = moduleColors)
genecolors_kME <- left_join(genecolors, datKME, by = "ID") 
###
```
```{r kME_files}
partner <- "coral"
# Assemble input for WGCNA GO_MWU: dataframe with gene IDs and kME values for all genes within a module, 0 for all genes without
for (module in uniqModules){
modkME <- genecolors_kME %>% dplyr::select(., "ID", "moduleColor", module)#paste("kME", module, sep = ""))
colnames(modkME) <- c("gene", "moduleColor", "kME")
###
modkME$kME[modkME$moduleColor!=module] <- 0  #also need to change column names for color you pick above
# modkME$kME[modkME$moduleColor==module] <- 1 #leave commented out to keep kME values
modkME <- modkME %>% dplyr::select(., -moduleColor)
# write csv
#write.csv(modkME, file = paste("./outputs/WGCNA-results/tables/", module,"_kme.csv",sep=""), quote=F, row.names=F)
write.csv(modkME, file = paste("./R/GO_MWU/", module,"_kme_", partner, ".txt", sep=""), quote=F, row.names=F)
#repeat for each module
}
```
### Module eigengene correlations
For this study, I can correlate against physiological data and bacteria community data.
It's a shame there's so few modules, but they all correlate significantly to experimental traits and physiology and microbiome data!
```{r treatment_trait_data}
# Get sample data into tibble
datTraits <- coldata %>%
  rownames_to_column("Sample_ID") %>%
  dplyr::select(Sample_ID) %>% 
  mutate(Control = ifelse(coldata$Treatment == "Control", 1, 0),
         Antibiotics = ifelse(coldata$Treatment == "Antibiotics", 1, 0)  #,
         # Panama = ifelse(coldata$Gulf == "Panama", 1, 0),
         # Chiriqui = ifelse(coldata$Gulf == "Chiriqui", 1, 0),
         # Cladocopium = ifelse(coldata$Symbiont_Genus == "Cladocopium", 1, 0),
         # Durusdinium = ifelse(coldata$Symbiont_Genus == "Durusdinium", 1, 0),
         # I = ifelse(coldata$haplotype == "I", 1, 0), 
         # III = ifelse(coldata$haplotype == "III", 1, 0), 
         # V = ifelse(coldata$haplotype == "V", 1, 0), 
         # VI = ifelse(coldata$haplotype == "VI", 1, 0)
) %>%
  column_to_rownames(var = "Sample_ID")
```
```{r physiology_trait_data}
ipam_traits <- read.csv("./outputs/physio/ipam_wgcna_traits.csv") %>%
   arrange(Sample) %>% 
  dplyr::rename("Fv/Fm (Day 5)" = mean_value) %>% 
  column_to_rownames("Sample")
datTraits_Ipam <- ipam_traits
#
respR_traits <- read.csv("./outputs/physio/respR/respR_wgcna_traits.csv") %>% 
  arrange(Sample) %>% 
  dplyr::select(-`rate`, -`Mass`) %>% 
  dplyr::rename("MO2 (Day 5)" = rate_norm) %>% # (MO[2])
  column_to_rownames("Sample")
datTraits_Resp <- respR_traits 
#
datTraits_Phys <- cbind(datTraits_Ipam, datTraits_Resp) # incorrect sample order - fixed!
```
```{r bacteria_taxa_abundance_trait_data}
# Combined analysis linking WGCNA modules and bacteria relative abundance
# # read in table with relative abundances of interesting bacteria taxa
datTaxa_Class <- read_csv("./outputs/phyloseq_results/datTaxa_Class.csv")
#
datTaxa_CoreASVs <- read_csv("./outputs/phyloseq_results/datTaxa_CoreASVs.csv")
#
#
#
```
```{r n_genes_samples}
# Define numbers of genes and samples
nGenes <- ncol(datExpr)
nSamples <- nrow(datExpr)
```
```{r ME_correlations_treatments}
# Correlate module eigengene-trait associations
moduleTraitCor <- cor(MEs1, datTraits, use = "p")
moduleTraitPvalue <- corPvalueStudent(moduleTraitCor, nSamples)
moduleTraitCors <- as.data.frame(moduleTraitCor, moduleTraitPvalue)
```
```{r ME_correlations_physiology}
modulePhysCor <- cor(MEs1, datTraits_Phys, use = "p")
modulePhysPvalue <- corPvalueStudent(modulePhysCor, nSamples)
modulePhysCors <- as.data.frame(modulePhysCor, modulePhysPvalue)
```
```{r ME_correlations_taxa}
moduleTaxaCor <- cor(MEs1, datTaxa_Class, use = "p")
moduleTaxaPvalue <- corPvalueStudent(moduleTaxaCor, nSamples)
moduleTaxaCors <- as.data.frame(moduleTaxaCor, moduleTaxaPvalue)
```
```{r ME_correlations_taxa}
moduleTaxaCor <- cor(MEs1, datTaxa_CoreASVs, use = "p")
moduleTaxaPvalue <- corPvalueStudent(moduleTaxaCor, nSamples)
moduleTaxaCors <- as.data.frame(moduleTaxaCor, moduleTaxaPvalue)
```
```{r significant_modules}
# Identify all unique and interesting modules based on trait correlations
# Filter for all the modules that have a significant correlation to any treatment, genotype, bacteria taxa, etc.
p_values <- cbind(moduleTraitPvalue, modulePhysPvalue)

modP <- as.data.frame(p_values) %>%
  rownames_to_column("module") %>%
  arrange(module) %>% 
  filter_all(any_vars(. < 0.05)) # %>% dplyr::select(module, lps)

#
sigModules <- sort(setdiff(modP$module,  "grey"))
# 
uniqModules
# would like to reorder the sigModules subset to match ME clustering dendrogram order
# sigModules <- uniqModules[uniqModules %in% sigModules]
# 
all(sigModules %in% uniqModules) # test that all the significant modules are actually within the modules
all(sigModules == uniqModules) # are all modules also significantly correlated to one or more traits? NO
sigModules
```
```{r significant_asvs}
moduleTaxaPvalue
#
modP_bacteria <- as.data.frame(moduleTaxaPvalue) %>%
  rownames_to_column("module") %>%
  arrange(module) %>% 
  filter_all(any_vars(. < 0.05)) %>% # select only modules correlated to one or more taxa
  select_if((~ any(. < 0.05))) # select only taxa that have a significantly correlated module (n = 20 / 37 ASVs)
#
wgcna_correlated_core <- colnames(modP_bacteria)
```

### Module gene significance (GS) and interconnectivity (kME)
```{r genesig}
# calculate gene trait significance for each treatment, genotype, etc.
geneTraitSignificance <- as.data.frame(cor(datExpr, datTraits, use = "p")) 
traitGS <- geneTraitSignificance %>% 
  rownames_to_column("ID")
```
```{r df_kME_GS}
longKME <- datKME %>% 
  # rename(gene = "ID") %>% 
  pivot_longer(cols= -ID, names_to = "module", values_to = "kME")
#
df_gene_GS_kME <- genecolors %>%
  # rename(gene = "ID") %>% 
  full_join(traitGS, by = "ID") %>% 
  full_join(longKME, by = "ID") %>%
  filter(moduleColor == module) %>%
  dplyr::select(-module)
```
```{r module_GS_kME_results_tables}
# write output tables with gene kME statistics and trait gene significance for each module
 for (module in uniqModules) {
# datKME <- datKME %>% column_to_rownames("gene")
df_mod_GS_kME <- df_gene_GS_kME %>%
  filter(moduleColor == module) %>% 
  dplyr::select(-moduleColor) %>% 
  dplyr::select(ID, kME, everything()) %>% 
  arrange(desc(kME)) %>% 
  left_join(., gene_annotation, by = "ID")
write_csv(df_mod_GS_kME, path = paste("./outputs/WGCNA_results/tables/", module, "_GS_kME.csv", sep = ""))
}
```

### WGCNA EuKaryotic Orthologous Groups (KOG) enrichment analysis
```{r choose_gene2kog}
# Ensure gene KOG annotation is present, choose partner
gene2kog <- gene2kog_coral
# gene2kog <- gene2kog_symbiont
```
```{r wgcna_kog}
# Create tibble to keep results of WGCNA module KOGMWU enrichment
# Choose whether to use Fisher's test for module membership or MWU test of enrichment of within-module kME ranks
modKOG <- tibble("module" = uniqModules)
modset <- uniqModules
for (i in seq_along(modset)) {
module <- modset[i]
modkME <- genecolors_kME %>% dplyr::select(., ID, moduleColor, module)
colnames(modkME) <- c("ID", "moduleColor", "kME")
###
modkME$kME[modkME$moduleColor!=module] <- 0  # also need to change column names for color you pick above
modkME$kME[modkME$moduleColor==module] <- 1 # leave uncommented for Fisher's test, comment out to keep kME values for MWU test
modkME <- modkME %>% dplyr::select(., -moduleColor)
modKOG$kME[i] <- list("kME" = modkME)
###
kogmwu_module <- kog.mwu(modkME, gene2kog) #%>% arrange("KOG_Class")
modKOG$KOG[i] <- list("KOG" = kogmwu_module)
#names(modKOGlist[i]) <- module
}
modKOG$KOG
```
```{r module_kog_results}
modKOG 
```
```{r signed_p_function}
signedpadj <- function(x){-1*log(x, base = 10)}
signedpadj(0.05)
```
```{r KOG_Fishers_purrr}
# Only use this section to extract results from Fisher's exact tests in KOGMWU
# Generate all KOG adjusted p-values (Fisher's test does not calculate delta rank)
modkogtables_fisher <- modKOG %>%
  unnest(KOG) %>%
  dplyr::select(module, term, padj) %>% 
  mutate(signedpadj = -log(padj, base = 10)) %>% 
  dplyr::select(-padj) %>%
  split(list(.$module), drop = TRUE) %>% 
  map(~ spread(., module, signedpadj)) %>% 
  map(~ column_to_rownames(., "term"))

# Generate KOG adjusted p-value stars
modkogpvals_fisher <- modKOG %>%
  unnest(KOG) %>%
  dplyr::select(module, term, padj) %>%
  split(list(.$module), drop = TRUE) %>% 
  map(~ spread(., module, padj)) %>%
  map(~ column_to_rownames(., "term")) %>%
  map(~ transmute_all(., gtools::stars.pval))
# 
modkogtable_fisher <- modKOG %>%
  unnest(KOG) %>%
  dplyr::select(module, term, padj) %>%
  mutate(signedpadj = -log(padj, base = 10)) %>% 
  dplyr::select(-padj) %>%
  spread(module, signedpadj) %>%
   column_to_rownames("term") 
write_csv(modkogtable_fisher, file = "./outputs/WGCNA_results/KOGMWU/modkogtable_fisher_coral.csv")
  
modkogpadj_fisher <- modKOG %>%
  unnest(KOG) %>%
  dplyr::select(module, term, padj) %>% 
  spread(module, padj) %>%
  column_to_rownames("term") %>%
  transmute_all(gtools::stars.pval)
```
```{r KOG_MWU_purrr}
# Only use this section to extract results from gene kME MWU tests in KOGMWU
modKOG <- modKOG %>% mutate(
         # If a KOG term has <10 genes associated with it, set delta rank to zero
         KOG = map(KOG, ~ mutate(., delta.rank = ifelse(nseqs < 10, 0, delta.rank))))

# Generate all KOG delta rank tables
modkogtables_mwu <- modKOG %>%
  unnest(KOG) %>%
  dplyr::select(module, term, delta.rank) %>% 
  split(list(.$module), drop = TRUE) %>% 
  map(~ spread(., module, delta.rank)) %>% 
  map(~ column_to_rownames(., "term"))

# Generate KOG adjusted p-value stars
modkogpvals_mwu <- modKOG %>%
  unnest(KOG) %>%
  dplyr::select(module, term, padj) %>%
  split(list(.$module), drop = TRUE) %>% 
  map(~ spread(., module, padj)) %>%
  map(~ column_to_rownames(., "term")) %>%
  map(~ transmute_all(., gtools::stars.pval))

#
modkogtable_mwu <- modKOG %>%
  unnest(KOG) %>%
  dplyr::select(module, term, padj) %>%
  mutate(signedpadj = -log(padj, base = 10)) %>% 
  dplyr::select(-padj) %>%
  spread(module, signedpadj) #%>%
  # column_to_rownames("term") 
write_csv(modkogtable_mwu, file = "./outputs/WGCNA_results/KOGMWU/modkogtable_mwu_coral.csv")
  
modkogpadj_mwu <- modKOG %>%
  unnest(KOG) %>%
  dplyr::select(module, term, padj) %>% 
  spread(module, padj) %>%
  column_to_rownames("term") %>%
  transmute_all(gtools::stars.pval)
```

### WGCNA Gene Ontology (GO) enrichment analysis
```{r module_GO_MWU}
# Run GO_MWU for any given module for Biological Process (BP) GO terms
commandArgs <- function(...) c("green_kme_symbiont.txt", "BP", "cladocopium_gene2go.tab") #"pdam_emapper_gene2go.tab"
source("./R/GO_MWU/GO_MWU_WGCNA.R")

# Run GO_MWU for Heat vs. control for Molecular Function GO terms
commandArgs <- function(...) c("brown_kme.txt", "MF", "pdam_emapper_gene2go.tab") #"cladocopium_gene2go.tab"
source("./R/GO_MWU/GO_MWU_WGCNA.R")

# Run GO_MWU for Heat vs. control for Cellular Component (CC) GO terms
commandArgs <- function(...) c("salmon_kme.txt", "CC", "pdam_emapper_gene2go.tab")
source("./R/GO_MWU/GO_MWU.R")
```
```{r module GO_MWU plot}
# Plot GO_MWU results
commandArgs <- function(...) c("salmon_kme.txt", "BP", "GO_MWU.salmon.png")
source("./R/GO_MWU/GO_MWU_WGCNA_plot.R")
commandArgs <- function(...) c("salmon_kme.txt", "MF", "GO_MWU.salmon.png")
source("./R/GO_MWU/GO_MWU_WGCNA_plot.R")
commandArgs <- function(...) c("salmon_kme.txt", "CC", "GO_MWU.salmon.png")
source("./R/GO_MWU/GO_MWU_WGCNA_plot.R")
```
```{r module_GO_MWU_loop}
kMEfiles <- list.files(path = "./R/GO_MWU/", pattern = "_kme_coral.txt$")
kMEfiles 
###
intModules_coral <- c("greenyellow","blue", "lightyellow", "darkorange",  "orange", "brown", "darkgrey", "cyan", "lightcyan", "tan", "darkgreen", "darkred", "darkturquoise", "black", "yellow", "royalblue", "magenta")
# intModules_symbiont <- c("green", "brown", "grey60", "magenta", "darkgrey", "darkturquoise", "purple", "darkgreen")
#
for (module in intModules_coral){
  print(module)
### command for GO_MWU analysis
  #input file name
tab <- paste(module, "_kme_", partner, ".txt", sep = "")
  print(tab)
  commandArgs <- function(...) c(tab, "MF", "pdam_emapper_gene2go.tab")
  source("./R/GO_MWU/GO_MWU_WGCNA.R")
}
```
```{r module_GO_results}
# find results files for GO_MWU enrichments
module_fisher_file <- "./R/GO_MWU/MWU_BP_greenyellow_kme_coral.txt"
module_fisher_file
module_fisher <- read.table(module_fisher_file, header = T)
module_fisher %>%
  filter(p.adj < 0.1) %>%
  View()
```
```{r all_modules_GO_results}
module_mwu_files <- str_c("./R/GO_MWU/MWU_", "BP_", intModules_coral, "_kme_", partner, ".txt", sep = "")
names(module_mwu_files) <- intModules_coral
module_mwu_files
# read in GO_MWU results
modules_GO <- as.list(module_mwu_files) %>%
  map(read.table, header = T) %>%
  bind_rows(.id = "module") %>%
  as_tibble()
```
```{r}
# save module enrichment
modules_GO_coral <- modules_GO
```
```{r}
# filter only significantly enriched GO terms
module_GO_terms <- modules_GO %>%
  dplyr::select(module, delta.rank, p.adj, name, nseqs, level, term) %>%
  filter(module %in% intModules_coral) %>%
  filter(p.adj < 0.05)
#
module_GO_terms$module <- factor(module_GO_terms$module, levels = intModules_coral, ordered = TRUE)
#
module_GO_terms %>% 
  arrange(module, p.adj) %>% 
  write_excel_csv(path = "./outputs/WGCNA_results/GO_MWU/modules_GO_coral.csv")
#
module_GO_counts <- modules_GO %>%
  filter(p.adj < 0.05) %>%
  # filter(module %in% intModules) %>%
  dplyr::count(module) %>% 
  arrange(desc(n))
# 
enrichedModules <- module_GO_counts$module
```
### Supplementary Table 10
```{r}
module_GO_terms %>% 
  arrange(module, level, p.adj) %>% 
  write_excel_csv(path = "./outputs/WGCNA_results/Table_S10_modules_GO_coral.csv")
```

```{r GO_MWU_figures}
# keyModules <- c("darkgrey", "midnightblue", "darkorange", "darkgreen", "ivory")
module <- "brown"
# for (module in enrichedModules){
###
tab <- paste(module, "_kme_", partner, ".txt", sep = "")
fig <- as.character(paste("GO_MWU_", module, ".pdf", sep = ""))
# levs <- c(0.05, 1e-10, 1e-15)
print(module)
##
commandArgs <- function(...) c(tab, "BP", fig, module, 0.05, 0.01, 0.001, 1.0)
source("./R/GO_MWU/GO_MWU_WGCNA_plot.R")
# }
```

```{r}

```


### WGCNA correlation heatmaps
```{r heatmap_tests}
Heatmap(moduleTraitCor)
Heatmap(modulePhysCor)
Heatmap(moduleTaxaCor)
```
#### Create heatmap annotations
```{r ComplexHeatmap_annotation_WGCNA_modules}
# row labels as module names and number of genes in parentheses
module_labels <- data.frame(table(moduleColors)) %>%
  filter(moduleColors != "grey") %>%
  # arrange(desc(Freq)) %>% 
  dplyr::mutate(freq_string = str_c("(", Freq ,")")) %>%
  dplyr::select(-Freq) %>% 
  unite(col = "module_label", moduleColors, freq_string, sep = "  ")
module_labels <- as.character(module_labels$module_label)

# annotation to include rectangles with module colors
moddf <- data.frame(module_colors)
module_anno  <- module_colors
names(module_anno) <- module_colors
module_sizes <- data.frame(table(moduleColors)) %>% filter(moduleColors != "grey") %>% arrange(desc(Freq))
colnames(module_sizes) <- c("module_colors", "module_size")
moddf <- left_join(moddf, module_sizes, by = "module_colors")
module_annotation <- HeatmapAnnotation(modules = moddf$module_colors,
                                       col = list(modules = module_anno),
                                       which = "row",
                                       annotation_name_rot = 0,
                                       gp = gpar(col = "black", lwd = 0.5),
                                       show_legend = FALSE,
                                       show_annotation_name = FALSE
                                       )
#
cn = colnames(moduleTraitCor)
```
```{r ComplexHeatmap_color_lists}
# Create colony and condition annotation colors list
anno_colors <- list(
  Treatment = c(Control = treatcolors[2],  Antibiotics = treatcolors[3])#,
  # Bacteria = bac_colors
  )
```
```{r ComplexHeatmap_annotation_traits_phys_taxa}
# ComplexHeatmap colors
# col_fun_ge <- function(mat) {
#   colorRamp2(c(max(mat), max(mat)/2, mean(mat), min(mat)/2, min(mat)), c("red", "yellow", "grey10", "cyan", "blue"))
# }
col_fun_cor <- colorRamp2(c(1, 0, -1), c("red", "white", "blue"))
# col_fun_KOG_fisher <- colorRamp2(c(80, 0), c("red", "white"))
# ComplexHeatmap sample annotations
# sample_annotation <- HeatmapAnnotation(Genotype = samples$Genotype,
#                                        Treatment = samples$Treatment,
#                                        col = anno_colors,
#                                        annotation_name_rot = 0,
#                                        show_legend = FALSE
#                                        )
# column annotation for treatments, physiology, and bacteria
treatment_annotation <- HeatmapAnnotation(Treatment = c("Control", "Antibiotics"),
                                          col = anno_colors[1],
                                          gp = gpar(col = "black", lwd = 0.5),
                                          show_legend = FALSE,
                                          show_annotation_name = FALSE
                                          )
phys_annotation <- HeatmapAnnotation(Metric = c("Fv/Fm", "Mass", "Resp. Rate", "Norm. Rate"),
                                          # col = anno_colors[1],
                                          gp = gpar(col = "black", lwd = 0.5),
                                          show_legend = FALSE,
                                          show_annotation_name = FALSE
                                          )

# bacteria_annotation <- HeatmapAnnotation(Family = bacteria_colors$Family,
#                                           col = anno_colors[4],
#                                           gp = gpar(col = "black", lwd = 0.5),
#                                           show_legend = FALSE,
#                                           show_annotation_name = FALSE
#                                           )
# anno_colors[4]

```
#### Create text matrices
```{r treatment_text_matrix}
# Create text matrix to display module-treatment correlations and their p-values
# Only displays correlation coefficients for significant correlations p < 0.05
textMatrix_treatment <- character()
for (i in 1:length(moduleTraitPvalue)){
  # reviewer request: add p-value to cell text
  if(moduleTraitPvalue[i] < 0.05){ text <- paste(signif(moduleTraitCor[i], 2),
                                                "\n", "(",
                                                 signif(moduleTraitPvalue[i], 2),
                                                 ")", sep = "")} else { text <- "" }
  textMatrix_treatment[i] <- text
}
dim(textMatrix_treatment) <- dim(moduleTraitCor)
# textMatrix_treatment
```
```{r physiology_text_matrix}
# Create text matrix to display module-treatment correlations and their p-values
# Only displays correlation coefficients for significant correlations p < 0.05
textMatrix_phys <- character()
for (i in 1:length(modulePhysPvalue)){
  # reviewer request: add p-value to cell text
  if(modulePhysPvalue[i] < 0.05){ text <- paste(signif(modulePhysCor[i], 2),
                                                "\n", "(",
                                                 signif(modulePhysPvalue[i], 2),
                                                 ")", sep = "")} else { text <- "" }
  textMatrix_phys[i] <- text
}
dim(textMatrix_phys) <- dim(modulePhysCor)
# textMatrix_phys
```
```{r bacteria_text_matrix}
# Create text matrix to display module-treatment correlations and their p-values
# Only displays correlation coefficients for significant correlations p < 0.05
textMatrix_taxa <- character()
for (i in 1:length(moduleTaxaPvalue)){
  # reviewer request: add p-value to cell text
  if(moduleTaxaPvalue[i] < 0.05){ text <- paste(signif(moduleTaxaCor[i], 2),
                                                "\n", "(",
                                                 signif(moduleTaxaPvalue[i], 2),
                                                 ")", sep = "")} else { text <- "" }
  textMatrix_taxa[i] <- text
}
dim(textMatrix_taxa) <- dim(moduleTaxaCor)
# textMatrix_taxa
```
#### Create individual heatmaps
```{r ComplexHeatmap_treatment}
# pdf(file = "./outputs/figures/ME_complexheatmap_treatments.pdf", width = 6.5, height = 5)
# 
me_treatment_heatmap <- Heatmap(moduleTraitCor,
                      col = col_fun_cor,
                      cell_fun = function(j, i, x, y, width, height, fill) {grid.text(sprintf("%s", textMatrix_treatment[i, j]), x, y, gp = gpar(fontsize = 4))},
                      cluster_rows = hclustME,
                      # top_annotation = treatment_annotation,
                      left_annotation = module_annotation,
                      # right_annotation = module_barplot,
                      # bottom_annotation = HeatmapAnnotation(text = anno_text(cn,
                      #                                                        rot = 330,
                      #                                                        offset = unit(1, "npc"),
                      #                                                        just = "left"),
                      #                                       annotation_height = max_text_width(cn)),
                      # row_split = 3,
                      bottom_annotation = treatment_annotation,
                      border = FALSE,
                      rect_gp = gpar(col = "black", lwd = 0.5),
                      column_order = colnames(datTraits),
                      show_column_names = TRUE,
                      column_names_rot = 90,
                      column_names_side = "bottom",
                      column_names_centered = F,
                      column_names_gp = gpar(fontsize = 8, fontface = "plain"),
                      # column_title = "Treatments",
                      # column_title_gp = gpar(fontsize = 12, fontface = "plain"),
                      # row_title = "Cluster %s",
                      row_labels = module_labels,
                      row_title_gp = gpar(fontsize = 10, fontface = "plain"),
                      row_dend_width = unit(12, "mm"),
                      row_names_side = "left",
                      row_names_gp = gpar(fontsize = 8, fontface = "plain", just = "left"),
                      row_title_rot = 0,
                      show_heatmap_legend = FALSE,
                      heatmap_legend_param = list(title = "Module eigengene correlation",
                                                  direction = "horizontal",
                                                  # title_position = "leftcenter-rot",
                                                  legend_height = unit(1.0, "cm"),
                      width = unit(2, "cm"))
                      )
draw(me_treatment_heatmap,
     show_annotation_legend = FALSE, heatmap_legend_side = "left"
     )
# dev.off()
```
```{r ComplexHeatmap_physiology}
# pdf(file = "./outputs/figures/ME_complexheatmap_physiology.pdf", width = 6.5, height = 5)
# 
me_physiology_heatmap <- Heatmap(modulePhysCor,
                      col = col_fun_cor,
                      cell_fun = function(j, i, x, y, width, height, fill) {grid.text(sprintf("%s", textMatrix_phys[i, j]), x, y, gp = gpar(fontsize = 4))},
                      cluster_rows = hclustME,
                      # top_annotation = treatment_annotation,
                      # left_annotation = module_annotation,
                      # right_annotation = module_barplot,
                      # bottom_annotation = HeatmapAnnotation(text = anno_text(cn,
                      #                                                        rot = 330,
                      #                                                        offset = unit(1, "npc"),
                      #                                                        just = "left"),
                      #                                       annotation_height = max_text_width(cn)),
                      # row_split = 3,
                      border = FALSE,
                      rect_gp = gpar(col = "black", lwd = 0.5),
                      column_order = colnames(datTraits_Phys),
                      show_column_names = TRUE,
                      column_names_rot = 90,
                      column_names_side = "bottom",
                      column_names_centered = F,
                      column_names_gp = gpar(fontsize = 8, fontface = "plain"),
                      # column_title = "Treatments",
                      # column_title_gp = gpar(fontsize = 12, fontface = "plain"),
                      # row_title = "Cluster %s",
                      row_labels = module_labels,
                      row_title_gp = gpar(fontsize = 10, fontface = "plain"),
                      row_dend_width = unit(1.25, "in"),
                      row_names_side = "left",
                      row_names_gp = gpar(fontsize = 8, fontface = "plain", just = "left"),
                      row_title_rot = 0,
                      show_heatmap_legend = FALSE,
                      heatmap_legend_param = list(title = "Module eigengene correlation",
                                                  direction = "horizontal",
                                                  # title_position = "leftcenter-rot",
                                                  legend_height = unit(1.0, "cm"),
                      width = unit(2, "cm"))
                      )
draw(me_physiology_heatmap,
     show_annotation_legend = FALSE, heatmap_legend_side = "left"
     )
# dev.off()
```
```{r ComplexHeatmap_bacteria_classes}
## Need to get phylogenetic tree of classes/families/etc. above column rows

# pdf(file = "./outputs/figures/ME_complexheatmap_bacteria.pdf", width = 11, height = 5)
# 
me_bacteria_heatmap <- Heatmap(moduleTaxaCor,
                      col = col_fun_cor,
                      cell_fun = function(j, i, x, y, width, height, fill) {grid.text(sprintf("%s", textMatrix_taxa[i, j]), x, y, gp = gpar(fontsize = 4))},
                      cluster_rows = hclustME,
                      # cluster_columns = classes_dendrogram,
                      # top_annotation = treatment_annotation,
                      # left_annotation = module_annotation,
                      # right_annotation = module_barplot,
                      # bottom_annotation = HeatmapAnnotation(text = anno_text(cn,
                      #                                                        rot = 330,
                      #                                                        offset = unit(1, "npc"),
                      #                                                        just = "left"),
                      #                                       annotation_height = max_text_width(cn)),
                      # row_split = 3,
                      border = FALSE,
                      rect_gp = gpar(col = "black", lwd = 0.5),
                      column_order = colnames(datTaxa_Class),
                      column_dend_height = unit(20, "mm"),
                      show_column_names = TRUE,
                      column_names_rot = 270,
                      column_names_side = "bottom",
                      column_names_centered = T,
                      column_names_gp = gpar(fontsize = 8, fontface = "plain"),
                      # column_title = "Treatments",
                      # column_title_gp = gpar(fontsize = 12, fontface = "plain"),
                      # row_title = "Cluster %s",
                      row_labels = module_labels,
                      row_title_gp = gpar(fontsize = 10, fontface = "plain"),
                      row_dend_width = unit(1.25, "in"),
                      row_names_side = "left",
                      row_names_gp = gpar(fontsize = 10, fontface = "plain", just = "left"),
                      row_title_rot = 0,
                      show_heatmap_legend = FALSE,
                      heatmap_legend_param = list(title = "Module eigengene correlation",
                                                  direction = "horizontal",
                                                  # title_position = "leftcenter-rot",
                                                  legend_height = unit(1.0, "cm"),
                      width = unit(2, "cm"))
                      )
draw(me_bacteria_heatmap,
     show_annotation_legend = FALSE, heatmap_legend_side = "left"
     )
# dev.off()
```
```{r ComplexHeatmap_bacteria_ASV}
## Need to get phylogenetic tree of classes/families/etc. above column rows

pdf(file = "./outputs/figures/ME_complexheatmap_bacteria_ASVs_core_correlated_phylogenetic.pdf", width = 11, height = 8.5)
# 
me_bacteria_heatmap <- Heatmap(moduleTaxaCor,
                      col = col_fun_cor,
                      cell_fun = function(j, i, x, y, width, height, fill) {grid.text(sprintf("%s", textMatrix_taxa[i, j]), x, y, gp = gpar(fontsize = 4))},
                      cluster_rows = hclustME,
                      cluster_columns = clustASVs, # need to add phylogenetic tree for select ASVs here
                      # top_annotation = treatment_annotation,
                      # left_annotation = module_annotation,
                      # right_annotation = module_barplot,
                      # bottom_annotation = HeatmapAnnotation(text = anno_text(cn,
                      #                                                        rot = 330,
                      #                                                        offset = unit(1, "npc"),
                      #                                                        just = "left"),
                      #                                       annotation_height = max_text_width(cn)),
                      # row_split = 3,
                      border = FALSE,
                      rect_gp = gpar(col = "black", lwd = 0.5),
                      # column_order = colnames(datTaxa_Class),
                      column_dend_height = unit(20, "mm"),
                      show_column_names = TRUE,
                      column_names_rot = 30,
                      column_names_side = "bottom",
                      column_names_centered = F,
                      column_names_gp = gpar(fontsize = 6, fontface = "plain"),
                      # column_title = "Treatments",
                      # column_title_gp = gpar(fontsize = 12, fontface = "plain"),
                      # row_title = "Cluster %s",
                      row_labels = module_labels,
                      row_title_gp = gpar(fontsize = 10, fontface = "plain"),
                      row_dend_width = unit(1.25, "in"),
                      row_names_side = "left",
                      row_names_gp = gpar(fontsize = 10, fontface = "plain", just = "left"),
                      row_title_rot = 0,
                      show_heatmap_legend = FALSE,
                      heatmap_legend_param = list(title = "Module eigengene correlation",
                                                  direction = "horizontal",
                                                  # title_position = "leftcenter-rot",
                                                  legend_height = unit(1.0, "cm"),
                      width = unit(2, "cm"))
                      )
draw(me_bacteria_heatmap,
     show_annotation_legend = FALSE, heatmap_legend_side = "left"
     )
dev.off()
```
#### Concatenate heatmaps
```{r}
# Concatenate heatmaps
me_treatment_heatmap + me_physiology_heatmap + me_bacteria_heatmap
```
### Figure 5
```{r}
# Complex heatmap with treatment, symbiont type, and physiology correlations, plus bacteria taxa correlations
heatmap_combined <- me_treatment_heatmap + me_physiology_heatmap + me_bacteria_heatmap
# combine with enriched GO term word clouds?
```
```{r Figure_5_correlation_legend}
cor_lgd <- Legend(col_fun = col_fun_cor,
         title = "Module correlation",
         title_gp = gpar(fontsize = 8, fontface = "plain"),
         at = c(-1, 0, 1),
         direction = "horizontal",
         title_position = "topcenter",
         legend_width = unit(25, "mm"),
         grid_width = unit(2.5, "mm"),
         labels_gp = gpar(fontsize = 8, fontface = "plain"),
         border = "black")
```
```{r}
pdf("./outputs/figures/Fig5_wgcna_complexheatmap_mean3_merge0.3.pdf", width = 136/25.4, height = 136/25.4)
draw(heatmap_combined,
     ht_gap = unit(1, "mm"),
     show_annotation_legend = FALSE, heatmap_legend_side = "left"
     )
draw(cor_lgd,
     x = unit(12, "mm"), y = unit(12, "mm"), just = c("left", "bottom"))
dev.off()
```


### WGCNA KOG enrichment heatmap
```{r ME_KOG_ComplexHeatmap_annotation}
# annotation to include module colors
module_annotation_col <- HeatmapAnnotation(modules = moddf$module_colors,
                                       col = list(modules = module_anno),
                                       which = "col",
                                       annotation_name_rot = 0,
                                       show_legend = FALSE
                                       )
# text matrix annotation for cell text
 cell_fun <- function(j, i, x, y, width, height, fill) {
        grid.text(sprintf("%.1f", t(modkogpadj_fisher)[i, j]), x, y, gp = gpar(fontsize = 6))
 }
 #
col_fun_KOGfisher <- colorRamp2(c(0, 14), c("white", "turquoise3"))
#
kog_lgd <- Legend(col_fun = col_fun_KOGfisher,
         title = "KOG enrichment -log(padj)",
         title_gp = gpar(fontsize = 8, fontface = "plain"),
         at = c(0, 14),
         direction = "horizontal",
         title_position = "topcenter",
         legend_width = unit(25, "mm"),
         grid_width = unit(2.5, "mm"),
         labels_gp = gpar(fontsize = 8, fontface = "plain"),
         border = "black")
```
```{r ME_KOG_ComplexHeatmap}
me_kog_heatmap_fisher <- Heatmap(t(modkogtable_fisher),
                      col = col_fun_KOGfisher,
                      cell_fun = function(j, i, x, y, width, height, fill) {grid.text(sprintf("%s", t(modkogpadj_fisher)[i, j]), x, y, gp = gpar(fontsize = 8))},
                      cluster_rows = hclustME,
                      # right_annotation = module_barplot,
                      row_split = NULL,
                      border = FALSE,
                      rect_gp = gpar(col = "black", lwd = 0.5),
                      # column_order = rownames(modkogtable_fisher),
                      column_dend_height = unit(0.5, "cm"),
                      column_names_rot = 90,
                      column_names_side = "bottom",
                      column_names_centered = F,
                      column_names_gp = gpar(fontsize = 6, fontface = "plain", vjust = 0, hjust = 0),
                      # column_title = "KOG enrichments",
                      # column_title_gp = gpar(fontsize = 10, fontface = "plain"),
                      show_row_names = FALSE,
                      # row_names_gp = gpar(fontsize = 10, fontface = "plain"),
                      show_heatmap_legend = FALSE,
                      heatmap_width = unit(100/25.4, "in"),
                      heatmap_legend_param = list(title = "KOG class -log(padj)",
                                                  direction = "vertical",
                                                  title_position = "leftcenter-rot",
                                                  legend_height = unit(1.0, "cm"),
                       width = unit(4, "cm")
                      ))
# 
draw(me_kog_heatmap_fisher, 
     show_annotation_legend = FALSE, heatmap_legend_side = "left"
     )
```
```{r ComplexHeatmap_Fisher}
# pdf("./outputs/WGCNA-results/figures/modules_KOGenrichment_heatmap.pdf", height = 8.5, width = 10)
kog_heatmap_fisher <- Heatmap(t(modkogtable_fisher),
                      col = col_fun_KOGfisher,
                      cell_fun = function(j, i, x, y, width, height, fill) {grid.text(sprintf("%s", t(modkogpadj_fisher)[i, j]), x, y, gp = gpar(fontsize = 6))},
                      cluster_rows = hclustME,
                      # right_annotation = module_barplot,
                      # row_split = NULL,
                      # border = FALSE,
                      # rect_gp = gpar(col = "white", lwd = 1.5),
                      # column_title_gp = gpar(fontsize = 8, fontface = "plain"),
                      # column_order = rownames(modkogtable_fisher),
                      column_names_rot = 270,
                      column_names_side = "bottom",
                      # column_names_centered = TRUE,
                      column_names_gp = gpar(fontsize = 8, fontface = "plain"),
                      # row_names_gp = gpar(fontsize = 8, fontface = "plain"),
                      # show_heatmap_legend = TRUE,
                      heatmap_legend_param = list(title = "KOG enrichment -log(p.adj)",
                                                  direction = "vertical",
                                                  title_position = "leftcenter-rot",
                                                  legend_height = unit(1.0, "in"),
                       height = unit(4, "in"),
                       width = unit(4, "in")
                      ))
draw(kog_heatmap_fisher, column_title = "Module KOG enrichment heatmap", 
     show_annotation_legend = FALSE, heatmap_legend_side = "left"
     )
```

### Updated Figure 5
```{r}
pdf("./outputs/figures/Fig5_wgcna_complexheatmap_mean3_merge0.3_KOG.pdf", width = 176/25.4, height = 150/25.4)
heatmap_kog_combined <- me_treatment_heatmap + me_physiology_heatmap + me_kog_heatmap_fisher
draw(heatmap_kog_combined,
     ht_gap = unit(1, "mm"),
     show_annotation_legend = FALSE, heatmap_legend_side = "left"
     )
draw(cor_lgd,
     x = unit(12, "mm"), y = unit(48, "mm"), just = c("left", "bottom"))
draw(kog_lgd,
     x = unit(12, "mm"), y = unit(32, "mm"), just = c("left", "bottom"))
dev.off()
```

# Exploratory Analysis

### Mega-Heatmap: The Forbidden Figure
```{r}
me_treatment_heatmap + me_physiology_heatmap + me_bacteria_heatmap + me_kog_heatmap_fisher
```

